---
title: "Supplementary material"
geometry: left=2cm,right=2cm,top=2cm,bottom=2cm
output:
  bookdown::pdf_document2:
    toc: true
    lot: true
    lof: true
    toc_depth: 3
    keep_tex: yes
    df_print: kable
    latex_engine: xelatex
  word_document: default
  html_document: default
  pdf_document: bookdown::pdf_document2
mainfont: Times New Roman
monofont: Courier
urlcolor: blue
bibliography: Methods_Softwares.bib
# bib-humanities: yes
biblio-style: authoryear
header-includes:
- \usepackage{fvextra}
- \usepackage{caption}
- \usepackage[justification=raggedright,labelfont=bf,singlelinecheck=false]{caption}
- \DefineVerbatimEnvironment{Highlighting}{Verbatim}{breaksymbolleft={},showspaces
  = false,showtabs = false,breaklines,commandchars=\\\{\}}
- \renewcommand{\figurename}{Figure S}
- \renewcommand{\tablename}{Table S}
- \makeatletter
- \def\fnum@figure{\figurename\thefigure}
- \def\fnum@table{\tablename\thetable}
- \makeatother
- \usepackage{pdflscape}
- \newcommand{\blandscape}{\begin{landscape}}
- \newcommand{\elandscape}{\end{landscape}}
- \usepackage{hyperref}
- \hypersetup{colorlinks=true,linkcolor=black,citecolor=black,urlcolor=blue}
- \usepackage[titles]{tocloft}
- \renewcommand{\cftdotsep}{2} 
- \renewcommand{\cftfigindent}{0em}  
- \renewcommand{\cftfignumwidth}{5em}  
- \renewcommand{\cftfigpresnum}{Figure S} 
- \renewcommand{\cfttabindent}{0em}  
- \renewcommand{\cfttabnumwidth}{5em}
- \renewcommand{\cfttabpresnum}{Table S} 
# toc: yes
editor_options:
  chunk_output_type: console
---

\listoftables

\listoffigures

```{r setup, include=FALSE, message = FALSE, warning = FALSE}
# tinytex::parse_install('Supplementary_material_all.log')
# tinytex::parse_packages('Supplementary_material_all.log')
setwd("~/Desktop/deepsea_worms_epigenomes/downstream_analyses_scripts/")

# Load libraries
library(ggplot2)
library(broom)
library(dplyr)
library(knitr)
library(kableExtra)
library(readr)
library(tidyr)
library(reshape2)
library('formatR')
library(tinytex)
library(scales)
library(viridis)
library(ggpattern)
library(lvplot)
library(gtable)
library('egg')
library(scales)
library("Hmisc")
# install.packages("devtools", dependencies = TRUE)
# devtools::install_github("haozhu233/kableExtra")
# remotes::install_github('yihui/knitr')
options(tinytex.clean = FALSE)
knitr::opts_chunk$set(cache = F, echo = T, tidy.opts=list(width.cutoff=60),tidy=TRUE, fig.cap = "",fig.pos = 'H',dev = "png",dpi=300)

def.chunk.hook  <- knitr::knit_hooks$get("chunk")
knitr::knit_hooks$set(chunk = function(x, options) {
  x <- def.chunk.hook(x, options)
  paste0("\n \\", "normalsize","\n\n", x, "\n\n \\")
})

```

## Genome-wide methylation profile 

### methylation coverage vs methylation depth
```{r prep for fig cov-vs-depth, eval=FALSE, include=FALSE}
df=read.csv('../data/Genome-wide_methylation_profiles/all.windowavg1000.tsv',header=T,sep='\t')

df=rbind(sample_n(df[df$genome=='P. echinospica',], 10000),
         sample_n(df[df$genome=='R. piscesae',], 10000),
         sample_n(df[df$genome=='P. palmiformis',], 10000))
df %>% tail
df$binary=ifelse((df$meth_cov>0.75) & (df$meth_depth>0.50),1,0)

df$genome=factor(df$genome,levels=c('P. echinospica','R. piscesae','P. palmiformis'))
ggplot(df)+
  # geom_bin2d(aes(x=meth_cov*100,y=meth_depth*100))+
  # scale_fill_gradient(low='grey90',high='black')+
  geom_rect(xmin = 75,ymin = 50,xmax=105,ymax=105,color='purple',fill='#F9F5FF',alpha=0.1)+
  geom_point(aes(x=meth_cov*100,y=meth_depth*100),alpha=0.008)+
  labs(x="Methylation coverage (%)",y="Methylation depth (%)")+
  scale_x_continuous(limits=c(0,105))+
  scale_y_continuous(limits=c(0,105))+
  facet_grid(.~genome)+
  theme_bw()+
  theme(legend.position = 'none',
      strip.background = element_blank(),
      strip.text.x= element_text(face=c('bold.italic'),angle=0),
      aspect.ratio = 1:1)

ggsave("../figures/meth_cov_vs_depth_v3.png", width = 350, height = 150, units = "mm")

```


## LTR methylation
```{r compute LTR summary fig, eval=FALSE, warning=FALSE, include=FALSE}

datt=read.csv('../data/Transposable_elements/All.downstream.LTR.tsv',header=T,sep='\t')

datt %>% str
datt %>% group_by(genome) %>% summarise(count=n())

# a=datt[datt$genome=='R. piscesae','LTR']
# df=read.csv('/Volumes/4T_Omics/Tubeworms/Epig/LTR/clean_R07B-5_wtdbg2.fasta.pass.list.gff3',header=F,sep='\t')
# df %>% str 
# b=df[df$V3=="repeat_region",'V9'] %>% strsplit(.,';|=') %>% lapply(.,'[[',4) %>% unlist
# setdiff(b,a)


#### LTR type ####

dat=datt %>% dcast(.,LTR_type~genome) %>% melt(value.name = "count", variable.name="genome")

dat$LTR_family=ifelse(dat$LTR_type %in% c('LINE','Penelope','Maverick','mixture',''),'unclassified',dat$LTR_type)
dat$LTR_type=factor(dat$LTR_type,levels=c("unclassified", "Bel-Pao", "Copia", "DIRS", "Gypsy"))

dat=dat %>% group_by(genome,LTR_family) %>% summarise(count=sum(count))
dat$LTR_type=dat$LTR_family
dat=rbind(dat,dat[4,],dat[4,])

# setting up the data break in the figure
dat[4,'count']=145
dat[16,'count']=9
dat[16,'LTR_type']='Gypsy2'
dat[17,'count']=5
dat[17,'LTR_type']='Gypsyblank'

dat$LTR_type=factor(dat$LTR_type, levels=c( "unclassified","Gypsy2","Gypsyblank","Gypsy", 
"Bel-Pao", "DIRS", "Copia") )
dat$LTR_family=factor(dat$LTR_family, levels=rev(c("unclassified","Gypsy", "Bel-Pao", "DIRS", "Copia")))


type_colors=c('grey','#1b4965','white','#1b4965','#5fa8d3','#62b6cb','#cae9ff')
names(type_colors)=c("unclassified", "Gypsy",'Gypsyblank','Gypsy2',"Bel-Pao","DIRS", "Copia")


F1=ggplot(dat[dat$genome=='P. echinospica',])+
  geom_bar(aes(y=LTR_family,x=count,fill=LTR_type),stat='identity',position='stack')+
  scale_x_continuous(limits=c(0,160),breaks=c(0,50,100,150),labels = c(1,50,100,1800))+
  scale_fill_manual(values=type_colors,guide='none')+
  labs(title='P. echinospica',
        x ="Count", y = "LTR family")+
  facet_wrap(~genome)+
  theme_bw()+
  theme(strip.text.x = element_blank(),
        axis.text.x.top = element_blank(),
        axis.ticks.x.top = element_blank(),
        axis.text.x = element_text(angle=45,vjust=1,hjust=1),
        axis.title.x = element_blank(),
        axis.title.y = element_text(angle=0,vjust=0.5,hjust=0),
        strip.background = element_blank(),
        plot.title = element_text(face = 'bold.italic',hjust=0.6,size=rel(0.8))
        )

F2<-ggplot(dat[dat$genome=='R. piscesae',])+
  geom_bar(aes(y=LTR_family,x=count,fill=LTR_type),stat='identity')+
  scale_x_continuous(limits=c(0,160),breaks=c(0,50,100,150),labels = c(1,50,100,''))+
  facet_wrap(~genome)+
  scale_fill_manual(values=type_colors,guide='none')+
  labs(title='R. piscesae',
        y ="Count", x = "TE superfamily")+
  theme_bw()+
  theme(strip.text.x = element_blank(),
      axis.text.x.top = element_blank(),
      axis.ticks.x.top = element_blank(),
      axis.text.y = element_blank(),
      axis.title.y = element_blank(),
      axis.title.x = element_blank(),
      strip.background = element_blank(),
      plot.title = element_text(face = 'bold.italic',hjust=0.5,size=rel(0.8))
      )

F3=ggplot(dat[dat$genome=='P. palmiformis',])+
  geom_bar(aes(y=LTR_family,x=count,fill=LTR_type),stat='identity')+
  scale_fill_manual(values=type_colors,guide='none')+
  scale_x_continuous(limits=c(0,160),breaks=c(0,50,100,150),labels = c(1,50,100,''))+
  facet_wrap(~genome)+
  labs(title='P. palmiformis',
        y ="Count", x = "TE superfamily")+
  theme_bw()+
  theme(strip.text.x = element_blank(),
      axis.text.x.top = element_blank(),
      axis.ticks.x.top = element_blank(),
      axis.text.y = element_blank(),
      axis.title.y = element_blank(),
      axis.title.x = element_blank(),
      strip.background = element_blank(),
      plot.title = element_text(face = 'bold.italic',hjust=0.5,size=rel(0.8))
      )


#### insertion time ####

T1=ggplot(datt[datt$genome=='P. echinospica',])+
  geom_histogram(aes(x=insertion_time_my),binwidth = 0.2,color='black')+
  facet_wrap(~genome,scales='free_y')+
  theme_bw()+
  labs(x='LTR estimated\ninsertion time\n(Mya)',y='Count')+
  theme(strip.text = element_blank(),
      strip.background = element_blank(),
      axis.title.y = element_text(angle=0,vjust=0.5,hjust=0),
      axis.title.x = element_blank())+
  scale_x_reverse(limits=c(5,-0.1))+
  coord_flip()


T2=ggplot(datt[datt$genome=='R. piscesae',])+
  geom_histogram(aes(x=insertion_time_my),binwidth = 0.2,color='black')+
  facet_wrap(~genome,scales='free_y')+
  theme_bw()+
  labs(x='LTR estimated insertion time (Mya)',y='Count')+
  theme(strip.text = element_blank(),
      strip.background = element_blank(),
      axis.title.x = element_blank(),
      axis.title.y = element_blank())+
  scale_x_reverse(limits=c(5,-0.1))+
  coord_flip()
  

T3=ggplot(datt[datt$genome=='P. palmiformis',])+
  geom_histogram(aes(x=insertion_time_my),bins=20,color='black')+
  facet_wrap(~genome,scales='free_y')+
  theme_bw()+
  labs(x='LTR estimated insertion time (Mya)',y='Count')+
  theme(strip.text = element_blank(),
      strip.background = element_blank(),
      axis.title.x = element_blank(),
      axis.title.y = element_blank())+
  scale_x_reverse(limits=c(5,-0.1))+
  coord_flip()


#### location ####
datt %>% str

dat=datt %>% dcast(.,location~genome) %>% melt(value.name = "count", variable.name="genome")
dat$location=factor(dat$location, levels=rev(c('.','gene')),labels=rev(c('outside genes','within genes')))


cbind(dat,dat %>% group_by(genome) %>% summarise(prop=count/sum(count)))

L1=ggplot(dat[dat$genome=='P. echinospica',])+
  geom_bar(aes(y=location,fill=location,x=count),stat='identity')+
  theme_bw()+
  labs(y='Location')+
  scale_fill_discrete(guide='none')+
  theme(strip.text.x = element_blank(),
    axis.text.x.top = element_blank(),
    axis.ticks.x.top = element_blank(),
    axis.title.y = element_text(angle=0,vjust=0.5,hjust=0),
    axis.title.x = element_blank(),
    strip.background = element_blank(),
    plot.title = element_text(face = 'bold.italic',hjust=0.5,size=rel(0.8))
    )

L2=ggplot(dat[dat$genome=='R. piscesae',])+
  geom_bar(aes(y=location,fill=location,x=count),stat='identity')+
  theme_bw()+
  labs(y='Location')+
  scale_fill_discrete(guide='none')+
  theme(strip.text.x = element_blank(),
    axis.text.x.top = element_blank(),
    axis.ticks.x.top = element_blank(),
    axis.text.y = element_blank(),
    axis.title.y = element_blank(),
    axis.title.x = element_blank(),
    strip.background = element_blank(),
    plot.title = element_text(face = 'bold.italic',hjust=0.5,size=rel(0.8))
    )

L3=ggplot(dat[dat$genome=='P. palmiformis',])+
  geom_bar(aes(y=location,fill=location,x=count),stat='identity')+
  theme_bw()+
  labs(y='Location')+
  scale_fill_discrete(guide='none')+
  theme(strip.text.x = element_blank(),
    axis.text.x.top = element_blank(),
    axis.ticks.x.top = element_blank(),
    axis.text.y = element_blank(),
    axis.title.y = element_blank(),
    axis.title.x = element_blank(),
    strip.background = element_blank(),
    plot.title = element_text(face = 'bold.italic',hjust=0.5,size=rel(0.8))
    )

#### LTR size ####
S1=ggplot(datt[datt$genome=='P. echinospica',])+
  geom_histogram(aes(x=feature_size/1000),binwidth=1,color='black')+
  labs(x='Size\n(Kbp)')+
  scale_x_reverse(limits=c(30,0))+
  coord_flip()+
  theme_bw()+
  theme(strip.text = element_blank(),
      strip.background = element_blank(),
      axis.title.y = element_text(angle=0,vjust=0.5,hjust=0),
      axis.title.x = element_blank())

# S1

S2=ggplot(datt[datt$genome=='R. piscesae',])+
  geom_histogram(aes(x=feature_size/1000),binwidth=1,color='black')+
  scale_x_reverse(limits=c(30,0))+
  coord_flip()+
  theme_bw()+
  theme(strip.text = element_blank(),
      strip.background = element_blank(),
      axis.title.y = element_blank(),
      axis.title.x = element_blank())
# S2

S3=ggplot(datt[datt$genome=='P. palmiformis',])+
  geom_histogram(aes(x=feature_size/1000),binwidth=1,color='black')+
  scale_x_reverse(limits=c(30,0))+
  coord_flip()+
  theme_bw()+
  theme(strip.text = element_blank(),
      strip.background = element_blank(),
      axis.title.y = element_blank(),
      axis.title.x = element_blank())

# S3

#### meth freq ####

M1=ggplot(datt[datt$genome=='P. echinospica',])+
  geom_histogram(aes(x=meth_meanfreq*100),bins=20,color='black')+
  labs(x='Methylation level\n(%)')+
  coord_flip()+
  theme_bw()+
  theme(strip.text = element_blank(),
      strip.background = element_blank(),
      axis.title.y = element_text(angle=0,vjust=0.5,hjust=0),
      axis.title.x = element_blank())

# M1

M2=ggplot(datt[datt$genome=='R. piscesae',])+
  geom_histogram(aes(x=meth_meanfreq),bins=20,color='black')+
  coord_flip()+
  theme_bw()+
  # labs(y='LTR counts')+
  theme(strip.text = element_blank(),
      strip.background = element_blank(),
      axis.title.y = element_blank(),
      strip.text.x =element_blank())
# M2

M3=ggplot(datt[datt$genome=='P. palmiformis',])+
  geom_histogram(aes(x=meth_meanfreq),bins=20,color='black')+
  coord_flip()+
  theme_bw()+
  theme(strip.text = element_blank(),
      strip.background = element_blank(),
      axis.title.y = element_blank(),
      axis.title.x = element_blank())

# M3

#### CpGoe ####

C1=ggplot(datt[datt$genome=='P. echinospica',])+
  geom_histogram(aes(x=CpGoe),bins=50,color='black')+
  scale_x_continuous(limits=c(0,2))+
  labs(x='CpG obs/exp')+
  coord_flip()+
  theme_bw()+
  theme(strip.text = element_blank(),
      strip.background = element_blank(),
      axis.title.y = element_text(angle=0,vjust=0.5,hjust=0),
      axis.title.x = element_blank())

C2=ggplot(datt[datt$genome=='R. piscesae',])+
  geom_histogram(aes(x=CpGoe),bins=50,color='black')+
  scale_x_continuous(limits=c(0,2))+
  coord_flip()+
  labs(y='LTR-retroposons count',x='')+
  theme_bw()+
  theme(strip.text = element_blank(),
      strip.background = element_blank(),
      axis.title.y = element_blank())

C3=ggplot(datt[datt$genome=='P. palmiformis',])+
  geom_histogram(aes(x=CpGoe),bins=50,color='black')+
  scale_x_continuous(limits=c(0,2))+
  coord_flip()+
  theme_bw()+
  theme(strip.text = element_blank(),
      strip.background = element_blank(),
      axis.title.y = element_blank(),
      axis.title.x = element_blank())


#### mount ####

plot_description=egg::ggarrange(F1,F2,F3,L1,L2,L3,S1,S2,S3,T1,T2,T3,M1,M2,M3,C1,C2,C3, nrow=6, ncol=3,widths = c(1, 1, 1))


ggplot2::ggsave("../figures/LTR_distrib.png",plot_description,device = 'png',width = 2136 , height = 2136, units = 'px', dpi = 300)
ggplot2::ggsave("../figures/LTR_distrib.pdf",plot_description,device = 'pdf',width = 2136 , height = 2136, units = 'px')


```

### LTR vs genomic context
#### computing statistics
```{r compute fig LTRvscontext, eval=FALSE, include=FALSE}
library(ggpmisc)

datt=read.csv('../data/Transposable_elements/All.downstream.LTR.tsv',header=T,sep='\t')
datt %>% str

dat=datt %>% group_by(genome,LTR) %>% summarise(meth_meanfreq=mean(meth_meanfreq),flank4kb_meth_meanfreq=mean(flank4kb_meth_meanfreq),meth_binary=mean(meth_binary),CpGoe=mean(CpGoe),reps_CpGoe=mean(reps_CpGoe),meth_depth=mean(meth_depth),meth_cov=mean(meth_cov),flank4kb_meth_cov=mean(flank4kb_meth_cov),flank4kb_meth_depth=mean(flank4kb_meth_depth),location=unique(location)
                                                )


dat$meth_binary2=ifelse(dat$meth_meanfreq<0.1,0,
                         ifelse(dat$meth_binary==0,0.5,1))
dat$meth_binary2=factor(dat$meth_binary2,levels=c(1,0.5,0), labels=c('high','intermediate','low'))

dat$meth_binary=factor(dat$meth_binary,levels=rev(c(1,0)), labels=rev(c('high','low-intermediate')))

dat$genome=factor(dat$genome, levels=c('P. echinospica', 'R. piscesae', 'P. palmiformis'))

line_colors_bright=c('low-intermediate'='#023eff','high'='#e8000b')
line_colors_dark=c('low-intermediate'='#001c7f','high'='#8c0800')


plot1<-ggplot(dat,aes(y=flank4kb_meth_meanfreq*100,x=meth_meanfreq*100,color=factor(meth_binary)))+
  geom_point(size=1,alpha=0.5)+
  geom_abline(intercept=0,slope=1,linetype=2)+
  facet_grid(.~genome)+
  scale_color_manual(values=line_colors_bright)+
  # stat_smooth(method='lm')+
  stat_poly_eq(method = 'lm',label.x='right', label.y='bottom',vstep = 0.1) +
  coord_fixed()+
  theme_bw()+
  labs(x="LTR methylation level (%)",y="LTR flanking regions\nmethylation level (%)",color='LTR methylation')+
  guides(color = guide_legend(override.aes = list(size = 4)))+
  theme(strip.background = element_blank(),
        strip.text=element_text(face = 'bold.italic',size=10))

plot1
# ggplot2::ggsave("../figures/LTRvscontext.png",plot1,device = 'png', width=9.5, height=3.5, units='in', dpi = 300)
# ggplot2::ggsave("../figures/LTRvscontext.pdf",plot1,device = 'pdf', width=9.5, height=3.5, units='in')

########

e <- function(x) {
  r <- quantile(x, probs = c(0.025, 0.25, NA, 0.75, 0.975))
  names(r) <- c("ymin", "lower", "middle", "upper", "ymax")
  r
}

f <- function(x) {
  r <- quantile(x, probs = c(NA, 0.4, 0.5, 0.6, NA))
  names(r) <- c("ymin", "lower", "middle", "upper", "ymax")
  r
}


dat$rel_meth_meanfreq=(dat$meth_meanfreq-dat$flank4kb_meth_meanfreq)/dat$flank4kb_meth_meanfreq
dat$rel_meth_cov=(dat$meth_cov-dat$flank4kb_meth_cov)/dat$flank4kb_meth_cov  
dat$rel_meth_depth=(dat$meth_depth-dat$flank4kb_meth_depth)/dat$flank4kb_meth_depth  
 
dat2=melt(dat[,c('genome','LTR','location','CpGoe','reps_CpGoe','meth_binary','rel_meth_meanfreq','rel_meth_cov','rel_meth_depth')],id.vars=c('genome','LTR','CpGoe','reps_CpGoe','meth_binary','location'))

dat2$meth_binary=factor(dat$meth_binary,levels=rev(c('high','low-intermediate')))
dat2$variable=factor(dat2$variable,levels=rev(c("rel_meth_meanfreq", "rel_meth_cov", 
"rel_meth_depth")), labels=rev(c('level','coverage','depth')))


annotation=cbind(label=c('Flanking region > LTR-retrotransposon','LTR-retrotransposon > Flanking region'),y=c(-0.75,0.75),yend=c(-1.25,1.25),ylab=c(-1,1),genome=c('P. echinospica','P. echinospica'),variable=c('rel_meth_depth','rel_meth_depth')) %>% as.data.frame
annotation$variable=factor(annotation$variable,levels=rev(c("rel_meth_meanfreq", "rel_meth_cov", 
"rel_meth_depth")), labels=rev(c('level','coverage','depth')))

annotation %>% str
annotation$y=as.numeric(annotation$y)
annotation$yend=as.numeric(annotation$yend)
annotation$ylab=as.numeric(annotation$ylab)
annotation$variable=as.factor(annotation$variable)

alpha_scale=c(0.2,0.4)
names(alpha_scale)=c('50','20')

# ggplot(dat2)+
#   geom_boxplot(aes(x=paste(meth_binary,variable,sep='_'),y=value,fill=as.factor(meth_binary)))+
#   scale_y_continuous(limits=c(-1,1))+
#   facet_grid(.~genome)+
#   theme_bw()

dat2$genome %>% levels

plot2<-ggplot(dat2)+
  # geom_boxplot(aes(x=variable,y=value,fill=as.factor(meth_binary)))+
  stat_summary(aes(x=variable,y=value,fill=meth_binary,color=meth_binary,alpha='50'),fun.data = e, geom="boxplot",position = 'dodge')+
  stat_summary(aes(x=variable,y=value,fill=meth_binary,color=meth_binary,alpha='20'),fun.data = f, geom="boxplot",position = 'dodge')+
  geom_hline(yintercept = 0, linetype=2,color='grey30')+
  geom_segment(data=annotation,aes(x=as.numeric(variable)-0.5, xend=as.numeric(variable)-0.5, y=y,yend=yend),arrow = arrow(length = unit(0.08, "inch")),color='grey30')+
  geom_text(data=annotation,aes(x=as.numeric(variable)-0.5,y=ylab,label=label,hjust=0),color='grey30',nudge_x = 0.1,size=3)+
  scale_fill_manual(values=line_colors_bright,guide='none')+
  scale_color_manual(values=line_colors_bright,guide='none')+
  scale_alpha_manual(values=alpha_scale,name='Interpercentile range')+
  scale_y_continuous(limits=c(-1.25,1.25),labels= function(x) paste0(symnum(x, c(-Inf, 0, Inf), c("", "+")), x*100),name='Pair-wise difference between\n LTR and flanking regions (%)')+
  scale_x_discrete(name='Methylation parameter')+
  facet_grid(.~genome)+
  coord_fixed()+
  theme_bw()+
  guides(alpha = guide_legend(override.aes = list(fill = 'grey20', alpha=c(0.1,0.4))))+
   theme(strip.background = element_blank(),
        strip.text=element_blank(),
        axis.text.x = element_text(angle=45,hjust=1))



plot=egg::ggarrange(plot1,plot2, nrow=2, ncol=1,heights = c(1, 1))

plot

ggplot2::ggsave("../figures/LTRvscontext_v4.2.png",plot,device = 'png', width=9.5, height=6, units='in', dpi = 300)
ggplot2::ggsave("../figures/LTRvscontext_v4.2.pdf",plot,device = 'pdf', width=9.5, height=6, units='in')

# dat2$contig=sub("\\:.*", "", dat2$LTR)
# 
dat2[dat2$genome=='P. echinospica',] %>% group_by(genome,contig) %>% summarise(count=n()) %>% .[order(.$count,decreasing = T),]

dat2[dat2$genome=='P. echinospica' & dat2$location=='gene' & dat2$value > 0.5 & !(is.na(dat2$value)),]
# 


dat2[dat2$genome=='P. echinospica' & dat2$location=='.' & dat2$value > 0.5 & !(is.na(dat2$value)),]

dat %>% str

dat[dat$genome=='P. echinospica' & dat$flank4kb_meth_meanfreq>0.25 & dat$flank4kb_meth_meanfreq<0.5 & dat$meth_meanfreq < .05 ,] %>% as.data.frame

dat[dat$genome=='P. echinospica' & dat$flank4kb_meth_meanfreq>0.75 &  dat$meth_meanfreq > 0.75 & dat$flank4kb_meth_meanfreq/dat$meth_meanfreq > 0.95 & dat$flank4kb_meth_meanfreq/dat$meth_meanfreq < 1.05,] %>% as.data.frame


dat2[dat2$genome=='P. echinospica' & dat2$location=='gene' & dat2$value < -0.98 & !(is.na(dat2$value)),]
# 
dat2[ dat2$meth_binary=='high' & dat2$value < -0.95 & !(is.na(dat2$value)),]
```


#### inspecting specific cases of LTR vs context

```{R}
meta=read.csv('../data/Transposable_elements/All.downstream.LTR.tsv',header=T,sep='\t')
colnames(meta) %>% dput
cols_to_rename=c("insertion_time_my", "CpGs_count", 
"CpGoe", "location", "meth_meancallcov", "meth_meanfreq", "callCpGs_count", 
"meth_cov", "meth_depth", "meth_binary", "recent_insertion", 
"feature_size","insertion_time_qbins_my", "flank4kb_meth_meanfreq", 
"flank4kb_meth_cov", "flank4kb_meth_depth", "flank4kb_callCpGs_count")
colnames(meta)[colnames(meta) %in% cols_to_rename]=paste0('LTR_',colnames(meta)[colnames(meta) %in% cols_to_rename])
meta$LTR_insertion_time_qbins_my2 = sapply( meta$LTR_insertion_time_qbins_my %>% strsplit(.,'[(*),]'), `[`, 2) %>% as.numeric(.)
meta$LTR_insertion_time_qbins_my3 = sapply( meta$LTR_insertion_time_qbins_my %>% strsplit(.,'\\(|\\]|,'), `[`, 3) %>% as.numeric(.)
meta$LTR_insertion_time_qbins_my4 = ((meta$LTR_insertion_time_qbins_my3-meta$LTR_insertion_time_qbins_my2)/2)+meta$LTR_insertion_time_qbins_my2
meta=merge(meta,meta[order(meta$genome,meta$LTR_insertion_time_qbins_my2,meta$LTR),] %>% group_by(genome) %>% summarise(LTR_insertion_time_qbins_my=unique(LTR_insertion_time_qbins_my),LTR_insertion_time_qbins_rel=seq_along(order(unique(LTR_insertion_time_qbins_my2)))) %>% as.data.frame,by=c('genome','LTR_insertion_time_qbins_my'),all.x=T)

meta %>% str

species=c('R. piscesae','P. palmiformis','P. echinospica',"R07B-5", "P08H-3", "Pec")
names(species)=c("R07B-5", "P08H-3", "Pec",'R. piscesae','P. palmiformis','P. echinospica')

df_all=data.frame()
for (ind in c('Pec','R07B-5','P08H-3')){

genome=species[ind]; genome
cov=ifelse(ind=='P08H-3','cov1','cov10')
file=paste0('../data/Transposable_elements/',ind,'.downstream.',cov,'.LTR_TSS.2.tsv')

df=read.csv(file,header=T,sep='\t')
df$LTR_boundary=ifelse(df$LTR_boundary=='start','5prime','3prime')
colnames(df)[colnames(df)=='mRNA']='LTR'
df$genome=genome

df[df$TSS_idx>=0,'TSS_idx']=df[df$TSS_idx>=0,'TSS_idx']-1 # shift LTR_idx downstream counts to start at 0 instead of 1

LTRbins=10 # number of visible bins for the LTR body
TSS_idx_max_dic=df %>% group_by(LTR) %>% summarise(TSS_idx_max=max(TSS_idx)/LTRbins) %>% deframe
# get max value for LTR body
df$TSS_idx_rel=df$TSS_idx
df[df$TSS_idx>=0,'TSS_idx_rel']=2*floor(df[df$TSS_idx>=0,'TSS_idx']/TSS_idx_max_dic[df[df$TSS_idx>=0,'LTR']]) # normalise LTR_body TSS_idx to LTRbins range
df[(df$TSS_idx>=0) & (df$LTR_boundary=='3prime'),'TSS_idx_rel']=2*LTRbins - df[(df$TSS_idx>=0) & (df$LTR_boundary=='3prime'),'TSS_idx_rel'] # set origin at 3prime LTR boundary and reverse order of TSS_idx negtive strand walk
df[(df$TSS_idx<0) & (df$LTR_boundary=='3prime'),'TSS_idx_rel']=2*LTRbins - df[(df$TSS_idx<0) & (df$LTR_boundary=='3prime'),'TSS_idx_rel'] # set origin at 3prime LTR boundary reverse order of TSS_idx negtive strand walk
meth_freq_rel=df %>% group_by(LTR,TSS_idx_rel) %>% summarise(meth_freq_rel=mean(meth_freq)) # windows over the LTR body may be merged following the LTR body size normalisation and thus their values of meth_freq should be averaged
df=merge(df,meth_freq_rel,by=c('LTR','TSS_idx_rel'),all.x=T) %>% unique

df_all=rbind(df_all,df)
}

df_all=df_all[between(df_all$TSS_idx_rel,-20,2*LTRbins+20), ]
df_all %>% str

df_all %>% summary

flank5p=df_all[df_all$TSS_idx_rel<0,] %>% group_by(genome,LTR) %>% summarise(flank5p_meth_cov=mean(meth_cov),flank5p_meth_depth=mean(meth_depth), flank5p_meth_meancallcov=mean(meth_meancallcov)) %>% as.data.frame

flank3p=df_all[df_all$TSS_idx_rel>20,] %>% group_by(genome,LTR) %>% summarise(flank3p_meth_cov=mean(meth_cov),flank3p_meth_depth=mean(meth_depth), flank3p_meth_meancallcov=mean(meth_meancallcov)) %>% as.data.frame

flank5p %>% str

flanks=merge(flank5p,flank3p, by=c('genome','LTR'))

dat=merge(meta,flanks,by=c('genome','LTR'), all.x=T)

dat %>% str

#############
dat$flank5p_meth_binary=ifelse(dat$flank5p_meth_cov > 0.75 & dat$flank5p_meth_depth > 0.5,1,0) 
dat$flank3p_meth_binary=ifelse(dat$flank3p_meth_cov > 0.75 & dat$flank3p_meth_depth > 0.5,1,0) 
dat$flanks_binary=paste0(dat$flank5p_meth_binary,dat$flank3p_meth_binary)

dat %>% group_by(genome,LTR_meth_binary,flanks_binary) %>% summarise(counts=n())
#########

#######
dat %>% str 

dat[dat$LTR_meth_meanfreq>0.9 & dat$LTR_flank4kb_meth_meanfreq>0.75 & dat$LTR_flank4kb_meth_meanfreq<0.8,c('LTR','LTR_location')]

dat[dat$genome=='P. echinospica' & dat$LTR_meth_binary==0 & dat$flanks_binary=='11',] %>% as.data.frame

dat[dat$genome=='P. echinospica' & dat$LTR_meth_binary==0 & dat$flanks_binary=='00',] %>% as.data.frame

dat[dat$genome=='P. echinospica' & dat$LTR_meth_binary==1 & dat$flanks_binary=='11',] %>% as.data.frame %>% .[,c('LTR','LTR_location')]

#########
list_of_examples=c('PE_Scaf7794:144452..148014','PE_Scaf10002:98572..103585','PE_Scaf9477:200430..205202','PE_Scaf10317:34581..37742')

dat[dat$LTR %in% list_of_examples,c('LTR','LTR_meth_meanfreq','flanks_binary','meth_binary','LTR_flank4kb_meth_meanfreq')]

dat$meth_binary=factor(dat$LTR_meth_binary,levels=rev(c(1,0)), labels=rev(c('high','low-intermediate')))

dat[dat$genome=='P. echinospica' & dat$LTR %in% list_of_examples,'LTR']
dat[dat$genome=='P. echinospica' & dat$LTR %in% list_of_examples,'label']<-c('D','A','B','C')

ggplot(dat[dat$genome=='P. echinospica',],aes(y=LTR_flank4kb_meth_meanfreq*100,x=LTR_meth_meanfreq*100))+
  # geom_point(shape=21,size=1,alpha=0.5,fill='white')+
  geom_point(size=1,alpha=0.2)+
  geom_point(data=dat[dat$genome=='P. echinospica' & dat$LTR %in% list_of_examples,],aes(y=LTR_flank4kb_meth_meanfreq*100,x=LTR_meth_meanfreq*100),shape=21,color='#f59eff',fill='#8b3395',size=2,alpha=1)+
  geom_text(data=dat[dat$genome=='P. echinospica' & dat$LTR %in% list_of_examples,],aes(y=LTR_flank4kb_meth_meanfreq*100,x=LTR_meth_meanfreq*100,label=label),color='#8b3395',hjust=0,nudge_x = 5,size=4,fontface = "bold")+
  geom_abline(intercept=0,slope=1,linetype=2)+
  scale_y_continuous(limits=c(0,105))+
  scale_x_continuous(limits=c(0,105))+
  coord_fixed()+
  theme_bw()+
  labs(x="LTR methylation level (%)",y="LTR flanking regions\nmethylation level (%)")+
  guides(color = guide_legend(override.aes = list(size = 4)))+
  theme(strip.background = element_blank(),
        strip.text=element_text(face = 'bold.italic',size=10),
        )

```

### LTR vs insertion time 

```{r LTR-insertiontime-pcor-quant, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}
library(ppcor)

datt=read.csv('../data/Transposable_elements/All.downstream.LTR.tsv',header=T,sep='\t')

datt %>% str
datt %>% group_by(genome) %>% summarise(count=n())

datt$insertion_time_qbins_my2 = sapply( datt$insertion_time_qbins_my %>% strsplit(.,'[(*),]'), `[`, 2) %>% as.numeric(.)
datt$insertion_time_qbins_my3 = sapply( datt$insertion_time_qbins_my %>% strsplit(.,'\\(|\\]|,'), `[`, 3) %>% as.numeric(.)
datt$insertion_time_qbins_my4 = ((datt$insertion_time_qbins_my3-datt$insertion_time_qbins_my2)/2)+datt$insertion_time_qbins_my2
datt$location=as.factor(datt$location) %>% as.numeric()
datt$CpGs_density=datt$CpGs_count/datt$feature_size

datt=datt[complete.cases(datt[,c('flank4kb_meth_meanfreq','insertion_time_my','meth_meanfreq'),]), ]
datt %>% group_by(genome) %>% summarise(count=n())
datt %>% head


# plot_ly(data=datt,x=~flank4kb_meth_meanfreq,y=~meth_binary,z = ~insertion_time_qbins_my2,color=~insertion_time_qbins_my,type='scatter3d',size=1)

##########
# variables=list(c(),c("flank4kb_meth_meanfreq"),c('reps_CpGoe' ))
variables=list(c(),c("flank4kb_meth_meanfreq"))

cols=c("estimate", "p.value", "statistic", "n", "gp", "Method",'ctrl','genome','method')

pcor_df=setNames(data.frame(matrix(ncol = length(cols), nrow = 0)), cols)

for (variable in c('insertion_time_my')){
for (method in c('spearman','pearson')){
for (genome in c(unique(datt$genome))){
for (ctrl in variables){
high=pcor.test(datt[datt$genome==genome,'meth_meanfreq'],
               datt[datt$genome==genome,variable],
               datt[datt$genome==genome,ctrl],
               method=method) 
?pcor.test  

high$ctrl=paste(ctrl,collapse = " + ") # %>% print
high$genome=genome

pval_thresh=0.05
high$method=method
high$variable=variable
high$signif=as.numeric(high$p.value<pval_thresh)

pcor_df=rbind(pcor_df,high)

}}}
 }


pcor_df$genome=factor(pcor_df$genome, levels=c("P. echinospica", "R. piscesae", "P. palmiformis"))
                                         
pcor_df=pcor_df[with(pcor_df, order(genome, ctrl)),c('genome','n','ctrl','estimate','p.value','method','variable')]
pcor_df$p.value_str=sprintf("%f",pcor_df$p.value)

pval_thresh=0.05
pcor_df$signif=as.numeric(pcor_df$p.value<pval_thresh)
pcor_df

write.table(pcor_df,'../data/Transposable_elements/All.downstream.LTR.pcor',row.names = F,col.names = T,quote=F,sep='\t')

```
```{r prep figure, eval=FALSE, include=FALSE}

# ++++++++++++++++++++++++++++
# flattenCorrMatrix
# ++++++++++++++++++++++++++++
# cormat : matrix of the correlation coefficients
# pmat : matrix of the correlation p-values
flattenCorrMatrix <- function(cormat, pmat) {
  ut <- upper.tri(cormat)
  data.frame(
    row = rownames(cormat)[row(cormat)[ut]],
    column = rownames(cormat)[col(cormat)[ut]],
    cor  =(cormat)[ut],
    p = pmat[ut]
    )
}


datt=read.csv('../data/Transposable_elements/All.downstream.LTR.tsv',header=T,sep='\t')

datt$insertion_time_qbins_my2 = sapply( datt$insertion_time_qbins_my %>% strsplit(.,'[(*),]'), `[`, 2) %>% as.numeric(.)
datt$location=as.factor(datt$location) %>% as.numeric()
datt$CpGs_density=datt$callCpGs_count/datt$feature_size
datt=datt[complete.cases(datt), ]
datt %>% colnames

library(Hmisc)

plot_pairwide_corr=function(datt){
cols=c("row", "column", "cor", "p", "genome")
rcor_df=setNames(data.frame(matrix(ncol = length(cols), nrow = 0)), cols)
for (genome in c(unique(datt$genome))){

tmp=datt[datt$genome==genome, c('meth_meanfreq', 'flank4kb_meth_meanfreq', 'location', 'insertion_time_my','feature_size','reps_CpGoe')]

res2 <- rcorr(as.matrix(tmp),type='pearson')

res=flattenCorrMatrix(res2$r, res2$P)
res$genome=genome
rcor_df=rbind(rcor_df,res)

}

lvls_row=c( "meth_meanfreq","flank4kb_meth_meanfreq","location","insertion_time_my",
 "feature_size",'reps_CpGoe')
lvls_col=c('reps_CpGoe',"feature_size", "meth_meanfreq","insertion_time_my","location","flank4kb_meth_meanfreq")


rcor_df$row=factor(rcor_df$row,levels=lvls_row, labels=c( "LTR-retrotransposon\nmethlation level","Genomic context\nmethylation level","LTR-retrotransposon\nlocation","LTR-retrotransposon\ninsertion time",
 "LTR-retrotransposon\nsize",'Mutational bias on\nterminal repeats (CpG o/e)\n'))
rcor_df$col=factor(rcor_df$col,levels=lvls_col,labels=c('Mutational bias on\nterminal repeats (CpG o/e)\n',"LTR-retrotransposon\nsize", "LTR-retrotransposon\nmethlation level", "LTR-retrotransposon\ninsertion time","LTR-retrotransposon\nlocation","Genomic context\nmethylation level"))
rcor_df$alpha=-log10(rcor_df$p+0.0000000001)
rcor_df$genome=factor(rcor_df$genome,levels=c("P. echinospica", "R. piscesae", "P. palmiformis"))

rcor_df$p2=ifelse(rcor_df$p<0.001,'< 0.001',
                        ifelse(rcor_df$p<0.01,'< 0.01',
                               ifelse(rcor_df$p<0.05,'< 0.05','> 0.05')))

alpha_values=c('> 0.05'=0.01,'< 0.05'=0.5,'< 0.01'=0.75,'< 0.001'=1)

plot<-ggplot(rcor_df)+
  geom_bin2d(aes(y=col,x=row,fill=cor,alpha=p2),stat='identity',color='grey')+
  geom_text(aes(y=col,x=row,label=round(cor*100)/100),alpha=0.75,size=rel(2))+
  # scale_alpha_continuous(range = c(1,0))+
  scale_fill_viridis(option = "plasma",limits=c(-0.8,1.2))+
  scale_alpha_manual(values=alpha_values)+
  labs(x='',y='',fill='Pearson correlation (r)',alpha='p.value')+
  # scale_fill_distiller(palette = "Spectral")+
  # scale_fill_gradient2(low = "red", high = "blue", mid = 'purple',midpoint=0.1)+
  facet_wrap(~genome)+
  theme_bw()+
  theme(axis.text.x=element_text(hjust=1,vjust=1,lineheight = 0.8,angle=45),
        axis.text.y=element_text(vjust=0.5,lineheight = 0.8),
        panel.grid = element_blank(),
        strip.background = element_blank(),
        strip.text = element_text(face = 'bold.italic'),
        axis.title = element_blank())+
  coord_fixed()

plot

return(list('plot'=plot,'rcor_df'=rcor_df))
}

########
Pall=plot_pairwide_corr(datt)$plot
Plow=plot_pairwide_corr(datt[datt$meth_binary==0,])$plot
Phigh=plot_pairwide_corr(datt[datt$meth_binary==1,])$plot

P1=Pall + labs(title='All LTR-retrotransposons')+
                           theme(legend.position="none",
                                 axis.text.x=element_blank(),
                                 title = element_text(angle=0))
P2=Plow + labs(title='LTR-retrotransposons with low-intermediate methylation')+
                           theme(legend.position="right",
                                 title = element_text(angle=0),
                                 axis.text.x=element_blank())
P3=Phigh + labs(title='LTR-retrotransposons with high methylation')+
                           theme(legend.position="none",
                                 title = element_text(angle=0),
                                 axis.text.x=element_text(lineheight = 0.6,size =rel(0.95)))


plot=egg::ggarrange(P1,P2,P3,nrow=3,ncol=1,heights=c(1,1,1))

ggsave("../figures/fig-corrmatrix_CpGoe.png",device = 'png',plot,width = 11.08 , height=8.72, units = 'in',dpi=300)
ggsave("../figures/fig-corrmatrix_CpGoe.pdf",plot,width = 11.08 , height=8.72, units = 'in')

```


## Gene body methylation
### Introns-exons

```{r prep for fig intronexon, eval=FALSE, include=FALSE}

df=read.csv('../data/Gene-body_methylation/ALL.downstream.introns_vs_exons_long.tsv',header=T,sep='\t',row.names = 1)

df$meth_meanfreq=df$meth_cov*df$meth_depth

df %>% summary

textsize=ggplot2::.pt

conditions=df$type %in% c('intron','exon') & df$callCpGs_count>=10
d=df[conditions,] %>% dcast(.,ind+Parent+feature_posi~type, value.var = 'meth_cov')
colnames(d)[4]='exon_cov'
colnames(d)[5]='intron_cov'

dd=df[conditions,] %>% dcast(.,ind+Parent+feature_posi~type, value.var = 'meth_depth')
colnames(dd)[4]='exon_depth'
colnames(dd)[5]='intron_depth'

dat_wide=merge(d,dd,by=c('ind','Parent','feature_posi'),all=T)

dat_wide=dat_wide[complete.cases(dat_wide),]

### exclude first 3 exons
#dat_wide=dat_wide[dat_wide$feature_posi>3,]
## only first 3 exons
#dat_wide=dat_wide[dat_wide$feature_posi<=3,]

# dat_wide %>% group_by(ind) %>% summarise(count=n())
dat_wide_sample=rbind(sample_n(dat_wide[dat_wide$ind=='Pec',],10000),
                 sample_n(dat_wide[dat_wide$ind=='R07B-5',],10000),
                 sample_n(dat_wide[dat_wide$ind=='P08H-3',],10000)) %>% .[,c('ind','Parent','feature_posi')]
dat_wide_sample$ind=factor(dat_wide_sample$ind,levels=c('Pec','R07B-5','P08H-3'), labels=c('P. echinospica','R. piscesae','P. palmiformis'))



d1=reshape2::melt(dat_wide[,c('ind','Parent','feature_posi','exon_depth','intron_depth')],id.vars=c('ind','Parent','feature_posi'))

colnames(d1)[4]='type'
colnames(d1)[5]='meth_depth'
d1$type=sapply(strsplit(as.character(d1$type),"_"), `[`, 1)

d2=reshape2::melt(dat_wide[,c('ind','Parent','feature_posi','exon_cov','intron_cov')],id.vars=c('ind','Parent','feature_posi'))
colnames(d2)[4]='type'
colnames(d2)[5]='meth_cov'
d2$type=sapply(strsplit(as.character(d2$type),"_"), `[`, 1)

dat=cbind(d1,d2$meth_cov)
colnames(dat)[6]='meth_cov'

dat$ind=factor(dat$ind,levels=c('Pec','R07B-5','P08H-3'), labels=c('P. echinospica','R. piscesae','P. palmiformis'))
dat$type=factor(dat$type, levels=c("intron","exon"))

library(ggridges)

p1_ridge<-ggplot(dat)+
  # geom_density_ridges(aes(y=type,x=meth_cov,fill=type),stat='binline',alpha=0.6,scale = 3, rel_min_height = 0.01) +
  geom_density_ridges(aes(y=type,x=meth_cov,fill=type,color=type),binwidth=0.02,alpha=0.6,stat='binline') +
guides(colour = guide_legend(override.aes = list(alpha = 1,size=5)))+
  facet_wrap(~ind)+
  scale_fill_manual(values=c('exon'='#0025ff','intron'='#ff8300'))+
  scale_color_manual(values=c('exon'='#0025ff','intron'='#ff8300'))+
  labs(fill='Feature type',y='density')+
  theme_bw()+
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        legend.position = 'none',
        strip.background = element_blank(),
        strip.text.x= element_text(face=c('bold.italic'),angle=0),
        aspect.ratio=1/3)

p3_ridge<-ggplot(dat)+
  # geom_density_ridges(aes(y=type,x=meth_depth,fill=type),alpha=0.6,scale = 3, rel_min_height = 0.01) +
  geom_density_ridges(aes(y=type,x=meth_depth,fill=type,color=type),binwidth=0.02,alpha=0.6,stat='binline') +
  coord_flip()+
  theme_bw()+
  labs(x='Feature count',y='')+
  scale_fill_manual(values=c('exon'='#0025ff','intron'='#ff8300'),guide='none')+
  scale_color_manual(values=c('exon'='#0025ff','intron'='#ff8300'),guide='none')+
  theme(axis.text.y=element_blank(),
        axis.title.y=element_blank(),
        axis.text.x=element_text(angle=60,hjust=1,vjust=1),
        strip.background = element_blank(),
        strip.text.y=element_text(face='bold'),
        aspect.ratio=3/1)

###############

p1<-ggplot(dat)+
  stat_bin(aes(color=type,x=meth_cov,y=..density..),drop=T,geom='step',position='identity',bins=50)+
  scale_y_log10(labels= scales::label_number_si(),
                minor_breaks= c(50,500,5000),
                position='left')+
  guides(colour = guide_legend(override.aes = list(alpha = 1,size=5)))+
  facet_wrap(~ind)+
  scale_color_manual(values=c('exon'='#0025ff','intron'='#ff8300'))+
  labs(color='Feature type')+
  theme_bw()+
  theme(axis.text.x = element_blank(),
        axis.title.x = element_blank(),
        legend.position = 'none',
        strip.background = element_blank(),
        strip.text.x= element_text(face=c('bold.italic'),angle=0),
        aspect.ratio=1/3)

p2<-ggplot(merge(dat,dat_wide_sample,by=c('ind','Parent','feature_posi'),all.y=T))+
  geom_point(aes(color=type,y=meth_depth,x=meth_cov),alpha=0.05,size=1)+
  scale_color_manual(values=c('exon'='#0025ff','intron'='#ff8300'),guide='none')+
  scale_x_continuous(labels=scales::percent_format(accuracy=1))+
  scale_y_continuous(labels=scales::percent_format(accuracy=1))+
  facet_wrap(~ind)+
  coord_fixed()+
  theme_bw()+
  labs(x='Methylation coverage', y='Methylation depth')+
  theme(strip.background = element_blank(),
        strip.text = element_blank(),
        legend.position = 'none',
        aspect.ratio=3/3)

p3<-ggplot(dat)+
  stat_bin(aes(color=type,x=meth_depth,y=..count..),geom='step',position='identity',bins=50,pad=F)+
  scale_y_log10(breaks=c(10,100,1000,10000),
                limits=c(10,20000),
                labels= scales::label_number_si(),
                minor_breaks= c(50,500,5000),
                position='left')+
  coord_flip()+
  theme_bw()+
  labs(x='Feature count')+
  # facet_grid('All species'~.)+
  scale_color_manual(values=c('exon'='#0025ff','intron'='#ff8300'),guide='none')+
  theme(axis.text.y=element_blank(),
        axis.title.y=element_blank(),
        axis.text.x=element_text(angle=60,hjust=1,vjust=1),
        strip.background = element_blank(),
        strip.text.y=element_text(face='bold'),
        aspect.ratio=3/1)

n=0.25
j=4
empty <- ggplot()+
  geom_rect(aes(xmin = 0,ymin = n+0,xmax=0.5/j,ymax=n+0.5/j,fill='intron'))+
  geom_rect(aes(xmin = 0,ymin = n+0.75/j,xmax=0.5/j,ymax=n+1.25/j,fill='exon'))+
  geom_text(aes(x=0.7/j,y = n+0.25/j, label='intron'),hjust = 0)+
  geom_text(aes(x=0.7/j,y = n+1/j, label='exon'),hjust = 0)+
  scale_fill_manual(values=c('exon'='#0025ff','intron'='#ff8300'),guide='none')+
  scale_x_continuous(limits = c(0,1))+
  scale_y_continuous(limits = c(0,1))+
  coord_fixed()+
  theme(axis.ticks=element_blank(), 
        panel.background=element_blank(), 
        axis.text.x=element_blank(), axis.text.y=element_blank(), 
        axis.title.x=element_blank(), axis.title.y=element_blank(),
        legend.position='none')

# egg::ggarrange(p1,empty,p2,p3, nrow = 2, ncol=2,heights = c(1, 5),widths=c(10,1))

plot=egg::ggarrange(p1_ridge,empty,p2,p3_ridge, nrow = 2, ncol=2,heights = c(1, 5),widths=c(10,1))
plot

```

```{r export figure}
ggsave('../figures/intronexon_cov_depth_distrib.png',width=6.5,height=3.5, units = 'in',plot, device = 'png',dpi=300)
ggsave('../figures/intronexon_cov_depth_distrib.pdf',width=6.5,height=3.5, units = 'in',plot, device = 'pdf')

```

```{r permutation test variance, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

### these permutation test shows DNA methylation coverage and depth have greater variance in exons than in introns

detach(package:'Hmisc')
library(tidyr)

#### methylation coverage ####

dat=df[df$type %in% c('intron','exon'),]
dat$ind=factor(dat$ind,levels=c('Pec','R07B-5','P08H-3'), labels=c('P. echinospica','R. piscesae','P. palmiformis'))
dat=dat[!(is.na(dat$meth_cov)),]

dat %>% summary


iter<-data.frame(ind=as.character(),var_exon=as.numeric(numeric()),var_intron=as.numeric(numeric()),iter=as.numeric(numeric()))

for (n in c(1:500)){

itern=rbind(sample_n(dat[dat$ind=='P. echinospica',],100),sample_n(dat[dat$ind=='R. piscesae',],100),sample_n(dat[dat$ind=='P. palmiformis',],100))
itern=itern %>% group_by(ind,type) %>% summarise(var=var(meth_cov)) %>% as.data.frame %>% pivot_wider(., names_from = type, values_from = c(var))
itern$iter=n
iter=rbind(iter,itern)
}

iter %>% group_by(ind) %>% summarise(prop=sum(exon>intron)/n())

#  ind             prop
#   <fct>          <dbl>
# 1 P. echinospica 0.944
# 2 R. piscesae    0.972
# 3 P. palmiformis 0.908

#### methylation depth ####

dat=df[df$type %in% c('intron','exon'),]
dat$ind=factor(dat$ind,levels=c('Pec','R07B-5','P08H-3'), labels=c('P. echinospica','R. piscesae','P. palmiformis'))
dat=dat[!(is.na(dat$meth_depth)),]


iter<-data.frame(ind=as.character(),var_exon=as.numeric(numeric()),var_intron=as.numeric(numeric()),iter=as.numeric(numeric()))

for (n in c(1:500)){
itern=rbind(sample_n(dat[dat$ind=='P. echinospica',],100),sample_n(dat[dat$ind=='R. piscesae',],100),sample_n(dat[dat$ind=='P. palmiformis',],100))
itern=itern %>% group_by(ind,type) %>% summarise(var=var(meth_depth)) %>% as.data.frame %>% pivot_wider(., names_from = type, values_from = c(var))
itern$iter=n
iter=rbind(iter,itern)
}

iter %>% group_by(ind) %>% summarise(prop=sum(exon>intron)/n())

# # A tibble: 3 × 2
#   ind             prop
#   <fct>          <dbl>
# 1 P. echinospica 0.949
# 2 R. piscesae    0.979
# 3 P. palmiformis 0.929

```

```{r get data for fig-intronexon-biplot, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}

df=read.csv('../data/Gene-body_methylation/ALL.downstream.introns_vs_exons_long.tsv',header=T,sep='\t',row.names = 1)
df$meth_meanfreq=df$meth_cov*df$meth_depth
df$ind=factor(df$ind,levels=c('Pec','R07B-5','P08H-3'), labels=c('P. echinospica','R. piscesae','P. palmiformis'))

feat='exon'
high_cov=df[df$type==feat & df$feature_posi==1 & df$meth_cov>0.75 ,c('ind','Parent')] %>% unique
high_cov$group='high_cov'
low_cov=df[df$type==feat & df$feature_posi==1 & df$meth_cov<0.25 ,c('ind','Parent')] %>% unique
low_cov$group='low_cov'

df=merge(df,rbind(high_cov,low_cov), by=c('ind','Parent'), all.x=T)
df[is.na(df$group),'group']<-'mid_cov'


conditions=(df$type %in% c('intron','exon') )

dat1=df[conditions ,] %>% dcast(.,Parent+feature_posi+ind+group~type,value.var='meth_depth')
dat1$var='meth_depth'

dat2=df[conditions ,] %>% dcast(.,Parent+feature_posi+ind+group~type,value.var='meth_cov')
dat2$var='meth_cov'

dat3=rbind(dat1,dat2) %>% .[complete.cases(.),]

dat=dat3  %>% melt(id.vars=c('Parent','feature_posi','ind','var','group'), variable.name = 'type')

dat$ind=factor(dat$ind,levels=c('P. echinospica','R. piscesae','P. palmiformis'))
dat$var=factor(dat$var, levels= c("meth_depth", "meth_cov"))

color_cov='#006400'
color_depth='#8b0000'

color_count='grey70'

?group_by
max_c=dat %>% group_by(feature_posi,ind,var) %>% summarise(count=n())

max_count=max(max_c$count)

p1<-ggplot(subset(dat,var=='meth_depth'))+
  geom_step(aes(x=feature_posi,y=..count../max_count),stat='count',color=color_count)+
  stat_summary(aes(x=feature_posi,y=value,color=type),geom = "line", fun = mean, linetype = "solid")+
  stat_summary(aes(x=feature_posi,y=value,color=type),geom = "point", fun = mean, linetype = "solid")+
  stat_summary(aes(x=feature_posi,y=value,fill=type),fun.data=mean_cl_normal, geom="ribbon",alpha=0.3)+
  scale_x_continuous(limits=c(1,20), breaks=c(1,5,10,15,20))+
  scale_y_continuous(sec.axis = sec_axis(~.*max_count/2, name = "Exon-Intron pair counts",labels= scales::label_number_si()), name='Mean\n methylation depth',labels=scales::percent_format(accuracy=1))+
  scale_color_manual(values=c('exon'='#0025ff','intron'='#ff8300'), name='Feature type')+
  scale_fill_manual(values=c('exon'='#0025ff','intron'='#ff8300'), name='Feature type')+
  facet_wrap(~ind)+
  theme_bw()+
  theme(strip.text.x = element_text(face='italic'),
        axis.line.y.right = element_line(color=color_count),
        axis.text.y.right = element_text(color=color_count),
        axis.ticks.y.right = element_line(color=color_count),
        axis.title.y.right = element_blank(),
        axis.line.y.left = element_line(color=color_depth),
        axis.text.y.left = element_text(color=color_depth),
        axis.ticks.y.left = element_line(color=color_depth),
        axis.title.y.left = element_text(color=color_depth),
        axis.text.x = element_blank(),
        axis.title.x=element_blank())

p2<-ggplot(subset(dat,var=='meth_cov'))+
  geom_step(aes(x=feature_posi,y=..count../max_count),stat='count',color=color_count)+
  stat_summary(aes(x=feature_posi,y=value,color=type),geom = "line", fun = mean, linetype = "solid")+
  stat_summary(aes(x=feature_posi,y=value,color=type),geom = "point", fun = mean, linetype = "solid")+
  stat_summary(aes(x=feature_posi,y=value,fill=type),fun.data=mean_cl_normal, geom="ribbon",alpha=0.3)+
  scale_x_continuous(limits=c(1,20), breaks=c(1,5,10,15,20),name='Position in gene',labels = label_ordinal())+
  scale_y_continuous(sec.axis = sec_axis(~.*max_count/2, name = "Exon-Intron pair counts",labels= scales::label_number_si()), name='Mean\n methylation coverage',labels=scales::percent_format(accuracy=1))+
  scale_color_manual(values=c('exon'='#0025ff','intron'='#ff8300'), name='Feature type',guide='none')+
  scale_fill_manual(values=c('exon'='#0025ff','intron'='#ff8300'), name='Feature type',guide='none')+
  facet_wrap(~ind)+
  theme_bw()+
  theme(strip.text.x = element_blank(),
        axis.line.y.right = element_line(color=color_count),
        axis.text.y.right = element_text(color=color_count),
        axis.ticks.y.right = element_line(color=color_count),
        axis.title.y.right = element_text(color=color_count),
        axis.line.y.left = element_line(color=color_cov),
        axis.text.y.left = element_text(color=color_cov),
        axis.ticks.y.left = element_line(color=color_cov),
        axis.title.y.left = element_text(color=color_cov),
        axis.text.x = element_blank(),
        axis.title.x=element_blank())

dat3=dcast(dat,Parent+feature_posi+ind+var+group~type, value.var='value')
dat3$deltaei=dat3$exon-dat3$intron

dat4=dcast(dat3,Parent+feature_posi+ind~var,value.var = 'deltaei') %>%  melt(id.vars=c('Parent','ind','feature_posi'))
dat4$variable=factor(dat4$variable,levels=c('meth_depth','meth_cov'), labels=c('methylation depth','methylation coverage'))

max_c2=dat4 %>% group_by(feature_posi,ind,variable) %>% summarise(count=n())
max_count2=max(max_c2$count)
min_c=dat4[dat4$feature_posi<=20,] %>% group_by(feature_posi,ind,variable) %>% summarise(min_diff=mean(value))
min_count=min(min_c$min_diff)
min_c_5th=subset(min_c,feature_posi==5 & variable=='methylation coverage' & ind=='P. echinospica')$min_diff

ann_text <- data.frame(feature_posi = c(5,5),y = c(0.05,(min_c_5th+min_count)/2),lab = c('Exon > Intron','Intron > Exon'),ind = factor(c('P. echinospica','P. echinospica'),levels = c('P. echinospica','R. piscesae','P. palmiformis')))

p3<-ggplot(dat4)+
  geom_step(aes(x=feature_posi,y=..count../(max_count2*8)+(min_count)),stat='count',color=color_count)+
  stat_summary(aes(x=feature_posi,y=value,fill=variable),fun.data=mean_se,geom = "bar", position='dodge')+
  scale_x_continuous(limits=c(1,20), breaks=c(1,5,10,15,20),name='Position in gene',labels = label_ordinal())+
  scale_y_continuous(sec.axis = sec_axis(~(.-min_count)*max_count2*4, name ="Exon-Intron pair counts",labels=scales::label_number_si()), name='Mean\n methylation difference\n in exon-intron pair', labels=scales::percent_format(accuracy=1))+
  # annotate('text',x=5,y=0.05,label='Exon > Intron',hjust=0)+
  geom_text(data=ann_text,aes(x=feature_posi,y=y,label=lab),hjust=0)+
  scale_fill_manual(values=c('methylation depth'=color_depth,'methylation coverage'=color_cov), name='Methylation estimate')+
  facet_wrap(~ind)+
  theme_bw()+
  theme(strip.text.x = element_blank(),
        axis.line.y.right = element_line(color=color_count),
        axis.text.y.right = element_text(color=color_count),
        axis.ticks.y.right = element_line(color=color_count),
        axis.title.y.right = element_blank())

plot=egg::ggarrange(p1,p2,p3, nrow = 3, ncol=1,heights = c(1, 1, 1))

plot
ggsave('../figures/Intonexon_depth_cov_vs_geneposition.png',plot, width=10.5,height=5.06, units = 'in', device = 'png',dpi=300)
ggsave('../figures/Intonexon_depth_cov_vs_geneposition.pdf',plot, width=10.5,height=5.06, units = 'in', device = 'pdf')

```

```{r fig-intronexon-biplot, include=FALSE, eval=FALSE, echo=FALSE, message=FALSE, warning=FALSE, fig=TRUE, fig.cap='Methylation is lower at the beginning of genes and more so in the exons.', fig.align='center',fig.width=15, fig.height=8,out.width = "10in"}

knitr::include_graphics("../figures/Intonexon_depth_cov_vs_geneposition.png")

```

```{r get data for fig-intronexon-boxplot, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}
df=read.csv('../data/Gene-body_methylation/ALL.downstream.introns_vs_exons_long.tsv',sep='\t',header=T,row.names = 1)
df$meth_freq=df$meth_cov*df$meth_depth
df$ind=factor(df$ind,levels=c('Pec','R07B-5','P08H-3'), labels=c('P. echinospica','R. piscesae','P. palmiformis'))

# conditions=(df$type %in% c('intron','exon') & (df$callCpGs_count >= 10))
conditions=(df$type %in% c('intron','exon'))


dat1=df[conditions ,] %>% dcast(.,Parent+feature_posi+ind~type,value.var='meth_depth')
dat1$var='meth_depth'

dat2=df[conditions ,] %>% dcast(.,Parent+feature_posi+ind~type,value.var='meth_cov')
dat2$var='meth_cov'

dat3=df[conditions ,] %>% dcast(.,Parent+feature_posi+ind~type,value.var='meth_freq')
dat3$var='meth_freq'

dat4=rbind(dat1,dat2,dat3) %>% .[complete.cases(.),]

dat=dat4  %>% melt(id.vars=c('Parent','feature_posi','ind','var'), variable.name = 'type')

dat %>% str
dat %>% colnames


dd=dat %>% group_by(ind,var,type,feature_posi) %>% summarise(
                                                          min_40=quantile(value,.4,na.rm=T),
                                                          max_60=quantile(value,.6,na.rm=T),
                                                          min_25=quantile(value,.25,na.rm=T),
                                                          max_75=quantile(value,.75,na.rm=T),
                                                          median=quantile(value,.5,na.rm=T),
                                                          min_10=quantile(value,.1,na.rm=T),
                                                          max_90=quantile(value,.9,na.rm=T),
                                                          mean=mean(value)) %>% as.data.frame


max_c=dat %>% group_by(feature_posi,ind,var) %>% summarise(count=n())


max_count=max(max_c$count)
color_count='#006400'

plot=ggplot(subset(dd,var=='meth_freq' & feature_posi < 21))+
  geom_step(data=subset(max_c,var=='meth_freq' & feature_posi < 21),aes(x=feature_posi,y=count/max_count),stat='identity',color=color_count)+
  geom_boxplot(aes(color=type,fill=type,x=factor(feature_posi),
                   ymin = min_25,
                   lower=min_40,
                   middle=median,
                   upper=max_60,
                   ymax=max_75,
),stat='identity',alpha=0.6)+
  # geom_line(aes(color=type,x=feature_posi,y = median),show.legend = F)+
  # geom_point(aes(color=type,x=feature_posi,y = median),size=0.8)+
  # scale_x_continuous(limits=c(1,20), breaks=c(1,5,10,15,20),name='Position in gene',labels = label_ordinal())+
    scale_x_discrete(labels=c('1st','','','','5th','','','','','10th','','','','','15th','','','','','20th'),name='Feature position in gene')+
    scale_y_continuous(sec.axis = sec_axis(~.*max_count/2, name = "Exon-intron pair count",labels= scales::label_number_si()), name='Methylation level',labels=scales::percent_format(accuracy=1))+
  scale_color_manual(values=c('exon'='#0025ff','intron'='#ff8300'), name='Feature type')+
  scale_fill_manual(values=c('exon'='#0025ff','intron'='#ff8300'), name='Feature type')+
  facet_wrap(~ind)+
  theme_bw()+
  theme(strip.text.x = element_text(face='bold.italic'),
        strip.background = element_blank(),
        axis.line.y.right = element_line(color=color_count),
        axis.text.y.right = element_text(color=color_count),
        axis.ticks.y.right = element_line(color=color_count),
        axis.title.y.right = element_text(color=color_count))

plot
ggsave('../figures/Intonexon_meth_summary_vs_geneposition.png',plot, device = 'png', width=10, height=3.5, units = "in",dpi=600)

ggsave('../figures/Intonexon_meth_summary_vs_geneposition.pdf',plot, device = 'pdf', width=10, height=3.5, units = "in",dpi=600)

```

```{r stats-wilcox, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}
df=read.csv('../data/Gene-body_methylation/ALL.downstream.introns_vs_exons_long.tsv',sep='\t',header=T,row.names = 1)
df$meth_freq=df$meth_cov*df$meth_depth
df$ind=factor(df$ind,levels=c('Pec','R07B-5','P08H-3'), labels=c('P. echinospica','R. piscesae','P. palmiformis'))

# conditions=(df$type %in% c('intron','exon') & (df$callCpGs_count >= 10))
conditions=(df$type %in% c('intron','exon'))


dat1=df[conditions ,] %>% dcast(.,Parent+feature_posi+ind~type,value.var='meth_depth')
dat1$var='meth_depth'

dat2=df[conditions ,] %>% dcast(.,Parent+feature_posi+ind~type,value.var='meth_cov')
dat2$var='meth_cov'

dat3=df[conditions ,] %>% dcast(.,Parent+feature_posi+ind~type,value.var='meth_freq')
dat3$var='meth_freq'

dat4=rbind(dat1,dat2,dat3) %>% .[complete.cases(.),]

dat=dat4  %>% melt(id.vars=c('Parent','feature_posi','ind','var'), variable.name = 'type')

dat %>% str
dat %>% colnames


#### stats ####

#### Does methylation (cov and depth) increase with gene length? ####
feat='exon'
d=dcast(subset(df,type==feat),ind+Parent~feature_posi, value.var = 'meth_cov')

cor.test(d$`1`,d$`2`, method='spearman', paired=T,alternative='greater') # p-value < 2.2e-16

wilcox.test(df[df$type=='exon',]$meth_depth,df[df$type=='exon',]$feature_posi,paired = T,alternative = 'less',)
wilcox.test(df[df$type=='intron',]$meth_cov,df[df$type=='intron',]$feature_posi,paired = T,alternative = 'less',)
wilcox.test(df[df$type=='intron',]$meth_depth,df[df$type=='intron',]$feature_posi,paired = T,alternative = 'less',)

#### Is methylation in intron and exon correlated? YES ####

df %>% str

d=dcast(subset(df),ind+Parent+feature_posi~type, value.var = 'meth_depth')

d %>% str
d %>% head

cor.test(d$exon,d$intron, method='spearman', paired=T,alternative='greater') # p-value < 2.2e-16

library("ppcor")
dd=d[complete.cases(d[,c('ind','Parent','feature_posi','exon','intron')]),]
pcor.test(dd$exon,dd$intron,dd$feature_posi,method='spearman') # controlling for feature position

#### Are introns more methylated than exons? Introns have lower methylation depth and overall level than exons but higher methylation coverage

dat1 %>% head
wilcox.test(dat1$intron,dat1$exon,paired = T,alternative = 'less')
ggplot(sample_n(dat1,1000))+
  geom_point(aes(x=intron,y=exon))+
  geom_abline()

wilcox.test(dat3$intron,dat3$exon,paired = T,alternative = 'less')
ggplot(sample_n(dat3,1000))+
  geom_point(aes(x=intron,y=exon))+
  geom_abline()

dat2 %>% head
wilcox.test(dat2$intron,dat2$exon,paired = T,alternative = 'less')
ggplot(sample_n(dat2,1000))+
  geom_point(aes(x=intron,y=exon))+
  geom_abline()

```

```{r stats-KS, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}
df=read.csv('../data/Gene-body_methylation/ALL.downstream.introns_vs_exons_long.tsv',sep='\t',header=T,row.names = 1)
df$meth_freq=df$meth_cov*df$meth_depth
df$ind=factor(df$ind,levels=c('Pec','R07B-5','P08H-3'), labels=c('P. echinospica','R. piscesae','P. palmiformis'))

# conditions=(df$type %in% c('intron','exon') & (df$callCpGs_count >= 10))
conditions=(df$type %in% c('intron','exon'))


dat1=df[conditions ,] %>% dcast(.,Parent+feature_posi+ind~type,value.var='meth_depth')
dat1$var='meth_depth'

dat2=df[conditions ,] %>% dcast(.,Parent+feature_posi+ind~type,value.var='meth_cov')
dat2$var='meth_cov'

dat3=df[conditions ,] %>% dcast(.,Parent+feature_posi+ind~type,value.var='meth_freq')
dat3$var='meth_freq'

dat4=rbind(dat1,dat2,dat3) %>% .[complete.cases(.),]

dat=dat4  %>% melt(id.vars=c('Parent','feature_posi','ind','var'), variable.name = 'type')

dat %>% str
dat %>% colnames


#### stats ####

#### Does methylation (cov and depth) increase with gene length? ####
feat='exon'
d=dcast(subset(df,type==feat),ind+Parent~feature_posi, value.var = 'meth_cov')

cor.test(d$`1`,d$`2`, method='spearman', paired=T,alternative='greater') # p-value < 2.2e-16

KS.test(df[df$type=='exon',]$meth_depth,df[df$type=='exon',]$feature_posi,paired = T,alternative = 'greater',)
KS.test(df[df$type=='intron',]$meth_cov,df[df$type=='intron',]$feature_posi,paired = T,alternative = 'greater',)
KS.test(df[df$type=='intron',]$meth_depth,df[df$type=='intron',]$feature_posi,paired = T,alternative = 'greater',)

#### Is methylation in intron and exon correlated? YES ####

df %>% str

d=dcast(subset(df),ind+Parent+feature_posi~type, value.var = 'meth_depth')

d %>% str
d %>% head

cor.test(d$exon,d$intron, method='spearman', paired=T,alternative='greater') # p-value < 2.2e-16

library("ppcor")
dd=d[complete.cases(d[,c('ind','Parent','feature_posi','exon','intron')]),]
pcor.test(dd$exon,dd$intron,dd$feature_posi,method='spearman') # controlling for feature position

#### Are introns more methylated than exons? Introns have lower methylation depth and overall level than exons but higher methylation coverage

dat1 %>% head
KS.test(dat1$intron,dat1$exon,paired = T,alternative = 'greater')
ggplot(sample_n(dat1,1000))+
  geom_point(aes(x=intron,y=exon))+
  geom_abline()

KS.test(dat3$intron,dat3$exon,paired = T,alternative = 'greater')
ggplot(sample_n(dat3,1000))+
  geom_point(aes(x=intron,y=exon))+
  geom_abline()

dat2 %>% head
KS.test(dat2$intron,dat2$exon,paired = T,alternative = 'greater')
ggplot(sample_n(dat2,1000))+
  geom_point(aes(x=intron,y=exon))+
  geom_abline()

```

### KOG/KEGG functional categories

```{r KOG methylation ecds figure, eval=FALSE, warning=FALSE, include=FALSE}
df=read.csv('../data/Gene-body_methylation/ALL.downstream_genes_withCG_CDS.tsv',header=T,sep='\t',check.names = F)
colnames(df)[which(colnames(df)=='Name')]='mRNA'

## add functional annotations
kog_hier=read.csv('../data/Assemblies_and_annotations/reference_KOG_hierarchy.txt',header=T,sep='\t',col.names = c('KOG_lvl1_id','KOG','KOG_lvl1'))
brite_hier=read.csv('../data/Assemblies_and_annotations/reference_BRITE_hierarchy_clean.tsv',header=T,sep='\t')
orth=read.csv('../data/Assemblies_and_annotations/Final_annotations.txt',header=T,sep='\t') %>% .[.$flag=='good_annot',c('genome','qseqid','Description','KO','KOG','KOG_longname')] %>% unique
colnames(orth)[which(colnames(orth)=='qseqid')]='mRNA'
orth=merge(orth,kog_hier,by='KOG',all.x=T)
orth=merge(orth,brite_hier,by='KO',all.x=T)

df=merge(df,orth,by=c('genome','mRNA'),all.x=T)

## KOG lvl1

df %>% str

KOG_lvl1=c("seq_id","start", "end", "CpGoe","CpGs_count", "mRNA", 
"callCpGs_count", "meth_meandepth", "meth_meancov", "meth_meanfreq", 
 "gene_Length",'cds_Length', "genome",'KOG_lvl1')
dat=df[df$callCpGs_count>=10,KOG_lvl1] %>% unique # remove genes with less than 10 CpGs called for methylation


dat$meth_binary=ifelse((dat$meth_meancov>0.75) & (dat$meth_meandepth>0.50),1,0)

dat$genome=factor(dat$genome, levels=c('P. echinospica','R. piscesae','P. palmiformis'))


dat %>% 
  group_by(KOG_lvl1,genome) %>%
  summarise(., count=n(), meth_meanfreq=mean(meth_meanfreq)) %>%
  as.data.frame 

# reorder KOG levels by overall highest to lowest methylation frequency
dat$KOG_lvl1[is.na(dat$KOG_lvl1)]=''
KOG_lvl_sorted=dat %>% 
  group_by(KOG_lvl1) %>%
  summarise(., count=n(), meth_meanfreq=mean(meth_meanfreq),meth_meanCpGoe=mean(CpGoe)) %>%
  as.data.frame %>% .[order(.$meth_meanfreq,decreasing = T),'KOG_lvl1']


KOG_lvl_sorted=c(KOG_lvl_sorted[KOG_lvl_sorted != ''],'')

dat$KOG_lvl1=factor(dat$KOG_lvl1, levels=rev(KOG_lvl_sorted), labels=c('not annotated',rev(KOG_lvl_sorted[-length(KOG_lvl_sorted)])))

###########


ann_text <- data.frame(x = c(0.374,0.825),y = c(0.825,0.125),lab = c('Less\nmethylation','More\nmethylation'),genome = factor(c('P. echinospica','P. echinospica'),levels = c('P. echinospica','R. piscesae','P. palmiformis')))
arrows <-tibble(x1 = c(0.375, 0.75), x2 = c(0.25, 0.875), y1 = c(0.75, 0.125), y2 = c(0.825, 0.05), genome = factor(c('P. echinospica','P. echinospica'),levels = c('P. echinospica','R. piscesae','P. palmiformis')))


textsize=ggplot2::.pt

mult_format <- function() {
     function(x) format(100*x,digits = 1) 
}

ggplot(dat)+
  stat_ecdf(aes(meth_meanfreq,color=KOG_lvl1),geom = "step",pad = FALSE)+
  geom_text(data=ann_text,aes(x=x,y=y,label=lab),size=textsize,hjust=0)+
  geom_curve(data = arrows, aes(x = x1, y = y1, xend = x2, yend = y2),arrow = arrow(length = unit(0.08, "inch")),curvature=0)+
  scale_colour_viridis(option="turbo",discrete = TRUE,direction = 1)+
  # scale_x_continuous(labels = label_percent(accuracy=1))+
    scale_x_continuous(labels=mult_format())+
  labs(y='Cummulative fraction of genes', x= 'Methylation level (%)',color='KOG functional category')+
  facet_wrap(~genome)+
  theme_bw()+
  theme(text=element_text(size=11),
    legend.title = element_text( size=9), 
    legend.text=element_text(size=9),
    strip.background = element_blank(),
    strip.text.x = element_text(face='bold.italic',size=11))

```

```{r}
ggsave("../figures/ECDF_KOGlvl1.png", width = 17.5, height = 3.8, units = "in", dpi=600)

ggsave("../figures/ECDF_KOGlvl1.pdf", width = 17.5, height = 3.8, units = "in")

```

```{r KOG CpGoe ecds figure, eval=FALSE, warning=FALSE, include=FALSE}
df=read.csv('../data/Gene-body_methylation/ALL.downstream_genes_withCG_CDS.tsv',header=T,sep='\t',check.names = F)
colnames(df)[which(colnames(df)=='Name')]='mRNA'

## add functional annotations
kog_hier=read.csv('../data/Assemblies_and_annotations/reference_KOG_hierarchy.txt',header=T,sep='\t',col.names = c('KOG_lvl1_id','KOG','KOG_lvl1'))
brite_hier=read.csv('../data/Assemblies_and_annotations/reference_BRITE_hierarchy_clean.tsv',header=T,sep='\t')
orth=read.csv('../data/Assemblies_and_annotations/Final_annotations.txt',header=T,sep='\t') %>% .[.$flag=='good_annot',c('genome','qseqid','Description','KO','KOG','KOG_longname')] %>% unique
colnames(orth)[which(colnames(orth)=='qseqid')]='mRNA'
orth=merge(orth,kog_hier,by='KOG',all.x=T)
orth=merge(orth,brite_hier,by='KO',all.x=T)

df=merge(df,orth,by=c('genome','mRNA'),all.x=T)

## KOG lvl1
df %>% str

KOG_lvl1=c("seq_id","start", "end", "CpGoe","CpGs_count", "mRNA", 
"callCpGs_count", "meth_meandepth", "meth_meancov", "meth_meanfreq", 
 "gene_Length",'cds_Length', "genome",'KOG_lvl1')
dat=df[df$callCpGs_count>=10,KOG_lvl1] %>% unique # remove genes with less than 10 CpGs called for methylation


dat$meth_binary=ifelse((dat$meth_meancov>0.75) & (dat$meth_meandepth>0.50),1,0)

dat$genome=factor(dat$genome, levels=c('P. echinospica','R. piscesae','P. palmiformis'))

# reorder KOG levels by overall highest to lowest methylation frequency
dat$KOG_lvl1[is.na(dat$KOG_lvl1)]=''

KOG_lvl_sorted=dat %>% 
  group_by(KOG_lvl1) %>%
  summarise(., count=n(), meth_meanCpGoe=mean(CpGoe)) %>%
  as.data.frame %>% .[order(.$meth_meanCpGoe,decreasing = F),'KOG_lvl1']

KOG_lvl_sorted=c(KOG_lvl_sorted[KOG_lvl_sorted != ''],'')

dat$KOG_lvl1=factor(dat$KOG_lvl1, levels=rev(KOG_lvl_sorted), labels=c('not annotated',rev(KOG_lvl_sorted[-length(KOG_lvl_sorted)])))

###########

dat %>% str

ggplot(dat)+
  stat_ecdf(aes(CpGoe,color=KOG_lvl1),geom = "step",pad = FALSE)+
  # geom_text(data=ann_text,aes(x=x,y=y,label=lab),size=textsize,hjust=0)+
  # geom_curve(data = arrows, aes(x = x1, y = y1, xend = x2, yend = y2),arrow = arrow(length = unit(0.08, "inch")),curvature=0)+
  scale_colour_viridis(option="turbo",discrete = TRUE,direction = 1)+
  scale_x_continuous(limits=c(0,2))+
    # scale_x_continuous(labels=mult_format())+
  labs(y='Cummulative fraction of genes', x= 'CpGoe',color='CpG expected/observed')+
  facet_wrap(~genome)+
  theme_bw()+
  theme(text=element_text(size=11),
    legend.title = element_text( size=9), 
    legend.text=element_text(size=9),
    strip.background = element_blank(),
    strip.text.x = element_text(face='bold.italic',size=11))

ggsave("../figures/ECDF_KOGlvl1_CpGoe.png", width = 17.5, height = 3.8, units = "in", dpi=600)

ggsave("../figures/ECDF_KOGlvl1_CpGoe.pdf", width = 17.5, height = 3.8, units = "in")

```

```{r KEGG methylation ecds figure, eval=FALSE, warning=FALSE, include=FALSE}
df=read.csv('../data/Gene-body_methylation/ALL.downstream_genes_withCG_CDS.tsv',header=T,sep='\t',check.names = F)
colnames(df)[which(colnames(df)=='Name')]='mRNA'

## add functional annotations
kog_hier=read.csv('../data/Assemblies_and_annotations/reference_KOG_hierarchy.txt',header=T,sep='\t',col.names = c('KOG_lvl1_id','KOG','KOG_lvl1'))
brite_hier=read.csv('../data/Assemblies_and_annotations/reference_BRITE_hierarchy_clean.tsv',header=T,sep='\t')
orth=read.csv('../data/Assemblies_and_annotations/Final_annotations.txt',header=T,sep='\t') %>% .[.$flag=='good_annot',c('genome','qseqid','Description','KO','KOG','KOG_longname')] %>% unique
colnames(orth)[which(colnames(orth)=='qseqid')]='mRNA'
orth=merge(orth,kog_hier,by='KOG',all.x=T)
orth=merge(orth,brite_hier,by='KO',all.x=T)

df=merge(df,orth,by=c('genome','mRNA'),all.x=T)

## KEGG_lvl2
df %>% colnames

KEGG_lvl2=c("seq_id","start", "end", "CpGoe","CpGs_count", "mRNA", 
"callCpGs_count", "meth_meandepth", "meth_meancov", "meth_meanfreq", 
 "gene_Length",'cds_Length', "genome",'KEGG_lvl1','KEGG_lvl2')

bad_KEGG_lvl1=c("Brite Hierarchies","Human Diseases", "Not Included in Pathway or Brite") # remove irrelevant KEGG_lvl1 categories

dat=df[df$callCpGs_count>=10,] # remove genes with less than 10 CpGs called for methylation
dat=dat[,KEGG_lvl2] %>% unique


dat$genome=factor(dat$genome, levels=c('P. echinospica','R. piscesae','P. palmiformis'))
dat$meth_binary=ifelse((dat$meth_meancov>0.75) & (dat$meth_meandepth>0.50),1,0)
dat[dat$KEGG_lvl1 %in% bad_KEGG_lvl1, c('KEGG_lvl1','KEGG_lvl2')]=''


bad_KEGG_lvl2=dat %>% 
  group_by(KEGG_lvl2,genome) %>%
  summarise(., count=n(), meth_meanfreq=mean(meth_meanfreq)) %>%
  as.data.frame %>% .[.$count<50,c('KEGG_lvl2','genome')] #poorly represented KEGG_lvl2 categories (with less than 50 genes in each worm genomes)

filter=dat %>% 
  group_by(KEGG_lvl2,genome) %>%
  summarise(., count=n(), meth_meanfreq=mean(meth_meanfreq)) %>%
  as.data.frame %>% .[.$count>=50,c('KEGG_lvl2','genome')]

dat=merge(dat,filter, by=c('KEGG_lvl2','genome'), all.y=T) #remove poorly represented KEGG_lvl2 categories

dat$KEGG_lvl2[is.na(dat$KEGG_lvl2)]=''
KEGG_lvl2_sorted=dat %>% 
  group_by(KEGG_lvl2) %>%
  summarise(., count=n(), meth_meanfreq=mean(meth_meanfreq)) %>%
  as.data.frame %>% .[order(.$meth_meanfreq,decreasing = T),'KEGG_lvl2']

KEGG_lvl2_sorted=c(KEGG_lvl2_sorted[KEGG_lvl2_sorted != ''],'')

dat$KEGG_lvl2=factor(dat$KEGG_lvl2, levels=rev(KEGG_lvl2_sorted), labels=c('not annotated',rev(KEGG_lvl2_sorted[1:length(KEGG_lvl2_sorted)-1])))

#####


ann_text <- data.frame(x = c(0.374,0.825),y = c(0.825,0.125),lab = c('Less\nmethylation','More\nmethylation'),genome = factor(c('P. echinospica','P. echinospica'),levels = c('P. echinospica','R. piscesae','P. palmiformis')))
arrows <-tibble(x1 = c(0.375, 0.75), x2 = c(0.25, 0.875), y1 = c(0.75, 0.125), y2 = c(0.825, 0.05), genome = factor(c('P. echinospica','P. echinospica'),levels = c('P. echinospica','R. piscesae','P. palmiformis')))

textsize=ggplot2::.pt


ggplot(dat,aes(meth_meanfreq,color=KEGG_lvl2))+
  stat_ecdf(geom = "step",pad = FALSE)+
  geom_text(data=ann_text,aes(x=x,y=y,label=lab),size=textsize,hjust=0,color='black')+
  geom_curve(data = arrows, aes(x = x1, y = y1, xend = x2, yend = y2),arrow = arrow(length = unit(0.08, "inch")),curvature=0,color='black')+
  scale_colour_viridis(option="turbo",discrete = TRUE,direction = 1)+
  scale_x_continuous(labels = label_percent())+
  labs(y='Cummulative fraction of genes', x= 'Mean methylation frequency',color='KEGG functional category (with 50+ genes)')+
  facet_wrap(~genome)+
  theme_bw()+
  theme(strip.text.x = element_text(face = "italic"),
        legend.title = element_text( size=8), 
        legend.text=element_text(size=8))


ggsave("../figures/ECDF_KEGGlvl2_min50genes.png", width = 17.5, height = 3.8, units = "in",dpi=600)

ggsave("../figures/ECDF_KEGGlvl2_min50genes.pdf", width = 17.5, height = 3.8, units = "in")

```

```{r function to run Fisher tests1, eval=FALSE, include=FALSE}
# dat=d
# branch='All'

montecarlo_test_pval<-function(df,br,FUNC_level){
  ecdf_fun <- function(x,perc) ecdf(x)(perc) # determine the quantile of a value against a distribution
  test=br
  #generate random subsample with same number of genes as in test category
  df1=cbind(df[,1],df[,c('Core',test,FUNC_level)])
  colnames(df1)[1]='gene_id'
  len_core=df1[df1[,'Core']==1,]$gene_id %>% unique %>% length
  len_selected=df1[df1[,test]==1,]$gene_id %>% unique %>% length
  
  pval_distrib=c()
  n=0
  while (n<100) {
    # pick len_selected genes from core at random
    random_gene_set=sample((subset(df1,Core==1)$gene_id %>% unique), len_selected)
    # add random column to 
    dd=cbind(df1,'random'=as.numeric(df1$gene_id %in% random_gene_set))
    formula=as.formula(paste0(".~ ",FUNC_level))
    d<-aggregate(formula,dd,FUN = function(x) {sum(x > 0, na.rm = F)}, na.action = na.pass) 
    pval_distrib<-rbind(pval_distrib,c(dmvhyper(n=d[,c('Core')],x=d[,c('random')],k=sum(d[,c('random')]))))
    n=n+1}
  pval_test=dmvhyper(n=d[,c('Core')],x=d[,c(test)],k=sum(d[,c(test)]))
  
  final_pval<-ecdf_fun(sort(pval_distrib),pval_test)
  return(final_pval)} ## compare prop of distribution if test data to that of 100 trials of randomly sampled genes (same sample size as test data)

run_multihypergeometric_montecarlo_tests<- function(df,FUNC_level){
library("extraDistr")
  branches_to_test=names(df)[which(sapply(df, is.numeric))]

  pval<-data.frame(p.value=as.numeric(numeric()))
  for(br in branches_to_test[-1]){
    p.value<-montecarlo_test_pval(df,br,FUNC_level)
    new_pval<-as.data.frame(p.value, row.names=br, col.names='p.value')
    pval<-rbind(pval,new_pval)
  }

  pval$comp<-row.names(pval)
  pval$symbol<-ifelse(pval$p.value<=0.01,'***',
                ifelse(pval$p.value<=0.05,'**',
                ifelse(pval$p.value<=0.1,'*','')))

return(pval)
}


run_fisher_tests<- function(dat){
fisher<-data.frame(matrix(ncol = length(colnames(dat)[-1]), nrow = length(row.names(dat))),row.names=rownames(dat))
colnames(fisher) <- colnames(dat)[-1]

differences<-data.frame(matrix(ncol = length(colnames(dat)[-1]), nrow = length(row.names(dat))),row.names=rownames(dat))
colnames(differences) <- colnames(dat)[-1]

# br='meth_binary'
# category="Amino acid transport and metabolism "

for(br in colnames(fisher)){
  data<-dat[,c(colnames(dat)[1],br)] %>% replace(., is.na(.), 0)
  for (category in rownames(data)){
    success_in_sample<-data[rownames(data)==category,br]
    success_in_left_part<-data[rownames(data)==category,colnames(dat)[1]]-success_in_sample
    failure_in_sample<-sum(data[,br])-success_in_sample
    failure_in_left_part<-sum(data[rownames(data)!=category,colnames(dat)[1]])+success_in_sample
    p.value<-fisher.test(matrix(c(success_in_sample, success_in_left_part, failure_in_sample, failure_in_left_part), 2, 2),alternative = "two.sided")
    
    fisher[category,br]=p.value
    
    differences[category,br]=(data[rownames(data)==category,br]/sum(data))-(sum(data[rownames(data)==category,])/sum(data))*(sum(data[,br])/sum(data))
  }
  }

fisher[,br]=p.adjust(fisher[,br],method = "holm") # Holms correction for mltiple testing
fisher$categories<-rownames(fisher)
differences$categories<-rownames(differences)
breakdown<-cbind(melt(fisher, id.vars='categories',value.name='pvalue'), melt(differences, id.vars=c('categories'))[,3])
colnames(breakdown)[2]='comp'
colnames(breakdown)[4]='diff'
breakdown$pvalue<-as.numeric(breakdown$pvalue)
breakdown$diff<-as.numeric(breakdown$diff)*100 # expresses the difference in probability of the event happening
breakdown$symbol<-ifelse(breakdown$pvalue<=0.01,'***',
              ifelse(breakdown$pvalue<=0.05,'**',
              ifelse(breakdown$pvalue<=0.1,'*','')))
breakdown$sign<-ifelse(breakdown$diff<0,'depletion' ,'enrichment')

return(breakdown)
} # runs fisher test for enrichment on each category

plot_breakdown_with_counts<-function(breakdown,dat){
# level<-rownames(dat)
dat$level<-rownames(dat)
add_counts<-melt(dat,variable.name = 'comp',value.name = 'freq')
breakdown_w_counts<-merge(breakdown,add_counts, by.y=c('comp','level'), by.x=c('comp','categories'),all=T)
breakdown_w_counts$freq[!(breakdown_w_counts$comp %in% c('Pan')) & breakdown_w_counts$pvalue>0.1] <- NA
tot<-aggregate(data=breakdown_w_counts,freq~comp,FUN='sum')
colnames(tot)[2]='tot'
breakdown_w_counts<-merge(breakdown_w_counts,tot,by='comp',all.x=T)
breakdown_w_counts$relfreq<-paste(round(breakdown_w_counts$freq/breakdown_w_counts$tot*100),'%')
breakdown_w_counts$relfreq[breakdown_w_counts$relfreq =='NA %'] <- NA
breakdown_w_counts$symbol[breakdown_w_counts$comp =='moderate'] <- NA
breakdown_w_counts$sign[breakdown_w_counts$comp =='moderate'] <- NA
breakdown_w_counts$diff[breakdown_w_counts$comp =='moderate'] <- NA

plot<-ggplot(data=subset(breakdown_w_counts,comp!='Pan'))+
    geom_point(aes(x=comp,y=categories,fill=sign,size=abs(diff),alpha=pvalue),shape=21,color='grey')+
    scale_x_discrete(limits = levels(dat$comp)) +
    scale_fill_manual(name="",breaks=c("depletion", "enrichment"), values =c( "lightblue", "#EC7063")) +
    scale_size_continuous(range=c(5,20),breaks=c(5,20,50,75,100))+
    scale_alpha_continuous(range=c(1,0.2), limits=c(0,0.1), na.value = 0,breaks=c(0,0.01,0.05,0.1,1),guide=FALSE)+
    geom_text(aes(x=comp,y=categories,label=symbol),vjust=-0.5)+
    geom_text(aes(x=comp,y=categories,label=freq),vjust=0.5)+
    geom_text(aes(x=comp,y=categories,label=relfreq),hjust=-1)+
    theme_bw() +
    theme(axis.title.x=element_blank(),
          axis.text.x = element_text(angle=60,hjust=1))
  return(plot)
}
```

```{r run KOG lvl1 hypergeometric and fisher tests and format data frame for figure, eval=FALSE, echo=FALSE, warning=FALSE, include=FALSE}
# before running this block, the block above "function to run Fisher tests" should be run
df=read.csv('../data/Gene-body_methylation/ALL.downstream_genes_withCG_CDS.tsv',header=T,sep='\t',check.names = F)
colnames(df)[which(colnames(df)=='Name')]='mRNA'

## add functional annotations
kog_hier=read.csv('../data/Assemblies_and_annotations/reference_KOG_hierarchy.txt',header=T,sep='\t',col.names = c('KOG_lvl1_id','KOG','KOG_lvl1'))
brite_hier=read.csv('../data/Assemblies_and_annotations/reference_BRITE_hierarchy_clean.tsv',header=T,sep='\t')
orth=read.csv('../data/Assemblies_and_annotations/Final_annotations.txt',header=T,sep='\t') %>% .[.$flag=='good_annot',c('genome','qseqid','Description','KO','KOG','KOG_longname')] %>% unique
colnames(orth)[which(colnames(orth)=='qseqid')]='mRNA'
orth=merge(orth,kog_hier,by='KOG',all.x=T)
orth=merge(orth,brite_hier,by='KO',all.x=T)

df=merge(df,orth,by=c('genome','mRNA'),all.x=T)

# #### 1) run hypergeometric montecarlo test ####

KOG_lvl1=c("seq_id","start", "end", "CpGoe","CpGs_count", "mRNA", 
"callCpGs_count", "meth_meandepth", "meth_meancov", "meth_meanfreq", 
 "gene_Length",'cds_Length', "genome",'KOG_lvl1')
dat=df[df$callCpGs_count>=10,KOG_lvl1] %>% unique # remove genes with less than 10 CpGs called for methylation
dat$meth_binary=ifelse((dat$meth_meancov>0.75) & (dat$meth_meandepth>0.50),1,0)

dat$genome=factor(dat$genome, levels=c('P. echinospica','R. piscesae','P. palmiformis'))

dat %>% head

dat$KOG_lvl1[is.na(dat$KOG_lvl1)]=''
dat$KOG_lvl1=factor(dat$KOG_lvl1, levels=sort(unique(dat$KOG_lvl1)), labels=c('not annotated',sort(unique(dat$KOG_lvl1))[-1]))

# data source for hypergeometric test

dat['Core']=1
dat %>% str

# data source for fisher test
dd = dat %>% 
  group_by(KOG_lvl1,genome) %>%
  summarise(., core=sum(Core), methylated=sum(meth_binary), nonmethylated=sum(meth_binary==0)) %>%
  as.data.frame


pval_df=data.frame(genome=as.character(),p.value=as.numeric(numeric()),comp=as.character(),symbol=as.character())
breakdown_df=data.frame(categories=as.character(), comp=as.character(), pvalue=as.numeric(numeric()), diff=as.numeric(numeric()), symbol=as.character(), sign=as.character(),genome=as.character())

# run hyper geometric and fisher test for each genome

for (gen in c('P. palmiformis','R. piscesae','P. echinospica')){
  
pval<-run_multihypergeometric_montecarlo_tests(dat[dat$genome==gen,c('mRNA','Core','meth_binary','KOG_lvl1')],'KOG_lvl1')
pval_df=rbind(pval_df,as.data.frame(c(genome=gen,pval)))

#### 2) fisher test for over-representation of certain categories ####

d<-dd[dd$genome==gen,!(colnames(dd) %in% c('genome'))]

row.names(d)<-d$KOG_lvl1
d=d[,!(c(colnames(d) %in% c('KOG_lvl1')))]


breakdown<-run_fisher_tests(d)
breakdown$genome=gen
breakdown_df=rbind(breakdown_df,breakdown)

}

pval_df
breakdown_df

#### Create dataframe for figure construction ####
order=d[order(d$methylated/d$core),] %>% rownames
core_lvl_order<-c('not annotated',order[order != "not annotated"])
# core_lvl_order=KOG_lvl_sorted_ECDF # same order as ECDF figure

dd$KOG_lvl1 <- factor(dd$KOG_lvl1,levels = core_lvl_order)
dd$nonmethylated<-dd$core-dd$methylated

dd2<-melt(dd,id.vars = c('genome','KOG_lvl1'))
colnames(dd2)[2]='categories'
colnames(dd2)[3]='comp'
colnames(dd2)[4]='count'

dat<-merge(breakdown_df,dd2,by=c('genome','categories','comp'),all=T)
setdiff(core_lvl_order,unique(dat$categories))

dat$categories <- factor(dat$categories,levels =core_lvl_order)
dat$sign[dat$pvalue>=0.05] <- 'NS'
dat$symbol[is.na(dat$symbol)] <- ''
tot<-aggregate(count~comp,dat, function(x) sum(x))
colnames(tot)[2]='tot'
dat<-merge(dat,tot, by=c('comp'),all=T)
head(dat)
tail(head(dat))
dput(levels(dat$comp))

dat=dat[order(dat$categories),]

dat[dat$symbol=='**','sign']<-'NS'
dat[is.na(dat$pvalue),'sign']<-'NA'
dat$sign=factor(dat$sign,levels=c('NA','enrichment','depletion','NS'),labels=c('NA','enrichment','depletion','non-significant'))

dat$comp=factor(dat$comp,levels=c('nonmethylated','methylated', 'core'),labels=c('low-intermediate','high', 'core'))
dat$genome=factor(dat$genome, levels=c('P. echinospica','R. piscesae','P. palmiformis'))

library(ggpattern)

dat$categories %>% unique %>% dput  

plt=dat[!(dat$comp %in% c('core')),]
plt[plt$comp=='low-intermediate','sign']='NA'
plt[plt$comp=='low-intermediate',]
# write.table(plt,'../data/Gene-body_methylation/Fisher_tests_results_genebody.txt',col.names=T,sep='\t',row.names=F)

# ggplot()+
#   geom_col_pattern(data=plt,aes(y=categories,x=count,fill=comp,pattern_fill=sign,pattern_colour=comp),
#                    pattern_spacing=0.2,
#                    pattern_density = 0.3,
#                    pattern_key_scale_factor=0.1,
#                    width=0.8,
#                    pattern = 'wave',
#                    colour='black',
#                    position='fill')+
#   scale_pattern_colour_manual(values=c('methylated'='grey30','non-methylated'='grey70'),guide='none')+
#   scale_pattern_fill_manual(values=c('NA'='transparent','depletion'='blue','enrichment'='red','non-significant'='grey30'),name="Fisher's exact test",limits = c('depletion', 'enrichment','non-significant'))+
#   scale_fill_manual(values=c('methylated'='grey30','non-methylated'='grey70'),name='Gene methylation state',guide = guide_legend(override.aes = list(pattern = "none")))+
# 
#   theme_bw()+
#   facet_wrap(~genome)+
#   theme(
#         axis.title.x = element_blank(),
#         axis.title.y = element_blank(),
#         )

plot=ggplot()+
  geom_col_pattern(data=plt,aes(y=categories,x=count,fill=comp,pattern_fill=sign,pattern_colour=comp),
                   # pattern_spacing=0.2,
                   pattern_density = 0.65,
                   pattern_key_scale_factor=0.2,
                   width=0.8,
                   pattern = 'circle',
                   colour='black',
                   position='fill')+
  # scale_x_continuous(labels = label_percent())+
  scale_pattern_colour_manual(values=c('high'='grey30','low-intermediate'='grey70'),guide='none')+
  scale_pattern_fill_manual(values=c('NA'='grey70','depletion'='blue','enrichment'='red','non-significant'='grey30'),name="Fisher's exact test",limits = c('depletion', 'enrichment'))+
  scale_fill_manual(values=c('high'='grey30','low-intermediate'='grey70'),name='Gene methylation',guide = guide_legend(override.aes = list(pattern = "none")))+
  labs(x='Relative abundance')+
  theme_bw()+
  facet_wrap(~genome)+
  theme(axis.title.y = element_blank(),
        strip.background = element_blank(),
        strip.text.x = element_text(face='bold.italic',size=11)
        )


plot

ggsave("../figures/Fisher_KOGlvl1.png",plot, width = 350, height = 150, units = "mm", dpi=600)

ggsave("../figures/Fisher_KOGlvl1.pdf",plot, width = 350, height = 150, units = "mm")

```

```{r run KEGG lvl2 hypergeometric and fisher tests and format data frame for figure, eval=FALSE, echo=FALSE, warning=FALSE, include=FALSE}
df=read.csv('../data/Gene-body_methylation/ALL.downstream_genes_withCG_CDS.tsv',header=T,sep='\t',check.names = F)
colnames(df)[which(colnames(df)=='Name')]='mRNA'

## add functional annotations
kog_hier=read.csv('../data/Assemblies_and_annotations/reference_KOG_hierarchy.txt',header=T,sep='\t',col.names = c('KOG_lvl1_id','KOG','KOG_lvl1'))
brite_hier=read.csv('../data/Assemblies_and_annotations/reference_BRITE_hierarchy_clean.tsv',header=T,sep='\t')
orth=read.csv('../data/Assemblies_and_annotations/Final_annotations.txt',header=T,sep='\t') %>% .[.$flag=='good_annot',c('genome','qseqid','Description','KO','KOG','KOG_longname')] %>% unique
colnames(orth)[which(colnames(orth)=='qseqid')]='mRNA'
orth=merge(orth,kog_hier,by='KOG',all.x=T)
orth=merge(orth,brite_hier,by='KO',all.x=T)

df=merge(df,orth,by=c('genome','mRNA'),all.x=T)

## KEGG_lvl2
df %>% colnames

KEGG_lvl2=c("seq_id","start", "end", "CpGoe","CpGs_count", "mRNA", 
"callCpGs_count", "meth_meandepth", "meth_meancov", "meth_meanfreq", 
 "gene_Length",'cds_Length', "genome",'KEGG_lvl1','KEGG_lvl2')

bad_KEGG_lvl1=c("Brite Hierarchies","Human Diseases", "Not Included in Pathway or Brite") # remove irrelevant KEGG_lvl1 categories

dat=df[df$callCpGs_count>=10,] # remove genes with less than 10 CpGs called for methylation
dat=dat[,KEGG_lvl2] %>% unique


dat$genome=factor(dat$genome, levels=c('P. echinospica','R. piscesae','P. palmiformis'))

dat$meth_binary=ifelse((dat$meth_meancov>0.75) & (dat$meth_meandepth>0.50),1,0)

dat[dat$KEGG_lvl1 %in% bad_KEGG_lvl1, c('KEGG_lvl1','KEGG_lvl2')]=''


bad_KEGG_lvl2=dat %>% 
  group_by(KEGG_lvl2,genome) %>%
  summarise(., count=n(), meth_meanfreq=mean(meth_meanfreq)) %>%
  as.data.frame %>% .[.$count<50,c('KEGG_lvl2','genome')] #poorly represented KEGG_lvl2 categories (with less than 50 genes in each worm genomes)

filter=dat %>% 
  group_by(KEGG_lvl2,genome) %>%
  summarise(., count=n(), meth_meanfreq=mean(meth_meanfreq)) %>%
  as.data.frame %>% .[.$count>=50,c('KEGG_lvl2','genome')]

dat=merge(dat,filter, by=c('KEGG_lvl2','genome'), all.y=T) #remove poorly represented KEGG_lvl2 categories

dat$KEGG_lvl2[is.na(dat$KEGG_lvl2)]=''
KEGG_lvl2_sorted=dat %>% 
  group_by(KEGG_lvl2) %>%
  summarise(., count=n(), meth_meanfreq=mean(meth_meanfreq)) %>%
  as.data.frame %>% .[order(.$meth_meanfreq,decreasing = T),'KEGG_lvl2']

KEGG_lvl2_sorted=c(KEGG_lvl2_sorted[KEGG_lvl2_sorted != 'not annotated'],'not annotated')

dat$KEGG_lvl2=factor(dat$KEGG_lvl2, levels=KEGG_lvl2_sorted, labels=KEGG_lvl2_sorted)


dat['Core']=1
dat$genome=factor(dat$genome, levels=c('P. echinospica','R. piscesae','P. palmiformis'))

dat %>% str


############

# data source for fisher test
dd = dat %>% 
  group_by(KEGG_lvl2,genome) %>%
  summarise(., core=sum(Core), methylated=sum(meth_binary)) %>%
  as.data.frame

pval_df=data.frame(genome=as.character(),p.value=as.numeric(numeric()),comp=as.character(),symbol=as.character())
breakdown_df=data.frame(categories=as.character(), comp=as.character(), pvalue=as.numeric(numeric()), diff=as.numeric(numeric()), symbol=as.character(), sign=as.character(),genome=as.character())

# run hyper geometric and fisher test for each genome

for (gen in c('P. palmiformis','R. piscesae','P. echinospica')){
  
pval<-run_multihypergeometric_montecarlo_tests(dat[dat$genome==gen,c('mRNA','Core','meth_binary','KEGG_lvl2')],'KEGG_lvl2')
pval_df=rbind(pval_df,as.data.frame(c(genome=gen,pval)))

#### 2) fisher test for over-representation of certain categories ####

d<-dd[dd$genome==gen,!(colnames(dd) %in% c('genome'))]

row.names(d)<-d$KEGG_lvl2
d=d[,!(c(colnames(d) %in% c('KEGG_lvl2')))]

breakdown<-run_fisher_tests(d)
breakdown$genome=gen
breakdown_df=rbind(breakdown_df,breakdown)
}

# bk<-plot_breakdown_with_counts(breakdown,d)
# bk

pval_df
breakdown_df

#### Create dataframe for figure construction ####
order=d[order(d$methylated/d$core),] %>% rownames
core_lvl_order<-c('not annotated',order[order != "not annotated"])

dd$KEGG_lvl2 <- factor(dd$KEGG_lvl2,levels = core_lvl_order)
dd$nonmethylated<-dd$core-dd$methylated

dat<-melt(dd,id.vars = c('genome','KEGG_lvl2'))
colnames(dat)[2]='categories'
colnames(dat)[3]='comp'
colnames(dat)[4]='count'
dat<-merge(breakdown_df,dat,by=c('genome','categories','comp'),all=T)

dat$categories <- factor(dat$categories,levels =core_lvl_order)
dat$sign[dat$pvalue>=0.05] <- 'NS'
dat$symbol[is.na(dat$symbol)] <- ''
tot<-aggregate(count~comp,dat, function(x) sum(x))
colnames(tot)[2]='tot'
dat<-merge(dat,tot, by=c('comp'),all=T)
head(dat)
dput(levels(dat$comp))

dat=dat[order(dat$categories),]

dat[dat$symbol=='**','sign']<-'NS'
dat[is.na(dat$pvalue),'sign']<-'NA'
dat$sign=factor(dat$sign,levels=c('NA','enrichment','depletion','NS'),labels=c('NA','enrichment','depletion','non-significant'))
dat$comp=factor(dat$comp,levels=c('nonmethylated','methylated', 'core'),labels=c('low-intermediate','high', 'core'))
dat$genome=factor(dat$genome, levels=c('P. echinospica','R. piscesae','P. palmiformis'))

library(ggpattern)

dat$categories %>% unique %>% dput  

plt=dat[!(dat$comp %in% c('core')),]

plot=ggplot()+
  geom_col_pattern(data=plt,aes(y=categories,x=count,fill=comp,pattern_fill=sign,pattern_colour=comp),
                   # pattern_spacing=0.2,
                   pattern_density = 0.65,
                   pattern_key_scale_factor=0.2,
                   width=0.8,
                   pattern = 'circle',
                   colour='black',
                   position='fill')+
  scale_pattern_colour_manual(values=c('high'='grey30','low-intermediate'='grey70'),guide='none')+
  scale_pattern_fill_manual(values=c('NA'='transparent','depletion'='blue','enrichment'='red','non-significant'='grey30'),name="Fisher's exact test",limits = c('depletion', 'enrichment'))+
  scale_fill_manual(values=c('high'='grey30','low-intermediate'='grey70'),name='Gene methylation level',guide = guide_legend(override.aes = list(pattern = "none")))+
  labs(x='Relative abundance')+
  theme_bw()+
  facet_wrap(~genome)+
  theme(strip.text.x = element_text(face = "italic"),
        axis.title.y = element_blank(),
        )
plot

ggsave("../figures/Fisher_KEGGlvl2.png",plot, width = 350, height = 150, units = "mm", dpi=600)

ggsave("../figures/Fisher_KEGGlvl2.pdf",plot, width = 350, height = 150, units = "mm")

```


### Gene body Methylation vs transcription 

There is a positive correlation between Gene methylation and gene expression...

```{r prep for fig gene-corr-TPM, eval=FALSE, include=FALSE}

species=c('R. piscesae','P. palmiformis','P. echinospica',"R07B-5", "P08H-3", "Pec")
names(species)=c("R07B-5", "P08H-3", "Pec",'R. piscesae','P. palmiformis','P. echinospica')

meta=read.csv('../data/Gene-body_methylation/ALL.downstream_genes_withCG_CDS.tsv',header=T,sep='\t',check.names = F)
df=meta[,c('genome', "Name", "callCpGs_count", "meth_meandepth", "meth_meancov", "meth_meanfreq", "TPM_source", "TPM_value",'CpGs_count','gene_Length','cds_Length','CG%','cds_CG%','CpGoe')] %>% unique # Some genes are represented un multiple functional categories so duplicate rows need to be removed after clearing the functional columns

colnames(df)[which(colnames(df)=='CG%')]='GC'
colnames(df)[which(colnames(df)=='cds_CG%')]='cds_CG'
df$cds_CG=as.numeric(df$cds_CG)

df$gene_binary=ifelse(df$meth_meancov>0.75 & df$meth_meandepth>0.5,1,0)
df$genome=factor(df$genome, levels=c('P. echinospica','R. piscesae','P. palmiformis'))

# ggplot(df,aes(x=TPM_value,y=meth_meanfreq,color=factor(gene_binary)))+
#   geom_point(alpha=0.01)+
#   geom_smooth(method='lm',aes(color=factor(gene_binary)))+
#   facet_wrap(~genome)+
#   scale_x_continuous(trans = pseudo_log_trans(base = 10),
#                      breaks=c(1,100,1000,10000,100000),
#                      minor_breaks = c(5,50,500,5000,50000))+
#   theme_bw()

library(ggpmisc)

#Gene expression vs Methylation frequency
df %>% str

textsize=ggplot2::.pt

plot=ggplot(df[df$CpGs_count>=1,],aes(x=TPM_value,y=meth_meanfreq))+
  geom_point(alpha=0.01)+
  stat_poly_line() +
  stat_poly_eq(aes(label = paste(..eq.label.., ..rr.label.., ..p.value.label.., sep = "*`,`~")),color='blue',label.x = 'right',label.y='top') +
  facet_wrap(~genome)+
  scale_x_continuous(trans = pseudo_log_trans(base = 10),
                     breaks=c(1,100,1000,10000,100000),
                     minor_breaks = c(5,50,500,5000,50000),
                     labels=scales::label_number_si(),
                     name='Gene expression (TPM)')+
  scale_y_continuous(breaks=c(0,0.25,0.5,0.75,1),limits=c(0,1),
                     expand = c(0,0.1,0,0.2),
                     labels=scales::percent_format(accuracy=1),
                     name='Gene methylation level')+
  theme_bw()+
  theme(strip.text.x = element_text(face = 'bold.italic'),
        strip.background=element_blank())

plot

ggsave("../figures/gene-corr-TPM.png",plot, width = 10.7, height = 5.07, units = "in",dpi=300)
ggsave("../figures/gene-corr-TPM.pdf",plot, width = 10.7, height = 5.07, units = "in")

```

...but transcription is also dependent CDS length, and gene GC% content,

```{r prep for fig genechar-corr-TPM, eval=FALSE, include=FALSE}
meta=read.csv('../data/Gene-body_methylation/ALL.downstream_genes_withCG_CDS.tsv',header=T,sep='\t',check.names = F)
colnames(meta)[which(colnames(meta)=='CG%')]='GC'
colnames(meta)[which(colnames(meta)=='cds_CG%')]='cds_CG'
meta$cds_CG=as.numeric(meta$cds_CG)

meta  %>% str
df=meta[,c("genome","Name",'TPM_value','TPM_source','meth_meanfreq','gene_Length','cds_Length','GC','cds_CG','CpGoe')] %>% unique %>% .[complete.cases(.),]
df=df[!(is.na(df$TPM_value)) & df$TPM_source != "R08H-5-V_TPM",] %>% .[!duplicated(.),] # filter out genes with missing methylation data and low qual transcriptome for Ridgeia

# sample=sample_n(df[df$TPM_value>0,],5000)$Name
# dat=melt(df[df$Name %in% sample,], id.vars=c('genome','Name','TPM_source','TPM_value'))
dat=melt(df, id.vars=c('genome','Name','TPM_source','TPM_value'))
dat$genome=factor(dat$genome, levels=c('P. echinospica','R. piscesae','P. palmiformis'))

library(ggpmisc) # for stat_poly_eq
?stat_poly_eq

p1<-ggplot(dat[df$TPM_value>0 & dat$variable=='cds_Length',],aes(x=TPM_value,y=value))+
  geom_point(alpha=0.1)+
  stat_poly_line() +
  stat_poly_eq(aes(label = paste(..eq.label.., ..rr.label.., ..p.value.label.., sep = "*`,`~")),color='blue',label.x = 'right',label.y='top') +
 scale_x_continuous(trans = pseudo_log_trans(base = 10),
                     breaks=c(1,100,1000,10000,100000),
                     minor_breaks = c(5,50,500,5000,50000),
                     labels=scales::label_number_si(),
                     name='Gene expression (TPM)')+
  scale_y_continuous(trans = pseudo_log_trans(base = 10),
                     breaks=c(1,100,1000,10000,100000),
                     minor_breaks = c(5,50,500,5000,50000))+
  expand_limits(y=0)+
  facet_grid(.~genome,scales='free_y')+
  labs(y='CDS length')+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text.x = element_text(face = 'bold.italic'),
        strip.background=element_blank())


p2<-ggplot(dat[df$TPM_value>0 & dat$variable=='GC',],aes(x=TPM_value,y=value*100))+
  geom_point(alpha=0.1)+
  stat_poly_line() +
  stat_poly_eq(aes(label = paste(..eq.label.., ..rr.label.., ..p.value.label.., sep = "*`,`~")),color='blue',label.x = 'right',label.y='top') +
 scale_x_continuous(trans = pseudo_log_trans(base = 10),
                     breaks=c(1,100,1000,10000,100000),
                     minor_breaks = c(5,50,500,5000,50000),
                     labels=scales::label_number_si(),
                     name='Gene expression (TPM)')+
  scale_y_continuous()+
  expand_limits(y=0)+
  facet_grid(.~genome,scales='free_y')+
  labs(y='gene CG %')+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text.x = element_blank(),
        strip.background=element_blank())


plot=egg::ggarrange(p1,p2, nrow = 2, ncol=1,heights = c(1,1))

plot

ggsave("../figures/geneticparams-corr-TPM.png",plot, width = 13.4, height = 5.19, units = "in",dpi=300)
ggsave("../figures/geneticparams-corr-TPM.pdf",plot, width = 13.4, height = 5.19, units = "in",dpi=300)

```

Partial correlation for methylation and gene expression controlling for CDS length and gene GC content

```{r ppcor for gene, eval=FALSE, echo=FALSE, warning=FALSE, include=FALSE}
# genebody methylation/transcription correlation controlled for gene length, CG% and CpG content bias

library("ppcor")

species=c('R. piscesae','P. palmiformis','P. echinospica',"R07B-5", "P08H-3", "Pec")
names(species)=c("R07B-5", "P08H-3", "Pec",'R. piscesae','P. palmiformis','P. echinospica')

meta=read.csv('../data/Gene-body_methylation/ALL.downstream_genes_withCG_CDS.tsv',header=T,sep='\t',check.names = F)
df=meta[,c('genome', "Name", "callCpGs_count", "meth_meandepth", "meth_meancov", "meth_meanfreq", "TPM_source", "TPM_value",'CpGs_count','gene_Length','cds_Length','CG%','cds_CG%','CpGoe')] %>% unique # Some genes are represented un multiple functional categories so duplicate rows need to be removed after clearing the functional columns

colnames(df)[which(colnames(df)=='CG%')]='GC'
colnames(df)[which(colnames(df)=='cds_CG%')]='cds_GC'
df$cds_GC=as.numeric(df$cds_GC)

df %>% str

variables=list(c('cds_Length'),c('GC'),c('cds_Length',"GC"))

df=df[!(is.na(df$meth_meanfreq)) & !(is.na(df$TPM_value)) & df$TPM_source != "R08H-5-V_TPM",] %>% .[!duplicated(.),] # filter out genes with missing methylation data and low qual transcriptome for Ridgeia

cols=c("estimate", "p.value", "statistic", "n", "gp", "Method",'ctrl','genome')
pcor_df=setNames(data.frame(matrix(ncol = length(cols), nrow = 0)), cols)

for (genome in c('P. echinospica','R. piscesae','P. palmiformis')){
for (ctrl in variables){
  
tmp=pcor.test(df[df$genome==genome,]$meth_meanfreq,df[df$genome==genome,]$TPM_value,df[df$genome==genome,ctrl],method='spearman') 
tmp$ctrl=paste(ctrl,collapse = " + ") %>% print
tmp$genome=genome

tmp %>% str
tmp$estimate

pcor_df=rbind(pcor_df,tmp)

}}

pcor_df=pcor_df[,c('genome','ctrl','n','estimate','p.value')]

unique(pcor_df$ctrl) %>% dput

pcor_df$ctrl=factor(pcor_df$ctrl, levels=c("GC", "cds_Length", "cds_Length + GC"), labels=c('gene GC%','CDS length','gene GC% + CDS length'))


write.table(pcor_df,'../data/Gene-body_methylation/ALL.pcor_genemethylation_vs_TPM.txt',col.names=T,row.names=F,quote=F,sep='\t')


```


## Promoter methylation
### Profiles across orthologs

```{r, eval=FALSE, echo=FALSE, warning=FALSE, include=FALSE}

orth=read.csv('../data/Assemblies_and_annotations/clean_worms_aa_diamond.poff.tsv',header=T,sep='\t',check.names = F)
colnames(orth)[4]='P08H'
colnames(orth)[5]='Pec'
colnames(orth)[6]='R07B'
orth$orth_id=rownames(orth)

orth %>% str

d1=orth %>% separate_rows(P08H,sep = ',') %>% .[,c('P08H','orth_id',"# Species")] %>% unique
colnames(d1)[1]='mRNA'
colnames(d1)[3]='species_count'
d1$genome='P. palmiformis'

d2=orth %>% separate_rows(R07B,sep = ',') %>% .[,c('R07B','orth_id',"# Species")] %>% unique
colnames(d2)[1]='mRNA'
colnames(d2)[3]='species_count'
d2$genome='R. piscesae'

d3=orth %>% separate_rows(Pec,sep = ',') %>% .[,c('Pec','orth_id',"# Species")] %>% unique
colnames(d3)[1]='mRNA'
colnames(d3)[3]='species_count'
d3$genome='P. echinospica'

d=rbind(d1,d2,d3)

meta=read.csv('../data/Gene-body_methylation/ALL.downstream_genes_withCG_CDS_withflank.tsv',header=T,sep='\t')
meta$genome=factor(meta$genome,levels=c('P. echinospica','R. piscesae','P. palmiformis'))
meta=meta[meta$binary_idx %in% c('00','10','11','01'),]
meta %>% str

df=merge(meta,d, by=c('mRNA','genome'), all.x=T)

df %>% head

dat=list(Pec=df[df$genome=='P. echinospica',]$orth_id %>% unique,
      R07B=df[df$genome=='R. piscesae',]$orth_id %>% unique,
      P08H=df[df$genome=='P. palmiformis',]$orth_id %>% unique)
      
      
library(ggVennDiagram)

venn <- Venn(dat)
data <- process_data(venn)

# relatively few genes orthologs could be identified. the two tubeworms ave more orthologs in common compared to tubeworms/alvinellid

ggplot() +
  geom_sf(data = venn_region(data))+
  geom_sf_text(aes(label = name), data = venn_setlabel(data))+
  geom_sf_label(aes(label=count), fontface = "bold", data = venn_region(data)) +
  theme_void()


df[df$genome=='P. echinospica' & df$orth_id %in% venn_region(data)$item[[1]], 'mRNA'] %>% unique %>% length
df[df$genome=='P. echinospica' & df$orth_id %in% venn_region(data)$item[[4]], 'mRNA'] %>% unique %>% length
df[df$genome=='P. echinospica' & df$orth_id %in% venn_region(data)$item[[5]], 'mRNA'] %>% unique %>% length
df[df$genome=='P. echinospica' & df$orth_id %in% venn_region(data)$item[[7]], 'mRNA'] %>% unique %>% length

df[df$genome=='R. piscesae' & df$orth_id %in% venn_region(data)$item[[2]], 'mRNA'] %>% unique %>% length
df[df$genome=='R. piscesae' & df$orth_id %in% venn_region(data)$item[[4]], 'mRNA'] %>% unique %>% length
df[df$genome=='R. piscesae' & df$orth_id %in% venn_region(data)$item[[6]], 'mRNA'] %>% unique %>% length
df[df$genome=='R. piscesae' & df$orth_id %in% venn_region(data)$item[[7]], 'mRNA'] %>% unique %>% length

df[df$genome=='P. palmiformis' & df$orth_id %in% venn_region(data)$item[[3]], 'mRNA'] %>% unique %>% length
df[df$genome=='P. palmiformis' & df$orth_id %in% venn_region(data)$item[[5]], 'mRNA'] %>% unique %>% length
df[df$genome=='P. palmiformis' & df$orth_id %in% venn_region(data)$item[[6]], 'mRNA'] %>% unique %>% length
df[df$genome=='P. palmiformis' & df$orth_id %in% venn_region(data)$item[[7]], 'mRNA'] %>% unique %>% length


```

```{r gene and flank binaries, eval=FALSE, include=FALSE}

df_all=read.csv('../data/Gene-body_methylation/ALL.downstream_genes_withCG_CDS_withflank.tsv',header=T,sep='\t')
df_all$genome=factor(df_all$genome,levels=c('P. echinospica','R. piscesae','P. palmiformis'))
df=df_all[df_all$binary_idx %in% c('00','11','01','10'), c('mRNA','genome','binary_idx')] %>% unique
df_all=df_all[df_all$binary_idx %in% c('00','10','11','01'),]


library(tidyr)

orth=read.csv('../data/Assemblies_and_annotations/clean_worms_aa_diamond.poff.tsv',header=T,sep='\t',check.names = F)
colnames(orth)[4]='P08H'
colnames(orth)[5]='Pec'
colnames(orth)[6]='R07B'
orth$orth_id=rownames(orth)

orth %>% str

d1=orth %>% separate_rows(P08H,sep = ',') %>% .[,c('P08H','orth_id',"# Species")] %>% unique
colnames(d1)[1]='mRNA'
colnames(d1)[3]='species_count'
d1$genome='P. palmiformis'

d2=orth %>% separate_rows(R07B,sep = ',') %>% .[,c('R07B','orth_id',"# Species")] %>% unique
colnames(d2)[1]='mRNA'
colnames(d2)[3]='species_count'
d2$genome='R. piscesae'

d3=orth %>% separate_rows(Pec,sep = ',') %>% .[,c('Pec','orth_id',"# Species")] %>% unique
colnames(d3)[1]='mRNA'
colnames(d3)[3]='species_count'
d3$genome='P. echinospica'

d=rbind(d1,d2,d3)

df=merge(df_all,d,by=c('mRNA','genome'), all.x=T)

######
df %>% str
# genome='P. echinospica'
# bin_idx='01'


cols=c("source_genome", "source_binary_idx",  "comp_genome", "comp_binary_idx","ortho_count")
dat=setNames(data.frame(matrix(ncol = length(cols), nrow = 0)), cols)

for (genome in c('P. echinospica','R. piscesae','P. palmiformis')){
for (bin_idx in c('00','10','11','01')){
cat_ortho=df[df$genome==genome & df$binary_idx==bin_idx & !(is.na(df$orth_id)),]$orth_id %>% unique
cat_ortho %>% length

tmp=df[df$orth_id %in% cat_ortho,] %>% group_by(genome,binary_idx) %>% summarise(ortho_count=n()) %>% as.data.frame()
tmp
tmp$source_genome=genome
tmp$source_binary_idx=bin_idx
colnames(tmp)=c("comp_genome", "comp_binary_idx", "ortho_count", "source_genome", "source_binary_idx")
tmp=tmp[,c("source_genome", "source_binary_idx","comp_genome", "comp_binary_idx","ortho_count")]
tmp
dat=rbind(dat,tmp)
}
}

dat$source_binary_idx=factor(dat$source_binary_idx, levels=c('00','10','11','01'),labels=c('low UG','high U low G', 'high UG', 'low U high G'))
dat$comp_binary_idx=factor(dat$comp_binary_idx, levels=c('00','10','11','01'),labels=c('low UG','high U low G', 'high UG', 'low U high G'))
dat$source_genome=factor(dat$source_genome, levels=c('P. echinospica','R. piscesae','P. palmiformis'))
dat$comp_genome=factor(dat$comp_genome, levels=c('P. echinospica','R. piscesae','P. palmiformis'))


dat=dat[order(dat$source_genome, 
          dat$source_binary_idx,
          dat$comp_genome,
          dat$comp_binary_idx),]

dat$mean=dat %>% group_by(source_genome,source_binary_idx,comp_genome) %>% summarise(mean=ortho_count/sum(ortho_count)*100) %>% .$mean 

dat$mean=round(dat$mean)
dat$group=ifelse(dat$source_binary_idx %in% c('low U high G','high U low G'),0,1)

line_colors_dark=c('low UG'='#001c7f','high U low G'='#12711c','high UG'='#8c0800','low U high G'='#b1400d')
line_colors_bright=c('low UG'='#023eff','high U low G'='#1ac938','high UG'='#e8000b','low U high G'='#ff7c00')



dat=dat %>%
  mutate(y.label = paste("*",dat$source_genome,"*",
                         " <span style = 'color: ",lapply(dat$source_binary_idx,function(x){line_colors_dark[x][[1]]}) %>% unlist, ";'>",source_binary_idx,"</span>", 
                         " orthologs in ",
                        "*", dat$comp_genome,"*",
                         sep = ""))
library("ggtext")


p1<-ggplot(dat[dat$source_genome != dat$comp_genome,])+
  geom_bar(aes(y=y.label,x=ortho_count,fill=comp_binary_idx),stat='identity',position='stack',alpha=0.8)+
  facet_grid(source_genome~.,scales='free')+
  scale_fill_manual(values=line_colors_bright,name='Gene methylation profile group')+
  scale_x_continuous(expand=c(0,0),name='Counts')+
  theme_bw()+
  theme(strip.background = element_blank(),
        strip.text.y = element_blank(),
        axis.text.y.left = element_blank(),
        axis.title.y=element_blank())

p<-ggplot(dat[dat$source_genome != dat$comp_genome,])+
  geom_bar(aes(y=y.label,x=ortho_count,fill=comp_binary_idx),stat='identity',position='fill')+
  facet_grid(source_genome~.,scales='free')+
  scale_fill_manual(values=line_colors_bright,guide='none')+
  scale_x_continuous(expand=c(0,0),name='Relative abundance')+
  theme_bw()+
  theme(strip.background = element_blank(),
        strip.text.y = element_blank(),
        axis.text.y.left = element_markdown(hjust=0),
        axis.title.y= element_blank())

pg <- ggplot_build(p)

percent_map=cbind(pg$data[[1]][,c('fill','y','xmax','xmin')],dat[dat$source_genome != dat$comp_genome,])


p2<-ggplot(dat[dat$source_genome != dat$comp_genome,])+
  geom_bar(aes(y=y.label,x=ortho_count,fill=comp_binary_idx),stat='identity',position='fill',alpha=0.8,color='grey')+
  geom_text(data=percent_map[percent_map$mean>=5,],aes(x=(((xmax-xmin)/2)+xmin),y=y.label,label=paste0(mean,'%')),size=3)+
  facet_grid(source_genome~.,scales='free')+
  scale_fill_manual(values=line_colors_bright,guide='none')+
  # scale_color_manual(values=line_colors_dark,guide='none')+
  scale_x_continuous(expand=c(0,0),name='Relative abundance')+
  scale_y_discrete()+
  theme_bw()+
  theme(strip.background = element_blank(),
        strip.text.y = element_blank(),
        axis.text.y.left = element_markdown(hjust=0),
        axis.title.y= element_blank(),
        panel.grid = element_blank())

plot=egg::ggarrange(p2,p1, nrow=1, ncol=2,widths = c(2, 1))

plot 

ggsave('../figures/gene_methylation_profile_groups_vs_orthologs.png',plot,width=11.23, height=4.52, units = 'in', dpi=300)
ggsave('../figures/gene_methylation_profile_groups_vs_orthologs.pdf',plot,width=11.23, height=4.52, units = 'in')

```

### Partial correlations with gene expression

```{r get busco gene names, eval=FALSE, include=FALSE}
library('tidyverse')
library('jsonlite')

bg=read.csv('../data/Assemblies_and_annotations/buscogold.ALL.mrna.final_gene_models.proteins.txt',header=T,sep='\t')
colnames(bg)=c('busco_id','mRNA','genome')
bg %>% head

busco_ids <- unique(bg$busco_id)
names(busco_ids) <- busco_ids # so map_df gets .id right

# a tibble of interpro profile associated with each busco_id
busco_ipr <- map_df(busco_ids, .id="busco_id",  function(busco_id){
  write(busco_id, stderr()) # just so we can monitor progress
  # map the BUSCO ID to OrthoDB group ID
  query <- read_json(paste0("https://www.orthodb.org/search?query=", busco_id))
  odb_id <- query$data[1]
  # get all info on the orthogroup
  odb_info <- read_json(paste0("https://www.orthodb.org/group?id=", odb_id),
    simplifyVector = TRUE)
  # return the interpro table
  odb_info$data$interpro_domains
})

busco_ipr %>% select(busco_id, description)

bg=merge(bg,busco_ipr, by='busco_id')

write.table(bg,'../data/Assemblies_and_annotations/buscogold.ALL.mrna.final_gene_models.proteins.with_busco_ids.txt',col.names = T,sep = '\t',row.names = F)

```

```{r get data sliding window, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}
bg=read.csv('../data/Assemblies_and_annotations/buscogold.ALL.mrna.final_gene_models.proteins.with_busco_ids.txt',header=T,sep='\t')
bg$genome=factor(bg$genome, levels=c('Pec','R07B-5','P08H-3'), labels=c('P. echinospica','R. piscesae','P. palmiformis'))

species=c('R. piscesae','P. palmiformis','P. echinospica',"R07B-5", "P08H-3", "Pec")
names(species)=c("R07B-5", "P08H-3", "Pec",'R. piscesae','P. palmiformis','P. echinospica')

meta=read.csv('../data/Gene-body_methylation/ALL.downstream_genes_withCG_CDS.tsv',header=T,sep='\t',check.names = F)
meta=meta[,c('genome', "Name", "meth_meandepth", "meth_meancov", "meth_meanfreq", "TPM_value",'CpGs_count','gene_Length','cds_Length','CG%','cds_CG%','CpGoe')] %>% unique # Some genes are represented un multiple functional categories so duplicate rows need to be removed after clearing the functional columns
colnames(meta)=c('genome', "mRNA", "gene_mean_depth", "gene_mean_cov", "gene_mean_freq", "gene_TPM_value",'gene_CpG_count','gene_Length','cds_Length','gene_CG','cds_CG','gene_CpGoe')

ind='Pec'
cov=ifelse(ind=='P08H-3','cov1','cov10')

file=paste0('../data/Gene-body_methylation/',ind,'.downstream.',cov,'.TSS.2.tsv')
df=read.csv(file,header=T,sep='\t')
df=df[df$TSS_idx %in% c(-50,-40,-30,-25,-20,-10,-5),]
df$genome=species[ind]

for (ind in c('R07B-5','P08H-3')){
cov=ifelse(ind=='P08H-3','cov1','cov10')

file=paste0('../data/Gene-body_methylation/',ind,'.downstream.',cov,'.TSS.2.tsv')
tmp=read.csv(file,header=T,sep='\t')
tmp=tmp[tmp$TSS_idx %in% c(-50,-40,-30,-25,-20,-10,-5),]
tmp$genome=species[ind]

df=rbind(df,tmp)
}

colnames(meta)

df=merge(df,meta %>% unique, by=c('genome','mRNA'),all.x=T) 

methylated_flank=df[df[df$TSS_idx==-5,]$meth_cov>0.75 & df[df$TSS_idx==-5,]$meth_depth>0.5,c('mRNA','genome')] %>% unique
methylated_flank$flank_mean_binary=1
df=merge(df,methylated_flank,by=c('genome','mRNA'), all.x=T)
df[is.na(df$flank_mean_binary),'flank_mean_binary'] <- 0

df$gene_mean_binary=ifelse(df$gene_mean_cov>0.75 & df$gene_mean_depth>0.5,1,0)
df$mean_binary_idx=paste0(as.character(df$flank_mean_binary),as.character(df$gene_mean_binary))

df %>% head

datt=df %>% .[.$TSS_idx %in% c(-50,-40,-30,-25,-20,-10,-5),]
datt$cds_CG=as.numeric(datt$cds_CG)
datt$busco='all'
datt %>% nrow



datt_busco=merge(datt,unique(bg[,c('genome','mRNA')]),by=c('genome','mRNA'))
datt_busco$busco='busco'
datt_busco %>% head
datt_busco %>% nrow

datt=rbind(datt,datt_busco)
datt$mean_binary_idx=as.character(datt$mean_binary_idx)
write.table(datt,'../data/Gene-body_methylation/ALL.1Ksliding_flankmethylation_vs_TPM_vs_geneset.txt',col.names = T,row.names = F,quote=F,sep='\t')

rm(df,datt,datt_busco,meta,tmp,methylated_flank, bg)

```

```{r run pcor highly methylated genes, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE}
library(ppcor)

datt=read.csv('../data/Gene-body_methylation/ALL.1Ksliding_flankmethylation_vs_TPM_vs_geneset.txt',header=T,sep='\t')
datt %>% str

variables=list(c("cds_CG"),c('cds_Length'),c('gene_Length'),c("cds_CG",'cds_Length'),c("cds_CG",'gene_Length'))
# variables=list(c("cds_CG"),c('cds_Length'),c('gene_Length'),c("cds_CG",'cds_Length'),c("cds_CG",'gene_Length'),c('CpGoe'),c("cds_CG",'CpGoe'),c('CpGoe'),c('CpGoe'))

cols=c("estimate", "p.value", "statistic", "n", "gp", "Method",'ctrl','TSS_idx','busco')
pcor_df=setNames(data.frame(matrix(ncol = length(cols), nrow = 0)), cols)

minCpG=1

?pcor.test

minCpG=1
method='spearman'

for (genome in c(unique(datt$genome))){
for (busco in c('all','busco')){
for (TSS_idx in c(-50,-40,-30,-20,-10,-5)){
for (ctrl in variables){

high=pcor.test(datt[datt$genome==genome & datt$busco==busco & datt$gene_mean_binary %in% c(1) & datt$callCpGs_count>=minCpG & datt$TSS_idx==TSS_idx,] %>% .[complete.cases(.),'meth_freq'],
               datt[datt$genome==genome & datt$busco==busco & datt$gene_mean_binary %in% c(1) & datt$callCpGs_count>=minCpG & datt$TSS_idx==TSS_idx,] %>% .[complete.cases(.),'gene_TPM_value'],
               datt[datt$genome==genome & datt$busco==busco & datt$gene_mean_binary %in% c(1) & datt$callCpGs_count>=minCpG & datt$TSS_idx==TSS_idx,]  %>% .[complete.cases(.),ctrl],method=method) 
  
high$ctrl=paste(ctrl,collapse = " + ") # %>% print
high$TSS_idx=TSS_idx
high$busco=busco
high$genome=genome
high$method=method
high$minCpG=minCpG
pcor_df=rbind(pcor_df,high)

}}}}

pcor_df=pcor_df[with(pcor_df, order(minCpG,method,genome,busco,TSS_idx, ctrl)),c('minCpG','method','genome','busco','TSS_idx','n','ctrl','estimate','p.value')]

pcor_df$p.value_str=sprintf("%f",pcor_df$p.value)

write.table(pcor_df,paste0('../data/Gene-body_methylation/ALL.1Ksliding_flankmethylation_vs_TPM_vs_geneset_minCpG',minCpG,'.pcor'),row.names = F,col.names = T,quote=F,sep='\t')

```

```{r plot for figure sliding1Kflank-vs-expression, eval=FALSE, echo=FALSE, message=FALSE, warning=FALSE, include=FALSE}
#fig.align='center', fig.height=10, out.width = "8in"
datt=read.csv('../data/Gene-body_methylation/ALL.1Ksliding_flankmethylation_vs_TPM_vs_geneset.txt',header=T,sep='\t')

TSS.labs <- c('-10 Kbp','-8 Kbp','-6 Kbp','-5 Kbp','-4 Kbp','-2 Kbp','-1 Kbp')
names(TSS.labs) <- c(-50,-40,-30,-25,-20,-10,-5)

busco.labs <- c("all genes", "BUSCO genes")
names(busco.labs) <- c("all", "busco")

pcor_df=read.csv("../data/Gene-body_methylation/ALL.1Ksliding_flankmethylation_vs_TPM_vs_geneset_minCpG1.pcor",header=T,sep='\t')

pcor_df %>% head

pcor_df_summary=pcor_df %>% group_by(genome,busco,TSS_idx) %>% summarise(all_neg=all(estimate<0),
                                                                         all_signif=as.numeric(all(p.value<0.05)),
                                                                         all_neg_signif=as.numeric(all(estimate<0) & all(p.value<0.05)),
                                                                         all_pos_signif=as.numeric(all(estimate>0) & all(p.value<0.05)),
                                                                         all_signif_01=as.numeric(all(p.value<0.1)),
                                                                         all_neg_signif_01=as.numeric(all(estimate<0) & all(p.value<0.1)),
                                                                         all_pos_signif_01=as.numeric(all(estimate>0) & all(p.value<0.1)),
                                                                         all_signif_001=as.numeric(all(p.value<0.01)),
                                                                         all_neg_signif_001=as.numeric(all(estimate<0) & all(p.value<0.01)),
                                                                         all_pos_signif_001=as.numeric(all(estimate>0) & all(p.value<0.01))  )

pcor_df_summary[pcor_df_summary$busco=='busco' & pcor_df_summary$TSS_idx==-5,] %>% unique

pcor_df_summary[pcor_df_summary$all_neg_signif_01==1,'all_neg_signif_str']<-'*'
pcor_df_summary[pcor_df_summary$all_pos_signif_01==1,'all_pos_signif_str']<-'*'
pcor_df_summary[pcor_df_summary$all_neg_signif==1,'all_neg_signif_str']<-'**'
pcor_df_summary[pcor_df_summary$all_pos_signif==1,'all_pos_signif_str']<-'**'
pcor_df_summary[pcor_df_summary$all_neg_signif_001==1,'all_neg_signif_str']<-'***'
pcor_df_summary[pcor_df_summary$all_pos_signif_001==1,'all_pos_signif_str']<-'***'

text_df=pcor_df_summary[!is.na(pcor_df_summary$all_neg_signif_str),c('genome','busco','TSS_idx','all_neg_signif_str')] %>% as.data.frame


datt=merge(datt,text_df,by=c('genome','busco','TSS_idx'), all.x=T)
datt$genome=factor(datt$genome, levels=c("P. echinospica", "R. piscesae", "P. palmiformis"
))

datt %>% str

##########

datt$group=paste0('atop(italic("',datt$genome,'"),textstyle("',busco.labs[datt$busco],'"))')

datt$group=factor(datt$group, levels=c("atop(italic(\"P. echinospica\"),textstyle(\"all genes\"))", 
"atop(italic(\"P. echinospica\"),textstyle(\"BUSCO genes\"))", 
"atop(italic(\"R. piscesae\"),textstyle(\"all genes\"))",
"atop(italic(\"R. piscesae\"),textstyle(\"BUSCO genes\"))",
"atop(italic(\"P. palmiformis\"),textstyle(\"all genes\"))", 
"atop(italic(\"P. palmiformis\"),textstyle(\"BUSCO genes\"))"))

datt[datt$busco=='busco' & datt$genome=='P. palmiformis' & datt$TSS_idx==-5,c('genome','busco','TSS_idx','all_neg_signif_str')] %>% unique

plot<-ggplot(subset(datt,TSS_idx!= -25 & gene_mean_binary==1))+
  geom_hex(aes(x=gene_TPM_value,y=meth_freq,fill=..ndensity..))+
  geom_smooth(aes(x=gene_TPM_value,y=meth_freq),method='lm')+
  geom_text(aes(x=1000,y=0.9,label=all_neg_signif_str),color='red')+
  facet_grid(group~TSS_idx,labeller = labeller(TSS_idx = TSS.labs, group = label_parsed))+
  labs(x='Gene expression (TPM)',y='Upstream region methylation level')+
  # scale_x_continuous(limits=c(0,1000))+
  scale_x_continuous(trans=scales::pseudo_log_trans(base = 10),
                breaks=c(0,1,100,10000),
                minor_breaks = c(5,50,500,5000,10000),
                labels=scales::label_number_si())+
  # scale_x_continuous(trans=scales::pseudo_log_trans(base = 10))+
  scale_fill_gradient(low='transparent',high='black',trans = "log",name='density',breaks=c(0.005,0.05,0.5))+
  theme_bw()+
  theme(strip.text.y = element_text(angle=0),
        legend.position = 'top')
plot


ggsave("../figures/pcor_hex_plot_CpG1.png",plot, width = 7 , height=5, units = 'in', dpi = 300)
ggsave("../figures/pcor_hex_plot_CpG1.pdf",plot,width = 7 , height=5, units = 'in')

```

### Functional category enrichment

```{r load function to run Fisher tests2, eval=FALSE, message=FALSE, warning=FALSE, include=FALSE, warnings=FALSE}
# dat=d
# branch='All'

montecarlo_test_pval<-function(df,br,FUNC_level){
  ecdf_fun <- function(x,perc) ecdf(x)(perc) # determine the quantile of a value against a distribution
  test=br
  #generate random subsample with same number of genes as in test category
  df1=cbind(df[,1],df[,c('Core',test,FUNC_level)])
  colnames(df1)[1]='gene_id'
  len_core=df1[df1[,'Core']==1,]$gene_id %>% unique %>% length
  len_selected=df1[df1[,test]==1,]$gene_id %>% unique %>% length
  
  pval_distrib=c()
  n=0
  while (n<100) {
    # pick len_selected genes from core at random
    random_gene_set=sample((subset(df1,Core==1)$gene_id %>% unique), len_selected)
    # add random column to 
    dd=cbind(df1,'random'=as.numeric(df1$gene_id %in% random_gene_set))
    formula=as.formula(paste0(".~ ",FUNC_level))
    d<-aggregate(formula,dd,FUN = function(x) {sum(x > 0, na.rm = F)}, na.action = na.pass) 
    pval_distrib<-rbind(pval_distrib,c(dmvhyper(n=d[,c('Core')],x=d[,c('random')],k=sum(d[,c('random')]))))
    n=n+1}
  pval_test=dmvhyper(n=d[,c('Core')],x=d[,c(test)],k=sum(d[,c(test)]))
  
  final_pval<-ecdf_fun(sort(pval_distrib),pval_test)
  return(final_pval)} ## compare prop of distribution if test data to that of 100 trials of randomly sampled genes (same sample size as test data)

run_multihypergeometric_montecarlo_tests<- function(df,FUNC_level){
library("extraDistr")
  # FUNC_level='kog_des'
  branches_to_test=names(df)[which(sapply(df, is.numeric))]

  pval<-data.frame(p.value=as.numeric(numeric()))
  for(br in branches_to_test[-1]){
    p.value<-montecarlo_test_pval(df,br,FUNC_level)
    new_pval<-as.data.frame(p.value, row.names=br, col.names='p.value')
    pval<-rbind(pval,new_pval)
  }

  pval$comp<-row.names(pval)
  pval$symbol<-ifelse(pval$p.value<=0.01,'***',
                ifelse(pval$p.value<=0.05,'**',
                ifelse(pval$p.value<=0.1,'*','')))

return(pval)
}

run_fisher_tests<- function(dat){
fisher<-data.frame(matrix(ncol = length(colnames(dat)[-1]), nrow = length(row.names(dat))),row.names=rownames(dat))
colnames(fisher) <- colnames(dat)[-1]

differences<-data.frame(matrix(ncol = length(colnames(dat)[-1]), nrow = length(row.names(dat))),row.names=rownames(dat))
colnames(differences) <- colnames(dat)[-1]

br='01'
category="Amino acid transport and metabolism "
dat=dd

for(br in colnames(fisher)){
  data<-dat[,c(colnames(dat)[1],br)] %>% replace(., is.na(.), 0)
  for (category in rownames(data)){
    success_in_sample<-data[rownames(data)==category,br]
    success_in_left_part<-data[rownames(data)==category,colnames(dat)[1]]-success_in_sample
    failure_in_sample<-sum(data[,br])-success_in_sample
    failure_in_left_part<-sum(data[rownames(data)!=category,colnames(dat)[1]])+success_in_sample
    p.value<-fisher.test(matrix(c(success_in_sample, success_in_left_part, failure_in_sample, failure_in_left_part), 2, 2),alternative = "two.sided")
    
    # p.adjust(fisher[,br],
    #            method = "holm")

    fisher[category,br]=p.value
    
    differences[category,br]=(success_in_sample/sum(data))-(sum(data[rownames(data)==category,])/sum(data))*(sum(data[,br])/sum(data))
  }
  }

fisher[,br]=p.adjust(fisher[,br],method = "holm")
fisher$categories<-rownames(fisher)
differences$categories<-rownames(differences)
breakdown<-cbind(melt(fisher, id.vars='categories',value.name='pvalue'), melt(differences, id.vars=c('categories'))[,3])
colnames(breakdown)[2]='comp'
colnames(breakdown)[4]='diff'
breakdown$pvalue<-as.numeric(breakdown$pvalue)
breakdown$diff<-as.numeric(breakdown$diff)*100 # expresses the difference in probability of the event happening/normalised differences between expected and observed relative abundances of functional categories
breakdown$symbol<-ifelse(breakdown$pvalue<=0.01,'***',
              ifelse(breakdown$pvalue<=0.05,'**',
              ifelse(breakdown$pvalue<=0.1,'*','')))
breakdown$sign<-ifelse(breakdown$diff<0,'depletion' ,'enrichment')

return(breakdown)
} # runs fisher test for enrichment on each category

plot_breakdown_with_counts<-function(breakdown,dat){
# level<-rownames(dat)
dat$level<-rownames(dat)
add_counts<-melt(dat,variable.name = 'comp',value.name = 'freq')
breakdown_w_counts<-merge(breakdown,add_counts, by.y=c('comp','level'), by.x=c('comp','categories'),all=T)
breakdown_w_counts$freq[!(breakdown_w_counts$comp %in% c('Pan')) & breakdown_w_counts$pvalue>0.1] <- NA
tot<-aggregate(data=breakdown_w_counts,freq~comp,FUN='sum')
colnames(tot)[2]='tot'
breakdown_w_counts<-merge(breakdown_w_counts,tot,by='comp',all.x=T)
breakdown_w_counts$relfreq<-paste(round(breakdown_w_counts$freq/breakdown_w_counts$tot*100),'%')
breakdown_w_counts$relfreq[breakdown_w_counts$relfreq =='NA %'] <- NA
breakdown_w_counts$symbol[breakdown_w_counts$comp =='moderate'] <- NA
breakdown_w_counts$sign[breakdown_w_counts$comp =='moderate'] <- NA
breakdown_w_counts$diff[breakdown_w_counts$comp =='moderate'] <- NA

plot<-ggplot(data=subset(breakdown_w_counts,comp!='Pan'))+
    geom_point(aes(x=comp,y=categories,fill=sign,size=abs(diff),alpha=pvalue),shape=21,color='grey')+
    scale_x_discrete(limits = levels(dat$comp)) +
    scale_fill_manual(name="",breaks=c("depletion", "enrichment"), values =c( "lightblue", "#EC7063")) +
    scale_size_continuous(range=c(5,20),breaks=c(5,20,50,75,100))+
    scale_alpha_continuous(range=c(1,0.2), limits=c(0,0.1), na.value = 0,breaks=c(0,0.01,0.05,0.1,1),guide=FALSE)+
    geom_text(aes(x=comp,y=categories,label=symbol),vjust=-0.5)+
    geom_text(aes(x=comp,y=categories,label=freq),vjust=0.5)+
    geom_text(aes(x=comp,y=categories,label=relfreq),hjust=-1)+
    theme_bw() +
    theme(axis.title.x=element_blank(),
          axis.text.x = element_text(angle=60,hjust=1))
  return(plot)
}
```

```{r prep for promoter methylation fisher figure KOG, eval=FALSE, include=FALSE}
shortest_string <- function(s) {
  s=s[which(s!='---NA---')]
  s[which.min(nchar(s))]
}
presence <- function(x) {
  if(length(x) > 0) {1}
  else {0}}

df_all=read.csv('../data/Gene-body_methylation/ALL.downstream_genes_withCG_CDS_withflank.tsv',header=T,sep='\t')
df_all$genome=factor(df_all$genome,levels=c('P. echinospica','R. piscesae','P. palmiformis'))
df=df_all[df_all$binary_idx %in% c('00','11','01','10'), c('mRNA','genome','binary_idx')] %>% unique
df_all=df_all[df_all$binary_idx %in% c('00','10','11','01'),]

kog_hier=read.csv('../data/Assemblies_and_annotations/reference_KOG_hierarchy.txt',header=T,sep='\t',col.names = c('KOG_lvl1_id','KOG','KOG_lvl1'))
kog_hier %>% head
orth=read.csv('../data/Assemblies_and_annotations/Final_annotations.txt',header=T,sep='\t') %>% .[.$flag=='good_annot',c('genome','qseqid','KOG','Description')] %>% unique
colnames(orth)[which(colnames(orth)=='qseqid')]='mRNA'
orth$KOG=strsplit(orth$KOG, split=', ')
orth=orth %>% unnest(.,KOG)
map_of_good_descriptions=aggregate(orth$Description, by=list(orth$KOG), FUN = shortest_string) %>% as.data.frame %>% setNames(c("KOG",'Description'))
orth=merge(orth[,colnames(orth)[which(colnames(orth)!='Description')]],map_of_good_descriptions,by='KOG',all.x=T)
orth=merge(orth,kog_hier, by=c('KOG'), all.x=T)

df=merge(df,orth, by=c('genome','mRNA'), all.x=T)

df$KOG_lvl1 %>% unique %>% length

######### upstream methylation vs not ######

data=reshape2::dcast(df[df$binary_idx %in% c('01','11'),],genome+mRNA+KOG_lvl1~binary_idx,fun.aggregate = presence)
data$Core=1
data=data[,c('genome','mRNA','KOG_lvl1','Core','01','11')]

data %>% head

breakdown=data.frame()
for (genome in c('P. echinospica','R. piscesae', 'P. palmiformis')){
  
pval<-run_multihypergeometric_montecarlo_tests(data[data$genome==genome,-1],'KOG_lvl1')
pval

data[data$genome==genome,-1] %>% head

dd=data[data$genome==genome,-c(1,2)] %>%
  aggregate(. ~ KOG_lvl1,., sum)

rownames(dd)=dd$KOG_lvl1
dd[1]<-NULL
breakdown_tmp<-run_fisher_tests(dd)
breakdown_tmp$genome=genome

breakdown=rbind(breakdown,breakdown_tmp)
}

breakdown$genome=factor(breakdown$genome,levels=c('P. echinospica','R. piscesae', 'P. palmiformis'))
breakdown$comp=factor(breakdown$comp,levels=c('11','01'), labels=c("Methylated promoter","Unmethylated promoter"))

breakdown %>% str
breakdown$diff2=breakdown$diff
breakdown[breakdown$diff<0, 'diff2' ]=NA
breakdown[breakdown$comp=="Unmethylated promoter" & breakdown$diff>=0, 'diff2']=-1*breakdown[breakdown$comp=="Unmethylated promoter" & breakdown$diff>=0, 'diff']

cat_levels=breakdown[breakdown$genome=='P. echinospica', ] %>% .[order(.$diff2),c("categories")] %>% unique
breakdown$categories=factor(breakdown$categories, levels=cat_levels)

breakdown_KOG=breakdown[(breakdown$symbol %in% c('***','**')),]

plot_KOG<-ggplot(breakdown_KOG)+
  geom_bar(aes(x=diff2,y=categories,fill=(diff2>0)),color='grey',stat='identity')+
  labs(y='KOG functional categories', fill='',x='normalised differences between expected and observed\nrelative abundances of functional categories')+
  facet_grid(.~genome)+
  scale_fill_manual(values=c('#ff7c00','#e8000b'),labels=c('unmethylated promoters','methylated promoters'),name='Enriched amongst genes with:')+
  guides(fill=guide_legend(title.position="top", title.hjust = 0,nrow=1))+
  scale_x_continuous(labels = abs)+
  theme_bw()+
  theme(strip.text.x = element_text(face = 'bold.italic'),
        strip.background.x=element_blank(),
        strip.text.y=element_text(angle=0),
        legend.position='top')

plot_KOG
ggsave('../figures/promotermethylation_fisher_KOG.png',plot_KOG, width=11.8,height=7,,units = 'in', dpi=300)
ggsave('../figures/promotermethylation_fisher_KOG.pdf',plot_KOG, width=11.8,height=7,,units = 'in')


# ggplot(breakdown_KOG)+
#   geom_bar(aes(x=diff,y=categories,fill=sign),stat='identity')+
#   labs(y='KOG functional categories', fill='',x='normalised differences between expected and observed\nrelative abundances of functional categories')+
#   facet_grid(comp~genome)+
#   guides(fill=guide_legend(title.position="top", title.hjust = 0,nrow=1))+
#   scale_x_continuous(labels = abs)+
#   theme_bw()+
#   theme(strip.text.x = element_text(face = 'bold.italic'),
#         strip.background.x=element_blank(),
#         strip.text.y=element_text(angle=0),
#         legend.position='top')


```

```{r prep for promoter methylation fisher figure KEGG, eval=FALSE, include=FALSE}
shortest_string <- function(s) {
  s=s[which(s!='---NA---')]
  s[which.min(nchar(s))]
}
presence <- function(x) {
  if(length(x) > 0) {1}
  else {0}}

df_all=read.csv('../data/Gene-body_methylation/ALL.downstream_genes_withCG_CDS_withflank.tsv',header=T,sep='\t')
df_all$genome=factor(df_all$genome,levels=c('P. echinospica','R. piscesae','P. palmiformis'))
df=df_all[df_all$binary_idx %in% c('00','11','01','10'), c('mRNA','genome','binary_idx')] %>% unique
df_all=df_all[df_all$binary_idx %in% c('00','10','11','01'),]

brite_hier=read.csv('../data/Assemblies_and_annotations/reference_BRITE_hierarchy_clean.tsv',header=T,sep='\t')
brite_hier %>% str

orth=read.csv('../data/Assemblies_and_annotations/Final_annotations.txt',header=T,sep='\t') %>% .[.$flag=='good_annot',c('genome','qseqid','KO','Description')] %>% unique
colnames(orth)[which(colnames(orth)=='qseqid')]='mRNA'
orth$KO=strsplit(orth$KO, split=', ')
orth=orth %>% unnest(.,KO)
orth=merge(orth,brite_hier,by='KO',all.x=T)
orth %>% str

map_of_good_descriptions=aggregate(orth$Description, by=list(orth$KO), FUN = shortest_string) %>% as.data.frame %>% setNames(c("KO",'Description'))
orth=merge(orth[,colnames(orth)[which(colnames(orth)!='Description')]],map_of_good_descriptions,by='KO',all.x=T)

df=merge(df,orth, by=c('genome','mRNA'), all.x=T)

KEGG_lvl2=c("seq_id","start", "end", "CpGoe","CpGs_count", "mRNA", 
"callCpGs_count", "meth_meandepth", "meth_meancov", "meth_meanfreq", 
 "gene_Length",'cds_Length', "genome",'KEGG_lvl1','KEGG_lvl2')

bad_KEGG_lvl1=c("Brite Hierarchies","Human Diseases", "Not Included in Pathway or Brite") # remove irrelevant KEGG_lvl1 categories
bad_KEGG_lvl2=c("Biosynthesis of other secondary metabolites", "Cellular community - prokaryotes", 
"Information processing in viruses", "Membrane transport", "Metabolism of terpenoids and polyketides", 
"Not included in regular maps") #poorly represented KEGG_lvl2 categories (with less than 50 genes in each worm genomes)

df=df[!(df$KEGG_lvl1 %in% bad_KEGG_lvl1 | df$KEGG_lvl2 %in% bad_KEGG_lvl2), ]
df %>% head

## KEGG_lvl2

data=reshape2::dcast(df[df$binary_idx %in% c('01','11'),],genome+mRNA+KEGG_lvl2~binary_idx,fun.aggregate = presence)
data$Core=1
data=data[,c('genome','mRNA','KEGG_lvl2','Core','01','11')]

data %>% head

breakdown=data.frame()
for (genome in c('P. echinospica','R. piscesae', 'P. palmiformis')){
  
pval<-run_multihypergeometric_montecarlo_tests(data[data$genome==genome,-1],'KEGG_lvl2')
pval

data[data$genome==genome,-1] %>% head

dd=data[data$genome==genome,-c(1,2)] %>%
  aggregate(. ~ KEGG_lvl2,., sum)

rownames(dd)=dd$KEGG_lvl2
dd[1]<-NULL
breakdown_tmp<-run_fisher_tests(dd)
breakdown_tmp$genome=genome

breakdown=rbind(breakdown,breakdown_tmp)
}

breakdown$genome=factor(breakdown$genome,levels=c('P. echinospica','R. piscesae', 'P. palmiformis'))
breakdown$comp=factor(breakdown$comp,levels=c('11','01'), labels=c("Methylated promoter","Unmethylated promoter"))

breakdown %>% str
breakdown$diff2=breakdown$diff
breakdown[breakdown$diff<0, 'diff2' ]=NA
breakdown[breakdown$comp=="Unmethylated promoter" & breakdown$diff>=0, 'diff2']=-1*breakdown[breakdown$comp=="Unmethylated promoter" & breakdown$diff>=0, 'diff']

cat_levels=breakdown[breakdown$genome=='P. echinospica', ] %>% .[order(.$diff2),c("categories")] %>% unique

breakdown$categories=factor(breakdown$categories, levels=cat_levels)

breakdown_KEGG=breakdown[(breakdown$symbol %in% c('***','**')),]

plot_KEGG<-ggplot(breakdown_KEGG)+
  geom_bar(aes(x=diff2,y=categories,fill=(diff2>0)),color='grey',stat='identity')+
  labs(y='KEGG functional categories', fill='',x='normalised differences between expected and observed\nrelative abundances of functional categories')+
  facet_grid(.~genome)+
  scale_fill_manual(values=c('#ff7c00','#e8000b'),labels=c('unmethylated promoters','methylated promoters'),name='Enriched amongst genes with:')+
  guides(fill=guide_legend(title.position="top", title.hjust = 0,nrow=1))+
  scale_x_continuous(labels = abs)+
  theme_bw()+
  theme(strip.text.x = element_text(face = 'bold.italic'),
        strip.background.x=element_blank(),
        strip.text.y=element_text(angle=0),
        legend.position='top')

plot_KEGG

ggsave('../figures/promotermethylation_fisher_KEGG.png',plot_KEGG, width=11.8,height=7,units = 'in', dpi=300)
ggsave('../figures/promotermethylation_fisher_KEGG.pdf',plot_KEGG, width=11.8,height=7,,units = 'in')

```

## Orthologous genes methylation profiles

```{r get orthology data, eval=FALSE, include=FALSE}
#### get data ####
TPM=read.csv('../data/Gene-body_methylation/ALL.downstream_genes_withCG_CDS.tsv',header=T,sep='\t',check.names = F)
colnames(TPM)[which(colnames(TPM)=='Name')]='mRNA'

TPM=TPM[TPM$TPM_source!="R08H-5-V_TPM",c('TPM_value','mRNA','genome')] %>% unique

df_all=read.csv('../data/Gene-body_methylation/ALL.downstream_genes_withCG_CDS_withflank.tsv',header=T,sep='\t')

df_all=merge(df_all,TPM,by=c('mRNA','genome'), all=T)
df_all %>% group_by(genome) %>% summarise(count=n())

colnames(df_all)[which(colnames(df_all)=='gene_meth_binary')]='gene_binary'
df_all$genome=factor(df_all$genome,levels=c('P. echinospica','R. piscesae','P. palmiformis'))

df_all %>% str
orth=read.csv('../data/Assemblies_and_annotations/clean_worms_aa_diamond.poff.tsv',header=T,sep='\t',check.names = F)
colnames(orth)[4]='P08H'
colnames(orth)[5]='Pec'
colnames(orth)[6]='R07B'
orth$orth_id=rownames(orth)

orth %>% str

d1=orth %>% separate_rows(P08H,sep = ',') %>% .[,c('P08H','orth_id',"# Species")] %>% unique
colnames(d1)[1]='mRNA'
colnames(d1)[3]='species_count'
d1$genome='P. palmiformis'

d2=orth %>% separate_rows(R07B,sep = ',') %>% .[,c('R07B','orth_id',"# Species")] %>% unique
colnames(d2)[1]='mRNA'
colnames(d2)[3]='species_count'
d2$genome='R. piscesae'

d3=orth %>% separate_rows(Pec,sep = ',') %>% .[,c('Pec','orth_id',"# Species")] %>% unique
colnames(d3)[1]='mRNA'
colnames(d3)[3]='species_count'
d3$genome='P. echinospica'

d=rbind(d1,d2,d3)

d[d$species_count==3,] %>% nrow

d[d$species_count==3,] %>% group_by(orth_id) %>% summarise(count=n()) %>% .[.$count==3,] %>% nrow

d[d$species_count==3,] %>% group_by(orth_id) %>% summarise(count=n()) %>% nrow

d[d$species_count==3 ,] %>%
  group_by(genome, orth_id) %>%
  mutate(dupe = n()>1) %>% .[.$dupe==F,] %>% .$orth_id %>% unique %>% length # core orthologs represented by 1 gene model in each genome

df=merge(df_all,d,by=c('mRNA','genome'), all.x=T)
df %>% group_by(genome) %>% summarise(count=n())

rm(df_all)
df %>% colnames
df %>% head
df %>% str


flank_CG=read.csv('../data/ALL.downstream.1KbPromoter_CG.tsv',header=T,sep='\t')
flank_CG$genome=factor(flank_CG$genome, levels=unique(flank_CG$genome), labels=c('P. echinospica','R. piscesae','P. palmiformis'))

df=merge(df,flank_CG, by=c('genome','mRNA'), all.x=T)

df$gene_binary=as.character(as.numeric(df$gene_meth_depth>0.5 & df$gene_meth_cov>0.75))
df$flank_meth_binary=as.character(as.numeric(df$flank_meth_depth>0.5 & df$flank_meth_cov>0.75))
df$binary_idx=paste0(df$flank_meth_binary,df$gene_binary)

```

### Gene-body

```{r theorical exp, eval=FALSE, include=FALSE}
th=read.csv('../downstream_analyses_scripts/theorical_expression_pattern.txt',header=T,sep='\t',check.names = F)
colnames(th)=c("logfd", "species1", 
"species2", "gene_set")
th$gene_set=factor(th$gene_set,levels=c("methylated in species 1 only",  "control methylated", "control unmethylated",
"methylated in species 2 only"))

n=0
ggplot(th,aes(x=gene_set, ymin=species1+n, ymax=species2+n))+
  geom_linerange()+
  geom_point(aes(y=species1+n,color='species1'))+
  geom_point(aes(y=species2+n,color='species2'))+
  scale_y_continuous(name='Gene expression')+
  coord_flip()
  
ggplot(th,aes(x=gene_set, ymin=species1+n, ymax=species2+n))+
  geom_linerange()+
  geom_point(aes(y=species1+n,color='species1'))+
  geom_point(aes(y=species2+n,color='species2'))+
  # geom_bar(aes(y=log(fold_change),x=gene_set),stat='identity')+
  coord_flip()

```

```{r parse data, eval=FALSE, include=FALSE}

#### prepare data ####
bad_orth_dups=df[df$species_count==3 ,] %>%
  group_by(genome, orth_id) %>%
  mutate(dupe = n()>1) %>% .[.$dupe==T,] %>% .$orth_id %>% unique # orths represented by more than one gene model

bad_orth_na=df[is.na(df$gene_binary),'orth_id'] %>% unique # orths missing methylation info in at least one genome

good_core_orth=df[!(df$orth_id %in% c(bad_orth_dups,bad_orth_na)),c('genome','orth_id')] %>% unique %>%
  group_by(orth_id) %>%
  mutate(dupe = n()==3) %>% .[.$dupe==T,] %>% .$orth_id %>% unique # keep only core orth (i.e represented by a gene model in all genomes)

good_core_orth %>% unique %>% length 

Pec_1_ortho=(df[df$orth_id %in% good_core_orth &
                 df$genome %in% c('P. echinospica') & df$gene_binary==1,]$orth_id %>% unique)
Pec_0_ortho=(df[df$orth_id %in% good_core_orth &
                 df$genome %in% c('P. echinospica') & df$gene_binary==0,]$orth_id %>% unique)
R07_1_ortho=(df[df$orth_id %in% good_core_orth &
                 df$genome %in% c('R. piscesae') & df$gene_binary==1,]$orth_id %>% unique)
R07_0_ortho=(df[df$orth_id %in% good_core_orth & 
                 df$genome %in% c('R. piscesae') & df$gene_binary==0,]$orth_id %>% unique)
P08_1_ortho=(df[df$orth_id %in% good_core_orth & 
                 df$genome %in% c('P. palmiformis') & df$gene_binary==1,]$orth_id %>% unique)
P08_0_ortho=(df[df$orth_id %in% good_core_orth & 
                 df$genome %in% c('P. palmiformis') & df$gene_binary==0,]$orth_id %>% unique)


```

#### venn diagram

```{r venn diagram orthologs gene-body methylation, eval=FALSE, include=FALSE}
library(ggVennDiagram)
library(ggtext)

dat_0=list(Pec=Pec_0_ortho,
      R07B=R07_0_ortho,
      P08H=P08_0_ortho)

dat_1=list(Pec=Pec_1_ortho,
      R07B=R07_1_ortho,
      P08H=P08_1_ortho)

c(Pec_0_ortho,R07_0_ortho,P08_0_ortho) %>% unique %>% length
c(Pec_1_ortho,R07_1_ortho,P08_1_ortho) %>% unique %>% length

venn_1 <- Venn(dat_1)
data_1 <- process_data(venn_1)
venn_0 <- Venn(dat_0)
data_0 <- process_data(venn_0)

line_colors_bright=c('high gene-body methylation'='#e8000b','low gene-body methylation'='blue')
# line_colors_bright=c('high UG'='#8c0800','low U high G'='#ff7c00')

venn_plot_genebody=ggplot() +
  geom_sf(data = venn_region(data_1),fill=NA)+
  geom_sf_text(aes(label = c('P. echinospica','R. piscesae','P. palmiformis')), fontface = "bold.italic", data = venn_setlabel(data_1))+
  geom_sf_text(aes(label=count,color='high gene-body methylation'),fontface='bold',nudge_y = 0.5, data = venn_region(data_1)) +
  geom_sf_text(aes(label=count,color='low gene-body methylation'),fontface='bold',nudge_y = -0.5, data = venn_region(data_0)) +
  scale_color_manual(values=line_colors_bright,
                     labels = paste("<span style='color:",
                                   line_colors_bright,
                                   "'>",
                                   names(line_colors_bright),
                                   "</span>"),
                     name='',
                     guide=guide_legend(override.aes = list(color = c(NA,NA)) ))+
  coord_sf(xlim = c(-6,11))+
  theme_void()+
  theme(legend.position = c(1,1),
        legend.justification = c(1, 1),
        legend.text = element_markdown(face = 'bold'))

venn_plot_genebody

ggsave('../figures/Venn_orthology_genebody_methylation.png',venn_plot_genebody,width=8.49,height=5.18, units = 'in', dpi=300)

ggsave('../figures/Venn_orthology_genebody_methylation.pdf',venn_plot_genebody,width=8.49,height=5.18, units = 'in')

###########

methylated_venn=data_1@region$item %>% reshape2::melt(value.name = 'orth_id')
methylated_venn$L1=factor(methylated_venn$L1, levels=unique(methylated_venn$L1), labels=data_1@region$name)
methylated_venn$meth='methylated_gene-body'
unmethylated_venn=data_0@region$item %>% reshape2::melt(value.name = 'orth_id')
unmethylated_venn$L1=factor(unmethylated_venn$L1, levels=unique(unmethylated_venn$L1), labels=data_0@region$name)
unmethylated_venn$meth='unmethylated_gene-body'

methylated_venn %>% group_by(L1) %>% summarise(count=n())
unmethylated_venn %>% group_by(L1) %>% summarise(count=n())

venn_data=rbind(methylated_venn,unmethylated_venn)
venn_data %>% head

write.table(venn_data,'../data/Gene-body_methylation/venn_genebody_methylation_orthologs.txt',col.names = T,sep='\t',row.names = F)

```
#### expression
```{r summarise data per genome1, eval=FALSE, include=FALSE}
demet_TPM1=df[df$orth_id %in% c(
  intersect(Pec_1_ortho,P08_1_ortho),
  intersect(Pec_1_ortho,P08_0_ortho),
  intersect(Pec_0_ortho,P08_1_ortho),
  intersect(Pec_0_ortho,P08_0_ortho)
  ),] %>% unique
demet_TPM1$demet=NA
demet_TPM1$demet[demet_TPM1$orth_id %in% intersect(Pec_1_ortho,P08_1_ortho)]='control_met'
demet_TPM1$demet[demet_TPM1$orth_id %in% intersect(Pec_1_ortho,P08_0_ortho)]='demet'
demet_TPM1$demet[demet_TPM1$orth_id %in% intersect(Pec_0_ortho,P08_1_ortho)]='remet'
demet_TPM1$demet[demet_TPM1$orth_id %in% intersect(Pec_0_ortho,P08_0_ortho)]='control_unmet'
demet_TPM1$demet=factor(demet_TPM1$demet,levels=c('control_met','demet','remet','control_unmet'))
demet_TPM1=demet_TPM1[demet_TPM1$genome != 'R. piscesae',]
demet_TPM1 %>% group_by(genome,demet) %>% summarise(count=n()) # gene counts 
demet_TPM1 %>% group_by(genome) %>% summarise(count=n()) # gene counts 
demet_TPM1 %>% nrow

demet_TPM2=df[df$orth_id %in% c(
  intersect(Pec_1_ortho,R07_1_ortho),
  intersect(Pec_1_ortho,R07_0_ortho),
  intersect(Pec_0_ortho,R07_1_ortho),
  intersect(Pec_0_ortho,R07_0_ortho)),] %>% unique
demet_TPM2$demet=NA
demet_TPM2$demet[demet_TPM2$orth_id %in% intersect(Pec_1_ortho,R07_1_ortho)]='control_met'
demet_TPM2$demet[demet_TPM2$orth_id %in% intersect(Pec_1_ortho,R07_0_ortho)]='demet'
demet_TPM2$demet[demet_TPM2$orth_id %in% intersect(Pec_0_ortho,R07_1_ortho)]='remet'
demet_TPM2$demet[demet_TPM2$orth_id %in% intersect(Pec_0_ortho,R07_0_ortho)]='control_unmet'
demet_TPM2$demet=factor(demet_TPM2$demet,levels=c('control_met','demet','remet','control_unmet'))
demet_TPM2=demet_TPM2[demet_TPM2$genome != 'P. palmiformis',]
demet_TPM2 %>% group_by(genome,demet) %>% summarise(count=n()) # gene counts 
demet_TPM2 %>% group_by(genome) %>% summarise(count=n()) # gene counts 
demet_TPM2 %>% nrow

demet_TPM3=df[df$orth_id %in% c(
  intersect(R07_1_ortho,P08_1_ortho),
  intersect(R07_0_ortho,P08_1_ortho),
  intersect(R07_1_ortho,P08_0_ortho),
  intersect(R07_0_ortho,P08_0_ortho)),] %>% unique
demet_TPM3$demet=NA
demet_TPM3$demet[demet_TPM3$orth_id %in% intersect(R07_1_ortho,P08_1_ortho)]='control_met'
demet_TPM3$demet[demet_TPM3$orth_id %in% intersect(R07_0_ortho,P08_1_ortho)]='demet'
demet_TPM3$demet[demet_TPM3$orth_id %in% intersect(R07_1_ortho,P08_0_ortho)]='remet'
demet_TPM3$demet[demet_TPM3$orth_id %in% intersect(R07_0_ortho,P08_0_ortho)]='control_unmet'
demet_TPM3$demet=factor(demet_TPM3$demet,levels=c('control_met','demet','remet','control_unmet'))
demet_TPM3=demet_TPM3[demet_TPM3$genome != 'P. echinospica',]
demet_TPM3 %>% group_by(genome,demet) %>% summarise(count=n())
demet_TPM3 %>% group_by(genome) %>% summarise(count=n())



```
```{r observed exp1, eval=FALSE, include=FALSE}

make_boxplot_plot=function(test1th,sp1,sp2,res,method='KS'){
  library(ggtext)  # remotes::install_github("clauswilke/ggtext")

e <- function(x) {
  r <- quantile(x, probs = c(0.05, 0.25, NA, 0.75, 0.95))
  names(r) <- c("ymin", "lower", "middle", "upper", "ymax")
  r
}

f <- function(x) {
  r <- quantile(x, probs = c(NA, 0.4, 0.5, 0.6, NA))
  names(r) <- c("ymin", "lower", "middle", "upper", "ymax")
  r
}

alpha_scale=c(0.4,0.6)
names(alpha_scale)=c('50','20')
ngenes=test1th[!(is.na(test1th$logfd)),] %>% group_by(demet) %>% summarise(count=n()) %>% as.data.frame

fill_scale=c("#F8766D", "#7CAE00", "#00BFC4", "#C77CFF")
names(fill_scale)=levels(test1th$demet)

fill_scale_labels=c(paste0("hypermethylated in both species (n = ",ngenes[1,2],')'),
                    paste0("hypermethylated in *",sp1,"* only (n = ",ngenes[2,2],')'),
                    paste0('hypermethylated in *',sp2,"* only (n = ",ngenes[3,2],')'),
                    paste0("low-intermediate methylation in both species  (n = ",ngenes[4,2],')'))

res_title=paste0("Summary of pairwise ",method," tests: ",
       paste(c(unlist(rbind(paste("<span style = 'color: ",fill_scale[strsplit(res[res$method==method,'string'],split = " ")[[1]][c(1,3,5,7)]], ";'>",'&#11044;',"</span>"),strsplit(res[res$method==method,'string'],split = " ")[[1]][c(2,4,6,8)]))) %>% .[1:7],collapse=''))

plot=ggplot(test1th)+
    stat_summary(aes(y=logfd, x=demet,fill=demet,alpha='50'),
                 fun.data = e, geom="boxplot",
                 width=0.8,
                 position=position_dodge(1))+
    stat_summary(aes(y=logfd, x=demet,fill=demet,alpha='20'),
                 fun.data = f, geom="boxplot",
                 width=0.8,
                 position=position_dodge(1))+
    # geom_flat_violin(aes(y=logfd, x=demet))+
    scale_y_continuous(name=paste0('Log10 expression (*',sp2,'* / *',sp1,'*)'))+
  scale_x_discrete(limits = rev(levels(test1th$demet)))+
    labs(title=res_title)+
   scale_alpha_manual(values=alpha_scale,name='Interpercentile range')+
  scale_fill_manual(values=fill_scale,labels=fill_scale_labels,name='Methylation state of gene-body')+
    coord_flip(ylim = c(-2,2))+
    guides(alpha = guide_legend(override.aes = list(fill = 'grey20', alpha=c(0.2,0.6))))+
    theme_bw()+
    theme(axis.ticks.y=element_blank(),
      axis.title.y = element_blank(),
          axis.text.y = element_blank(),
          legend.text = element_markdown(),
          axis.title.x = element_markdown(),
      plot.title = element_markdown(size=rel(0.8)))

return(plot)
}
make_plot=function(test1th,sp1,sp2){
  demet_labels=c('control_unmet'='Unmethylated in both species','remet'='Methylated in species 2 only','demet'='Methylated in species 1 only','control_met'='Methylated in both species')
  ngenes=test1th %>% nrow
  t1 <- test1th %>% group_by(demet) %>% slice_sample(n=10)

  n=0.5
  plot1 <- ggplot(t1,aes(x=demet, ymin=species1, ymax=species2, group = row.names(t1)))+
    geom_linerange(position = position_dodge(n))+
    geom_point(aes(y=species1,color=paste0('species 1 : ', sp1)),position = position_dodge(n),alpha=1,size=2)+
    geom_point(aes(y=species2,color=paste0('species 2 : ', sp2)),position = position_dodge(n),alpha=1,size=2)+
    scale_y_continuous(trans=pseudo_log_trans(),name='Expression (TPM)')+
    scale_x_discrete(name='Gene-body methylation profiles\nof orthologous genes\nacross species',labels=demet_labels)+
    scale_color_discrete(name='')+
    labs(title='Expression of representative orthologous genes across species\n(random sample of 10 genes per category amongst expressed genes)')+
    coord_flip()+
    theme_bw()+
    theme(legend.position='bottom',
          plot.title = element_text(size=rel(0.8)))

source("https://raw.githubusercontent.com/datavizpyr/data/master/half_flat_violinplot.R")
  plot2<-ggplot(test1th)+
    geom_flat_violin(aes(y=logfd, x=demet))+
    scale_y_continuous(name='Log10 expression (species 2/species 1)')+
    labs(title=paste0('Distribution of expression fold change of orthologous genes across\nspecies (all orthologous genes expressed in both species, n = ',ngenes,')'))+
    coord_flip()+
    theme_bw()+
    theme(axis.title.y = element_blank(),
          axis.text.y = element_blank(),
          plot.title = element_text(size=rel(0.8)))

  plot=egg::ggarrange(plot1,plot2,nrow = 1,ncol = 2)

  return(plot)

  }
run_KS_test= function(fold_diff_test1){
cols=unique(fold_diff_test1$demet)
diff_test_higher=matrix(rep(1,length(cols)^2),ncol=length(cols),nrow=length(cols)) # is x>y? yes if pval <0.01
diff_test_lower=matrix(rep(1,length(cols)^2),ncol=length(cols),nrow=length(cols)) # is x<y? yes if pval <0.01

colnames(diff_test_higher)=cols
rownames(diff_test_higher)=cols
colnames(diff_test_lower)=cols
rownames(diff_test_lower)=cols

#### KS test ####
pairs=expand.grid(cols,cols)
for (i in c(1:nrow(pairs))){
l1=pairs[i,1]
l2=pairs[i,2]

set1=fold_diff_test1[fold_diff_test1$demet==l1,'logfd']
set2=fold_diff_test1[fold_diff_test1$demet==l2,'logfd']
val=ks.test(set1,set2,alternative='greater')
val=val$p.value

diff_test_lower[l1,l2]<-val
}
for (i in c(1:nrow(pairs))){
l1=pairs[i,1]
l2=pairs[i,2]

set1=fold_diff_test1[fold_diff_test1$demet==l1,'logfd']
set2=fold_diff_test1[fold_diff_test1$demet==l2,'logfd']
val=ks.test(set1,set2,alternative='less')
val=val$p.value

diff_test_higher[l2,l1]<-val
}

#### sumarise results  ####
thresh=0.05
diff_test_lower=ifelse(diff_test_lower<thresh,1,0)
diff_test_higher=ifelse(diff_test_higher<thresh,1,0)

# print(diff_test_higher*diff_test_lower)

str_list=rowSums(diff_test_higher*diff_test_lower)[rev(order(rowSums(diff_test_higher*diff_test_lower)))] %>% as.data.frame
colnames(str_list)[1]='count'
str_list$id=rownames(str_list)

str_list=str_list %>%
  group_by(count) %>%
  summarise(loc=min(count), id=paste(id, collapse=" = "))


return(t(c(comp=fold_diff_test1$comp[1],pair=fold_diff_test1$pair[1],string=paste0(rev(str_list$id),collapse =" < "))))
}
run_wilcox_test= function(fold_diff_test1){
cols=unique(fold_diff_test1$demet)
diff_test_higher=matrix(rep(1,length(cols)^2),ncol=length(cols),nrow=length(cols)) # is x>y? yes if pval <0.01
diff_test_lower=matrix(rep(1,length(cols)^2),ncol=length(cols),nrow=length(cols)) # is x<y? yes if pval <0.01

colnames(diff_test_higher)=cols
rownames(diff_test_higher)=cols
colnames(diff_test_lower)=cols
rownames(diff_test_lower)=cols

#### wilcox ####
pairs=expand.grid(cols,cols)
wilcox_res=data.frame()
for (i in c(1:nrow(pairs))){
l1=pairs[i,1]
l2=pairs[i,2]

set1=fold_diff_test1[fold_diff_test1$demet==l1,'logfd']
set2=fold_diff_test1[fold_diff_test1$demet==l2,'logfd']
val=wilcox.test(set1,set2,alternative='less')
val=val$p.value

diff_test_lower[l1,l2]<-val
}
for (i in c(1:nrow(pairs))){
l1=pairs[i,1]
l2=pairs[i,2]

set1=fold_diff_test1[fold_diff_test1$demet==l1,'logfd']
set2=fold_diff_test1[fold_diff_test1$demet==l2,'logfd']
val=wilcox.test(set1,set2,alternative='greater')
val=val$p.value

diff_test_higher[l2,l1]<-val
}
diff_test_lower=ifelse(diff_test_lower<0.01,1,0)
diff_test_higher=ifelse(diff_test_higher<0.01,1,0)

### summarise results ####
str_list=rowSums(diff_test_higher*diff_test_lower)[rev(order(rowSums(diff_test_higher*diff_test_lower)))] %>% as.data.frame
colnames(str_list)[1]='count'
str_list$id=rownames(str_list)

str_list=str_list %>%
  group_by(count) %>%
  summarise(loc=min(count), id=paste(id, collapse=" = "))

return(t(c(comp=fold_diff_test1$comp[1],pair=fold_diff_test1$pair[1],string=paste0(rev(str_list$id),collapse =" < ")))) 

}
run_test=function(test1,filter,sp1,sp2){
  run_KS_wilcox_tests=function(test1){
  ks_res=as.data.frame(run_KS_test(test1))
  wilcox_res=as.data.frame(run_wilcox_test(test1))
  ks_res$method='KS'
  wilcox_res$method='Wilcox'
  
  res=rbind(ks_res,wilcox_res)
  res$filter=filter
  return(res)
  }

  res=run_KS_wilcox_tests(test1)
  plot=make_boxplot_plot(test1,sp1,sp2,res)

  return(list('data'=test1,'plot'=plot,'res'=res))
}

pseudocounts=0.0000001
# pseudocounts=0
##############

sp1='P. echinospica'
sp2='P. palmiformis'
test1th=demet_TPM1[!(is.na(demet_TPM1$TPM_value)),] %>% dcast(.,demet+orth_id~genome,value.var='TPM_value', drop=T)
test1th$comp=paste0(sp2,'/',sp1)
test1th=test1th[,c('demet','orth_id',sp1,sp2,'comp')]
colnames(test1th)=c("demet", "orth_id", "species1", "species2",'comp')

test1th$logfd=log10((test1th$`species2`+pseudocounts)/(test1th$species1+pseudocounts))


test1th %>% tail(n=10)
test1th %>% str
test1th[!(is.na(test1th$logfd)),] %>% group_by(demet,.drop = T) %>% summarise(count=n())
test1th[(is.na(test1th$species1) & is.na(test1th$species2)),] %>% group_by(demet,.drop = T) %>% summarise(count=n())
test1th[,] %>% group_by(demet,.drop = T) %>% summarise(count=n())

test1th %>% head

filter='none'
t1=run_test(test1th,filter,sp1,sp2)
t1$res
t1$plot

# filter='TPM>0'
# test1=test1th[!(is.na(test1th$species1) & is.na(test1th$species2)) & test1th$species1>0 & test1th$species2>0,] %>% .[!(is.na(.$demet)),]
# t1=run_test(test1,filter,sp1,sp2)
# t1$res
# t1$plot


ggsave('../figures/PecPpalm_orthologs_genebody_methylation.png',t1$plot,width=10,height=3, units = 'in', dpi=300)



# sp1='P. echinospica'
# sp2='R. piscesae'
# test2th=demet_TPM2[!(is.na(demet_TPM2$TPM_value)),] %>% dcast(.,demet+orth_id~genome,value.var='TPM_value', drop=F)
# test2th$comp=paste0(sp2,'/',sp1)
# colnames(test2th)=c("demet", "orth_id", "species1", "species2",'comp')
# test2th$logfd=log10((test2th$`species2`+pseudocounts)/(test2th$species1+pseudocounts))
# test2th %>% summarise(count=n())
# 
# filter='none'
# t2=run_test(test2th,filter,sp1,sp2)
# t2$res
# t2$plot

# filter='TPM>0'
# test2=test2th[!(is.na(test2th$species1) & is.na(test2th$species2)) & test2th$species1>0 & test2th$species2>0,] %>% .[!(is.na(.$demet)),]
# t2=run_test(test2,filter,sp1,sp2)
# t2$res
# t2$plot


# sp1='P. palmiformis'
# sp2='R. piscesae'
# test3th=demet_TPM3[!(is.na(demet_TPM3$TPM_value)),] %>% dcast(.,demet+orth_id~genome,value.var='TPM_value', drop=F)
# test3th$comp=paste0(sp2,'/',sp1)
# colnames(test3th)=c("demet", "orth_id", "species1", "species2",'comp')
# test3th$logfd=log10((test3th$`species2`+pseudocounts)/(test3th$species1+pseudocounts))
# test3th %>% summarise(count=n())

# filter='none'
# t3=run_test(test3th,filter,sp1,sp2)
# t3$res
# t3$plot
# 
# filter='TPM>0'
# test3=test3th[!(is.na(test3th$species1) & is.na(test3th$species2)) & test3th$species1>0 & test3th$species2>0,] %>% .[!(is.na(.$demet)),]
# test3 %>% group_by(demet) %>% summarise(n())
# t3=run_test(test3,filter,sp1,sp2)
# t3$res


```

#### comparisons across taxonomic groups
```{r summarise data per taxon, eval=FALSE, include=FALSE}
tubeworms_1_ortho=intersect(Pec_1_ortho,R07_1_ortho)
tubeworms_0_ortho=intersect(Pec_0_ortho,R07_0_ortho)
end_1_ortho=intersect(R07_1_ortho,P08_1_ortho)
end_0_ortho=intersect(R07_0_ortho,P08_0_ortho)

intersect(Pec_0_ortho,Pec_1_ortho) #orthologs with both gene-body methylation profiles in Pec (when an ortholog is represented by more than one gene)
intersect(R07_0_ortho,R07_1_ortho)
intersect(P08_0_ortho,P08_1_ortho)

bad_orth=c(intersect(Pec_0_ortho,Pec_1_ortho),intersect(R07_0_ortho,R07_1_ortho),intersect(P08_0_ortho,P08_1_ortho)) %>% unique

genebod_demethylation_in_P08=intersect(tubeworms_1_ortho,P08_0_ortho) %>% setdiff(.,bad_orth)
genebod_nodemethylation_in_P08=intersect(tubeworms_1_ortho,P08_1_ortho) %>% setdiff(.,bad_orth)
genebod_remethylation_in_P08=intersect(tubeworms_0_ortho,P08_1_ortho) %>% setdiff(.,bad_orth)
genebod_noremethylation_in_P08=intersect(tubeworms_0_ortho,P08_0_ortho) %>% setdiff(.,bad_orth)

genebod_demethylation_in_R07=intersect(Pec_1_ortho,R07_0_ortho) %>% setdiff(.,bad_orth)
genebod_nodemethylation_in_R07=intersect(Pec_1_ortho,R07_1_ortho) %>% setdiff(.,bad_orth)
genebod_remethylation_in_R07=intersect(Pec_0_ortho,R07_1_ortho) %>% setdiff(.,bad_orth)
genebod_noremethylation_in_R07=intersect(Pec_0_ortho,R07_0_ortho) %>% setdiff(.,bad_orth)

genebod_demethylation_in_end=intersect(Pec_1_ortho,end_0_ortho) %>% setdiff(.,bad_orth)
genebod_nodemethylation_in_end=intersect(Pec_1_ortho,end_1_ortho) %>% setdiff(.,bad_orth)
genebod_remethylation_in_end=intersect(Pec_0_ortho,end_1_ortho) %>% setdiff(.,bad_orth)
genebod_noremethylation_in_end=intersect(Pec_0_ortho,end_0_ortho) %>% setdiff(.,bad_orth)

#### TPM1 ####
demet_TPM1=df[df$orth_id %in% c(genebod_nodemethylation_in_P08,genebod_demethylation_in_P08,genebod_remethylation_in_P08,genebod_noremethylation_in_P08),] %>% unique
demet_TPM1$demet=NA
demet_TPM1$demet[demet_TPM1$orth_id %in% genebod_nodemethylation_in_P08]='control_met'
demet_TPM1$demet[demet_TPM1$orth_id %in% genebod_demethylation_in_P08]='demet'
demet_TPM1$demet[demet_TPM1$orth_id %in% genebod_remethylation_in_P08]='remet'
demet_TPM1$demet[demet_TPM1$orth_id %in% genebod_noremethylation_in_P08]='control_unmet'
demet_TPM1$demet=factor(demet_TPM1$demet,levels=c('control_met','demet','remet','control_unmet'))
# demet_TPM1=demet_TPM1[demet_TPM1$genome != 'R. piscesae',]
demet_TPM1 %>% group_by(genome,demet) %>% summarise(count=n()) # gene counts 



#### TPM2 ####

demet_TPM2=df[df$orth_id %in% c(genebod_demethylation_in_R07, genebod_nodemethylation_in_R07,genebod_remethylation_in_R07,genebod_noremethylation_in_R07),] %>% unique
demet_TPM2$demet=NA
demet_TPM2$demet[demet_TPM2$orth_id %in% genebod_nodemethylation_in_R07]='control_met'
demet_TPM2$demet[demet_TPM2$orth_id %in% genebod_demethylation_in_R07]='demet'
demet_TPM2$demet[demet_TPM2$orth_id %in% genebod_remethylation_in_R07]='remet'
demet_TPM2$demet[demet_TPM2$orth_id %in% genebod_noremethylation_in_R07]='control_unmet'
demet_TPM2$demet=factor(demet_TPM2$demet,levels=c('control_met','demet','remet','control_unmet'))
demet_TPM2=demet_TPM2[demet_TPM2$genome != 'P. palmiformis',]
demet_TPM2 %>% group_by(genome,demet) %>% summarise(count=n()) 


#### TPM3 ####
demet_TPM3=df[df$orth_id %in% c(genebod_demethylation_in_end, genebod_nodemethylation_in_end,genebod_remethylation_in_end,genebod_noremethylation_in_end),] %>% unique
demet_TPM3$demet=NA
demet_TPM3$demet[demet_TPM3$orth_id %in% genebod_nodemethylation_in_end]='control_met'
demet_TPM3$demet[demet_TPM3$orth_id %in% genebod_demethylation_in_end]='demet'
demet_TPM3$demet[demet_TPM3$orth_id %in% genebod_remethylation_in_end]='remet'
demet_TPM3$demet[demet_TPM3$orth_id %in% genebod_noremethylation_in_end]='control_unmet'
demet_TPM3$demet=factor(demet_TPM3$demet,levels=c('control_met','demet','remet','control_unmet'))
# demet_TPM3=demet_TPM3[demet_TPM3$genome != 'R. piscesae',]


```

```{r param distrib figures1, eval=FALSE, include=FALSE}
e <- function(x) {
  r <- quantile(x, probs = c(0.0025, 0.25, 0.5, 0.75, 0.9975))
  names(r) <- c("ymin", "lower", "middle", "upper", "ymax")
  r
}

library('ggimage')

#### plot tpm1 ####

demet_TPM1=demet_TPM1[demet_TPM1$genome != 'R. piscesae',]
demet_TPM1 %>% group_by(genome,demet) %>% summarise(count=n()) # gene counts 

demet_TPM1[demet_TPM1$demet=='control_met','image']='../figures/c1.svg'
demet_TPM1[demet_TPM1$demet=='control_unmet','image']='../figures/c0.svg'
demet_TPM1[demet_TPM1$demet=='demet' & demet_TPM1$genome=='P. echinospica','image']='../figures/e1.svg'
demet_TPM1[demet_TPM1$demet=='demet' & demet_TPM1$genome!='P. echinospica','image']='../figures/p0.svg'
demet_TPM1[demet_TPM1$demet=='remet' & demet_TPM1$genome=='P. echinospica','image']='../figures/e0.svg'
demet_TPM1[demet_TPM1$demet=='remet' & demet_TPM1$genome!='P. echinospica','image']='../figures/p1.svg'

count_TPM1=demet_TPM1 %>% group_by(genome,demet) %>% summarise(count=n()) %>% as.data.frame
count_TPM1_control_met=count_TPM1[count_TPM1$demet=='control_met','count'] %>% unique %>% as.character
count_TPM1_control_unmet=count_TPM1[count_TPM1$demet=='control_unmet','count'] %>% unique %>% as.character
count_TPM1_demet=count_TPM1[count_TPM1$demet=='demet','count'] %>% unique %>% as.character
count_TPM1_remet=count_TPM1[count_TPM1$demet=='remet','count'] %>% unique %>% as.character

p11<-ggplot(demet_TPM1,fill=demet,color=demet)+
  geom_image(aes(y=10000,x=demet, image=image,fill=demet),size=0.2,by='width')+
  scale_size_identity()+
  stat_summary(aes(x=demet,y=TPM_value,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=TPM_value),shape=21,color='black',fill=NA,width=0.2,alpha=0.1)+
  scale_y_continuous(trans=scales::pseudo_log_trans(base = 10),
                     labels = label_number(accuracy = 1),
                     breaks=c(1,10,100,1000,10000),
                     minor_breaks=c(0,5,50,500,5000),
                     expand = c(0,0.6),
                     name='Gene expression\n(TPM)')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved gene-body methylation (n=',count_TPM1_control_met,')'),
    paste0('with unmethylated gene-body in *P. palmiformis* (n=',count_TPM1_demet,')'),
    paste0('with unmethylated gene-body in tubeworms (n=',count_TPM1_remet,')'),
    paste0('with conserved  lack of gene-body methylation (n=',count_TPM1_control_unmet,')')),
    name='Core gene orthologs:')+
guides(color=guide_legend(title.position="top", title.hjust = 0,nrow=2))+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_text(face = 'bold.italic'),
        strip.background = element_blank(),
        legend.position='top',
        legend.text = element_markdown())

p12<-ggplot(demet_TPM1)+
  stat_summary(aes(x=demet,y=cds_Length,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=cds_Length),shape=21,color='black',fill=NA,width=0.2,alpha=0.1)+
  scale_y_continuous(trans=scales::pseudo_log_trans(base = 10),
                     labels = label_number(accuracy = 1),
                     breaks=c(1,10,100,1000,10000),
                     minor_breaks=c(0,5,50,500,5000),
                     name='cds length (bp)')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved gene-body methylation (n=',count_TPM1_control_met,')'),
    paste0('with unmethylated gene-body in *P. palmiformis* only (n=',count_TPM1_demet,')'),
    paste0('with unmethylated gene-body in tubeworms only (n=',count_TPM1_remet,')'),
    paste0('with conserved  lack of gene-body methylation (n=',count_TPM1_control_unmet,')')),
    name='Core gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
          legend.position='none')

p13<-ggplot(demet_TPM1)+
  stat_summary(aes(x=demet,y=gene_CG*100,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=gene_CG*100),shape=21,color='black',fill=NA,width=0.2,alpha=0.1)+
  scale_y_continuous(name='gene CG%')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved gene-body methylation (n=',count_TPM1_control_met,')'),
    paste0('with unmethylated gene-body in *P. palmiformis* only (n=',count_TPM1_demet,')'),
    paste0('with unmethylated gene-body in tubeworms only (n=',count_TPM1_remet,')'),
    paste0('with conserved  lack of gene-body methylation (n=',count_TPM1_control_unmet,')')),
    name='Core gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
        legend.position='none')

demet_TPM1 %>% str
p14<-ggplot(demet_TPM1)+
  stat_summary(aes(x=demet,y=gene_CpGoe,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=gene_CpGoe),shape=21,color='black',fill=NA,width=0.2,alpha=0.1)+
  scale_y_continuous(name='gene-body\nCpG obs/exp')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved gene-body methylation (n=',count_TPM1_control_met,')'),
    paste0('with unmethylated gene-body in *P. palmiformis* only (n=',count_TPM1_demet,')'),
    paste0('with unmethylated gene-body in tubeworms only (n=',count_TPM1_remet,')'),
    paste0('with conserved  lack of gene-body methylation (n=',count_TPM1_control_unmet,')')),
    name='Core gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
        legend.position='none')

#### plot tpm2 ####

demet_TPM2[demet_TPM2$demet=='control_met','image']='../figures/c1.svg'
demet_TPM2[demet_TPM2$demet=='control_unmet','image']='../figures/c0.svg'
demet_TPM2[demet_TPM2$demet=='demet' & demet_TPM2$genome=='P. echinospica','image']='../figures/e1.svg'
demet_TPM2[demet_TPM2$demet=='demet' & demet_TPM2$genome!='P. echinospica','image']='../figures/p0.svg'
demet_TPM2[demet_TPM2$demet=='remet' & demet_TPM2$genome=='P. echinospica','image']='../figures/e0.svg'
demet_TPM2[demet_TPM2$demet=='remet' & demet_TPM2$genome!='P. echinospica','image']='../figures/p1.svg'

count_TPM2=demet_TPM2 %>% group_by(genome,demet) %>% summarise(count=n()) %>% as.data.frame
count_TPM2_control_met=count_TPM2[count_TPM2$demet=='control_met','count'] %>% unique %>% as.character
count_TPM2_control_unmet=count_TPM2[count_TPM2$demet=='control_unmet','count'] %>% unique %>% as.character
count_TPM2_demet=count_TPM2[count_TPM2$demet=='demet','count'] %>% unique %>% as.character
count_TPM2_remet=count_TPM2[count_TPM2$demet=='remet','count'] %>% unique %>% as.character

p21<-ggplot(demet_TPM2,fill=demet,color=demet)+
  geom_image(aes(y=10000,x=demet, image=image,fill=demet),size=0.2,by='width')+
  scale_size_identity()+
  stat_summary(aes(x=demet,y=TPM_value,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=TPM_value),shape=21,color='black',fill=NA,width=0.2,alpha=0.01)+
  scale_y_continuous(trans=scales::pseudo_log_trans(base = 10),
                     labels = label_number(accuracy = 1),
                     breaks=c(1,10,100,1000,10000),
                     minor_breaks=c(0,5,50,500,5000),
                     expand = c(0,0.6),
                     name='Gene expression\n(TPM)')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved gene-body methylation (n=',count_TPM2_control_met,')'),
    paste0('with unmethylated gene-body in *R. piscesae* (n=',count_TPM2_demet,')'),
    paste0('with unmethylated gene-body in *P. echinospica* (n=',count_TPM2_remet,')'),
    paste0('with conserved  lack of gene-body methylation (n=',count_TPM2_control_unmet,')')),
    name='Core gene orthologs:')+
guides(color=guide_legend(title.position="top", title.hjust = 0,nrow=2))+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_text(face = 'bold.italic'),
        strip.background = element_blank(),
        legend.position='top',
        legend.text=element_markdown())

p22<-ggplot(demet_TPM2)+
  stat_summary(aes(x=demet,y=cds_Length,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=cds_Length),shape=21,color='black',fill=NA,width=0.2,alpha=0.01)+
  scale_y_continuous(trans=scales::pseudo_log_trans(base = 10),
                     labels = label_number(accuracy = 1),
                     breaks=c(1,10,100,1000,10000),
                     minor_breaks=c(0,5,50,500,5000),
                     name='cds length (bp)')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved gene-body methylation (n=',count_TPM2_control_met,')'),
    paste0('with unmethylated gene-body in *R. piscesae* (n=',count_TPM2_demet,')'),
    paste0('with unmethylated gene-body in *P. echinospica* (n=',count_TPM2_remet,')'),
    paste0('with conserved  lack of gene-body methylation (n=',count_TPM2_control_unmet,')')),
    name='Core gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
          legend.position='none')

p23<-ggplot(demet_TPM2)+
  stat_summary(aes(x=demet,y=gene_CG*100,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=gene_CG*100),shape=21,color='black',fill=NA,width=0.2,alpha=0.01)+
  scale_y_continuous(name='gene CG%')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved gene-body methylation (n=',count_TPM2_control_met,')'),
    paste0('with unmethylated gene-body in *R. piscesae* (n=',count_TPM2_demet,')'),
    paste0('with unmethylated gene-body in *P. echinospica* (n=',count_TPM2_remet,')'),
    paste0('with conserved  lack of gene-body methylation (n=',count_TPM2_control_unmet,')')),
    name='Core gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
        legend.position='none')

p24<-ggplot(demet_TPM2)+
  stat_summary(aes(x=demet,y=gene_CpGoe,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=gene_CpGoe),shape=21,color='black',fill=NA,width=0.2,alpha=0.01)+
  scale_y_continuous(name='gene-body\nCpG obs/exp')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved gene-body methylation (n=',count_TPM2_control_met,')'),
    paste0('with unmethylated gene-body in *R. piscesae* (n=',count_TPM2_demet,')'),
    paste0('with unmethylated gene-body in *P. echinospica* (n=',count_TPM2_remet,')'),
    paste0('with conserved  lack of gene-body methylation (n=',count_TPM2_control_unmet,')')),
    name='Core gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
        legend.position='none')


#### plot tpm3 ####
demet_TPM3=demet_TPM3[demet_TPM3$genome != 'R. piscesae',]

demet_TPM3[demet_TPM3$demet=='control_met','image']='../figures/c11.svg'
demet_TPM3[demet_TPM3$demet=='control_unmet','image']='../figures/c01.svg'
demet_TPM3[demet_TPM3$demet=='demet' & demet_TPM3$genome=='P. echinospica','image']='../figures/e11.svg'
demet_TPM3[demet_TPM3$demet=='demet' & demet_TPM3$genome!='P. echinospica','image']='../figures/p01.svg'
demet_TPM3[demet_TPM3$demet=='remet' & demet_TPM3$genome=='P. echinospica','image']='../figures/e01.svg'
demet_TPM3[demet_TPM3$demet=='remet' & demet_TPM3$genome!='P. echinospica','image']='../figures/p11.svg'
count_TPM3=demet_TPM3 %>% group_by(genome,demet) %>% summarise(count=n()) %>% as.data.frame
count_TPM3_control_met=count_TPM3[count_TPM3$demet=='control_met','count'] %>% unique %>% as.character
count_TPM3_control_unmet=count_TPM3[count_TPM3$demet=='control_unmet','count'] %>% unique %>% as.character
count_TPM3_demet=count_TPM3[count_TPM3$demet=='demet','count'] %>% unique %>% as.character
count_TPM3_remet=count_TPM3[count_TPM3$demet=='remet','count'] %>% unique %>% as.character


p31<-ggplot(demet_TPM3,fill=demet,color=demet)+
  geom_image(aes(y=10000,x=demet, image=image,fill=demet),size=0.2,by='width')+
  scale_size_identity()+
  stat_summary(aes(x=demet,y=TPM_value,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=TPM_value),shape=21,color='black',fill=NA,width=0.2,alpha=0.01)+
  scale_y_continuous(trans=scales::pseudo_log_trans(base = 10),
                     labels = label_number(accuracy = 1),
                     breaks=c(1,10,100,1000,10000),
                     minor_breaks=c(0,5,50,500,5000),
                     expand = c(0,0.6),
                     name='Gene expression\n(TPM)')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved gene-body methylation (n=',count_TPM3_control_met,')'),
    paste0('with unmethylated gene-body in vents worms (n=',count_TPM3_demet,')'),
    paste0('with unmethylated gene-body in *P. echinospica* (n=',count_TPM3_remet,')'),
    paste0('with conserved  lack of gene-body methylation (n=',count_TPM3_control_unmet,')')),
    name='Core gene orthologs:')+
guides(color=guide_legend(title.position="top", title.hjust = 0,nrow=2))+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_text(face = 'bold.italic'),
        strip.background = element_blank(),
        legend.position='top',
        legend.text=element_markdown())

p32<-ggplot(demet_TPM3)+
  stat_summary(aes(x=demet,y=cds_Length,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=cds_Length),shape=21,color='black',fill=NA,width=0.2,alpha=0.01)+
  scale_y_continuous(trans=scales::pseudo_log_trans(base = 10),
                     labels = label_number(accuracy = 1),
                     breaks=c(1,10,100,1000,10000),
                     minor_breaks=c(0,5,50,500,5000),
                     name='cds length (bp)')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved gene-body methylation (n=',count_TPM3_control_met,')'),
    paste0('with unmethylated gene-body in vent worms (n=',count_TPM3_demet,')'),
    paste0('with unmethylated gene-body in *P. echinospica* (n=',count_TPM3_remet,')'),
    paste0('with conserved  lack of gene-body methylation (n=',count_TPM3_control_unmet,')')),
    name='Core gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
          legend.position='none')

p33<-ggplot(demet_TPM3)+
  stat_summary(aes(x=demet,y=gene_CG*100,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=gene_CG*100),shape=21,color='black',fill=NA,width=0.2,alpha=0.01)+
  scale_y_continuous(name='gene CG%')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved gene-body methylation (n=',count_TPM3_control_met,')'),
    paste0('with unmethylated gene-body in vent worms (n=',count_TPM3_demet,')'),
    paste0('with unmethylated gene-body in *P. echinospica* (n=',count_TPM3_remet,')'),
    paste0('with conserved  lack of gene-body methylation (n=',count_TPM3_control_unmet,')')),
    name='Core gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
        legend.position='none')

p34<-ggplot(demet_TPM3)+
  stat_summary(aes(x=demet,y=gene_CpGoe,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=gene_CpGoe),shape=21,color='black',fill=NA,width=0.2,alpha=0.01)+
  scale_y_continuous(name='gene-body\nCpG obs/exp')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved gene-body methylation (n=',count_TPM3_control_met,')'),
    paste0('with unmethylated gene-body in vent worms (n=',count_TPM3_demet,')'),
    paste0('with unmethylated gene-body in *P. echinospica*  (n=',count_TPM3_remet,')'),
    paste0('with conserved  lack of gene-body methylation (n=',count_TPM3_control_unmet,')')),
    name='Core gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
        legend.position='none')


#### PLOT ####
plot_demet1<-egg::ggarrange(p11,p12,p13,p14, nrow=4, ncol=1,heights = c(3, 2, 2, 2))
plot_demet2<-egg::ggarrange(p21,p22,p23,p24, nrow=4, ncol=1,heights = c(3, 2, 2, 2))
plot_demet3<-egg::ggarrange(p31,p32,p33,p34, nrow=4, ncol=1,heights = c(3, 2, 2, 2))


ggsave('../figures/Tbws_vs_Ppalm_orthology_gene-body_methylation2.png',plot_demet1, width=7.5,height=6, units = 'in',dpi=300)
ggsave('../figures/Tbws_vs_Ppalm_orthology_gene-body_methylation2.pdf',plot_demet1, width=7.5,height=6, units = 'in')


ggsave('../figures/Pec_vs_Rid_orthology_gene-body_methylation.png',plot_demet2, width=7.5,height=6, units = 'in',dpi=300)
ggsave('../figures/Pec_vs_Rid_orthology_gene-body_methylation.pdf',plot_demet2, width=7.5,height=6, units = 'in')

ggsave('../figures/Pec_vs_endeavour_orthology_gene-body_methylation.png',plot_demet3, width=7.5,height=6, units = 'in',dpi=300)
ggsave('../figures/Pec_vs_endeavour_orthology_gene-body_methylation.pdf',plot_demet3, width=7.5,height=6, units = 'in')

```

```{r genetic-epigenetic coupling1, eval=FALSE, include=FALSE}
demet_TPM1$conservation=ifelse(demet_TPM1$demet %in% c('control_met','control_unmet'),'conserved','unconserved')
demet_TPM1$comp='inter-fam'
demet_TPM2$conservation=ifelse(demet_TPM2$demet %in% c('control_met','control_unmet'),'conserved','unconserved')
demet_TPM2$comp='tubeworms'
demet_TPM3$conservation=ifelse(demet_TPM3$demet %in% c('control_met','control_unmet'),'conserved','unconserved')
demet_TPM3$comp='inter-habitat'


#### concatenate data frames ####
demet_TPM=rbind(demet_TPM1,demet_TPM2)

demet_TPM$group=paste(demet_TPM$comp,demet_TPM$conservation,sep='_')
demet_TPM=demet_TPM[!(is.na(demet_TPM$gene_binary)),]

# demet_TPM %>% group_by(comp,conservation) %>% summarise(count=n())
# demet_TPM %>% group_by(comp,conservation,genome,gene_binary) %>% summarise(count=n())


e <- function(x) {
  r <- quantile(x, probs = c(0.0025, 0.25, 0.5, 0.75, 0.9975))
  names(r) <- c("ymin", "lower", "middle", "upper", "ymax")
  r
}

demet_TPM$group=demet_TPM$group=factor(demet_TPM$group,
                     levels=c("tubeworms_unconserved","tubeworms_conserved",
                              "inter-fam_unconserved", "inter-fam_conserved"),
                     labels=c('Contrasting methylation\namongst tubeworms',
                             "Conserved methylation\namongst tubeworms",
                             'Contrasting methylation\nacross families',
                             "Conserved methylation\nacross families"))

inspect=demet_TPM[,c('group','comp','genome','orth_id','cds_Length')] %>% dcast(group+orth_id~genome,value.var = 'cds_Length')

inspect$rel=round(apply(inspect[,c(3:5)], 1,FUN=sd,na.rm=T)/inspect$`P. echinospica`*100)

# % standard deviation of cds_length of orthologs between tubeworms and Pec normalised to Pec cds length
ggplot(inspect,aes(x=rel))+
  stat_ecdf(geom = "step")+
  scale_x_continuous(trans=pseudo_log_trans(),limits=c(0,100))+
  facet_wrap(~group)  

inspect %>% group_by(group) %>% summarise(quant=quantile(rel,probs = c(0.25, 0.5, 0.75, 0.8,0.9,0.95,0.99,1), na.rm=T)) %>% as.data.frame()

inspect %>% group_by(group) %>% summarise(quant=quantile(rel,probs = c(0.85), na.rm=T)) %>% as.data.frame() #85% of othologs have 40% or less product length difference

inspect %>% group_by(group) %>% summarise(quant=quantile(rel,probs = c(0.75), na.rm=T)) %>% as.data.frame() #75% of othologs have 27% or less product length difference

df_n=demet_TPM %>% group_by(comp,conservation,gene_binary) %>% summarise(count=n()) %>% as.data.frame

textsize=ggplot2::.pt

p0<-ggplot(demet_TPM)+
  stat_summary(aes(x=as.factor(gene_binary),y=gene_CpGoe, color=group),
               fun.data = e, geom="boxplot",
               fatten = 1)+
  geom_jitter(aes(x=as.factor(gene_binary),y=gene_CpGoe),
              shape=21,color='black',fill=NA,width=0.2,alpha=0.02)+
  geom_text(data=df_n,aes(x=as.factor(gene_binary),y=2,label=paste0('n = ',as.character(count))),size=textsize,vjust=0,color='grey20')+
  scale_x_discrete(labels=c("Low","High"), name='Gene-body methylation level')+
  scale_y_continuous(name='Gene-body CpG observed/expected')+
  facet_grid(comp~conservation,
             labeller = labeller(conservation=setNames(c("Contrasting\ngene-body methylation",
                                                         "Conserved\ngene-body methylation"),
                                                       c('unconserved', 'conserved')),
                                 comp=setNames(c("Across families","Amongst tubeworms","Across habitats"),
                                               c("inter-fam", 'tubeworms','inter-habitat'))))+
  coord_flip()+
  theme_bw()+
  theme(legend.position='none',
        axis.title.x=element_text(size=rel(0.8)),
        axis.title.y=element_text(size=rel(0.8)))

p0

#### 1000 bootstraps ####

set.seed(2022)    # for reproduciblility
iter=1000
tot_res=data.frame()
demet_TPM$group=paste(demet_TPM$comp,demet_TPM$conservation,sep='_')

demet_TPM$group=demet_TPM$group=factor(demet_TPM$group,
                     levels=c("tubeworms_unconserved","tubeworms_conserved",
                              "inter-fam_unconserved", "inter-fam_conserved"),
                     labels=c('Contrasting methylation\namongst tubeworms',
                             "Conserved methylation\namongst tubeworms",
                             'Contrasting methylation\nacross families',
                             "Conserved methylation\nacross families"))

sample_count=demet_TPM %>% group_by(genome,group) %>% summarise(count = n()) # get number of samples
orth_id_tbl=demet_TPM[demet_TPM$genome=='P. echinospica',c('comp','conservation','group','orth_id')] # get orth_id table
library('purrr')
for (i in (1:iter)){
  if(i%%100==0){print(i)}
  
  tmp=demet_TPM
  
  orth_id_map=orth_id_tbl %>% 
  nest(-group) %>% 
  left_join(sample_count, by = "group") %>%
  mutate(Sample = map2(data, count, replace = T, sample_n)) %>%
  unnest(Sample,names_repair='universal') %>% .[,c('orth_id','comp','conservation')] 

  tmp=merge(tmp[,colnames(tmp)[colnames(tmp)!='conservation']],orth_id_map,by=c('orth_id','comp'),all.y=T)
  
  
  res=tmp %>% nest(-group) %>% 
  mutate(fit = purrr::map(data, ~ glm(as.numeric(gene_binary) ~ gene_CpGoe, data = .,family = binomial)),
         results = map(fit, glance)) %>% 
  unnest(results) %>% 
  summarise(group=group,r.squared=1 - deviance/null.deviance) 
  
  res$i=i
  tot_res=rbind(tot_res,res)
}


write.table(tot_res,'../data/Gene-body_methylation/Bootstrap_results_gene-body_coupling.txt',col.names=T,row.names = F,sep='\t',quote = T)

error_bar=tot_res %>% group_by(group) %>% summarise(sd=sd(r.squared),mean=mean(r.squared),sdmin=mean(r.squared)-sd(r.squared),sdmax=mean(r.squared)+sd(r.squared))

res_glm=demet_TPM %>% nest(-group) %>% 
  mutate(fit = map(data, ~ glm(as.numeric(gene_binary) ~ gene_CpGoe, data = .,family = binomial)),
         results = map(fit, glance)) %>% 
  unnest(results) %>% 
  summarise(group=group,r.squared=1 - deviance/null.deviance) %>% as.data.frame




res_glm=merge(res_glm,error_bar ,by='group',all.x=T)


p1<-ggplot(res_glm,aes(y=group,xmin=sdmin,xmax=sdmax))+
  geom_bar(aes(fill=group,x=mean),stat='identity')+
  # geom_bar(aes(fill=group,,x=r.squared),stat='identity')+
  geom_errorbar(width = 0.5)+
  labs(x='Strength of correlation between\ngene-body methylation and its genetic signature')+
  theme_bw()+
  theme(axis.title.y=element_blank(),
        legend.position='none',
        axis.title.x=element_text(size=rel(0.8)))
p1

#### inspect bootstraps ####

tot_res=read.csv('../data/Gene-body_methylation/Bootstrap_results_gene-body_coupling.txt',header=T,sep='\t',allowEscapes = F)


tot_res %>% summary
tot_res %>% str

ggplot(tot_res)+
  geom_histogram(aes(x=r.squared))+
  facet_grid(.~group, scales='free')

ggplot(tot_res)+
  geom_boxplot(aes(x=r.squared,y=group))+
  theme_bw()

run_KS_test= function(tot_res){
cols=unique(tot_res$group)
diff_test_higher=matrix(rep(1,length(cols)^2),ncol=length(cols),nrow=length(cols)) # is x>y? yes if pval <0.01
diff_test_lower=matrix(rep(1,length(cols)^2),ncol=length(cols),nrow=length(cols)) # is x<y? yes if pval <0.01

colnames(diff_test_higher)=cols
rownames(diff_test_higher)=cols
colnames(diff_test_lower)=cols
rownames(diff_test_lower)=cols

#### KS test ####
pairs=expand.grid(cols,cols)
for (i in c(1:nrow(pairs))){
l1=pairs[i,1]
l2=pairs[i,2]

set1=tot_res[tot_res$group==l1,'r.squared'] %>% .[!(is.na(.))]
set2=tot_res[tot_res$group==l2,'r.squared'] %>% .[!(is.na(.))]
val=ks.test(set1,set2,alternative='greater')
val=val$p.value

diff_test_lower[l1,l2]<-val
}
for (i in c(1:nrow(pairs))){
l1=pairs[i,1]
l2=pairs[i,2]

set1=tot_res[tot_res$group==l1,'r.squared'] %>% .[!(is.na(.))]
set2=tot_res[tot_res$group==l2,'r.squared'] %>% .[!(is.na(.))]
val=ks.test(set1,set2,alternative='less')
val=val$p.value

diff_test_higher[l2,l1]<-val
}

#### sumarise results  ####
thresh=0.01
diff_test_lower=ifelse(diff_test_lower<thresh,1,0)
diff_test_higher=ifelse(diff_test_higher<thresh,1,0)

# print(diff_test_higher*diff_test_lower)

str_list=rowSums(diff_test_higher*diff_test_lower)[rev(order(rowSums(diff_test_higher*diff_test_lower)))] %>% as.data.frame
colnames(str_list)[1]='count'
str_list$id=rownames(str_list)

str_list=str_list %>%
  group_by(count) %>%
  summarise(loc=min(count), id=paste(id, collapse=" = "))


return(t(c(comp=tot_res$comp[1],pair=tot_res$pair[1],string=paste0(rev(str_list$id),collapse =" > "))))
}

pairwise_KS_test_results=run_KS_test(tot_res)


#### mount plot ####

plot=egg::ggarrange(p0,p1,ncol=2)

plot

ggsave('../figures/gene-bodyCpGoe_vs_gene-bodymeth.png',plot,width=9,height=3.5, units = 'in',dpi=300)
ggsave('../figures/gene-bodyCpGoe_vs_gene-bodymeth.pdf',plot,width=9,height=3.5, units = 'in')
```

#### KOG/KEGG fisher tests on core orthologs

```{r function to run Fisher tests3, eval=FALSE, include=FALSE}
# dat=d
# branch='All'

montecarlo_test_pval<-function(df,br,FUNC_level){
  ecdf_fun <- function(x,perc) ecdf(x)(perc) # determine the quantile of a value against a distribution
  test=br
  #generate random subsample with same number of genes as in test category
  df1=cbind(df[,1],df[,c('Core',test,FUNC_level)])
  colnames(df1)[1]='gene_id'
  len_core=df1[df1[,'Core']==1,]$gene_id %>% unique %>% length
  len_selected=df1[df1[,test]==1,]$gene_id %>% unique %>% length
  
  pval_distrib=c()
  n=0
  while (n<100) {
    # pick len_selected genes from core at random
    random_gene_set=sample((subset(df1,Core==1)$gene_id %>% unique), len_selected)
    # add random column to 
    dd=cbind(df1,'random'=as.numeric(df1$gene_id %in% random_gene_set))
    formula=as.formula(paste0(".~ ",FUNC_level))
    d<-aggregate(formula,dd,FUN = function(x) {sum(x > 0, na.rm = F)}, na.action = na.pass) 
    pval_distrib<-rbind(pval_distrib,c(dmvhyper(n=d[,c('Core')],x=d[,c('random')],k=sum(d[,c('random')]))))
    n=n+1}
  pval_test=dmvhyper(n=d[,c('Core')],x=d[,c(test)],k=sum(d[,c(test)]))
  
  final_pval<-ecdf_fun(sort(pval_distrib),pval_test)
  return(final_pval)} ## compare prop of distribution if test data to that of 100 trials of randomly sampled genes (same sample size as test data)

run_multihypergeometric_montecarlo_tests<- function(df,FUNC_level){
library("extraDistr")
  branches_to_test=names(df)[which(sapply(df, is.numeric))]

  pval<-data.frame(p.value=as.numeric(numeric()))
  for(br in branches_to_test[-1]){
    p.value<-montecarlo_test_pval(df,br,FUNC_level)
    new_pval<-as.data.frame(p.value, row.names=br, col.names='p.value')
    pval<-rbind(pval,new_pval)
  }

  pval$comp<-row.names(pval)
  pval$symbol<-ifelse(pval$p.value<=0.01,'***',
                ifelse(pval$p.value<=0.05,'**',
                ifelse(pval$p.value<=0.1,'*','')))

return(pval)
}


run_fisher_tests<- function(dat){
fisher<-data.frame(matrix(ncol = length(colnames(dat)[-1]), nrow = length(row.names(dat))),row.names=rownames(dat))
colnames(fisher) <- colnames(dat)[-1]

differences<-data.frame(matrix(ncol = length(colnames(dat)[-1]), nrow = length(row.names(dat))),row.names=rownames(dat))
colnames(differences) <- colnames(dat)[-1]

# br='meth_binary'
# category="Amino acid transport and metabolism "

for(br in colnames(fisher)){
  data<-dat[,c(colnames(dat)[1],br)] %>% replace(., is.na(.), 0)
  for (category in rownames(data)){
    success_in_sample<-data[rownames(data)==category,br]
    success_in_left_part<-data[rownames(data)==category,colnames(dat)[1]]-success_in_sample
    failure_in_sample<-sum(data[,br])-success_in_sample
    failure_in_left_part<-sum(data[rownames(data)!=category,colnames(dat)[1]])+success_in_sample
    p.value<-fisher.test(matrix(c(success_in_sample, success_in_left_part, failure_in_sample, failure_in_left_part), 2, 2),alternative = "two.sided")
    
    fisher[category,br]=p.value
    
    differences[category,br]=(data[rownames(data)==category,br]/sum(data))-(sum(data[rownames(data)==category,])/sum(data))*(sum(data[,br])/sum(data))
  }
  }

fisher[,br]=p.adjust(fisher[,br],method = "holm") # Holms correction for mltiple testing
fisher$categories<-rownames(fisher)
differences$categories<-rownames(differences)
breakdown<-cbind(melt(fisher, id.vars='categories',value.name='pvalue'), melt(differences, id.vars=c('categories'))[,3])
colnames(breakdown)[2]='comp'
colnames(breakdown)[4]='diff'
breakdown$pvalue<-as.numeric(breakdown$pvalue)
breakdown$diff<-as.numeric(breakdown$diff)*100 # expresses the difference in probability of the event happening
breakdown$symbol<-ifelse(breakdown$pvalue<=0.01,'***',
              ifelse(breakdown$pvalue<=0.05,'**',
              ifelse(breakdown$pvalue<=0.1,'*','')))
breakdown$sign<-ifelse(breakdown$diff<0,'depletion' ,'enrichment')

return(breakdown)
} # runs fisher test for enrichment on each category

plot_breakdown_with_counts<-function(breakdown,dat){
# level<-rownames(dat)
dat$level<-rownames(dat)
add_counts<-melt(dat,variable.name = 'comp',value.name = 'freq')
breakdown_w_counts<-merge(breakdown,add_counts, by.y=c('comp','level'), by.x=c('comp','categories'),all=T)
breakdown_w_counts$freq[!(breakdown_w_counts$comp %in% c('Pan')) & breakdown_w_counts$pvalue>0.1] <- NA
tot<-aggregate(data=breakdown_w_counts,freq~comp,FUN='sum')
colnames(tot)[2]='tot'
breakdown_w_counts<-merge(breakdown_w_counts,tot,by='comp',all.x=T)
breakdown_w_counts$relfreq<-paste(round(breakdown_w_counts$freq/breakdown_w_counts$tot*100),'%')
breakdown_w_counts$relfreq[breakdown_w_counts$relfreq =='NA %'] <- NA
breakdown_w_counts$symbol[breakdown_w_counts$comp =='moderate'] <- NA
breakdown_w_counts$sign[breakdown_w_counts$comp =='moderate'] <- NA
breakdown_w_counts$diff[breakdown_w_counts$comp =='moderate'] <- NA

plot<-ggplot(data=subset(breakdown_w_counts,comp!='Pan'))+
    geom_point(aes(x=comp,y=categories,fill=sign,size=abs(diff),alpha=pvalue),shape=21,color='grey')+
    scale_x_discrete(limits = levels(dat$comp)) +
    scale_fill_manual(name="",breaks=c("depletion", "enrichment"), values =c( "lightblue", "#EC7063")) +
    scale_size_continuous(range=c(5,20),breaks=c(5,20,50,75,100))+
    scale_alpha_continuous(range=c(1,0.2), limits=c(0,0.1), na.value = 0,breaks=c(0,0.01,0.05,0.1,1),guide=FALSE)+
    geom_text(aes(x=comp,y=categories,label=symbol),vjust=-0.5)+
    geom_text(aes(x=comp,y=categories,label=freq),vjust=0.5)+
    geom_text(aes(x=comp,y=categories,label=relfreq),hjust=-1)+
    theme_bw() +
    theme(axis.title.x=element_blank(),
          axis.text.x = element_text(angle=60,hjust=1))
  return(plot)
}
```

```{r KOG functional enrichment of orthologs, eval=FALSE, include=FALSE}

#### prepare data frame ####

shortest_string <- function(s) {
  s=s[which(s!='---NA---')]
  s[which.min(nchar(s))]
}
presence <- function(x) {
  if(length(x) > 0) {1}
  else {0}}

kog_hier=read.csv('../data/Assemblies_and_annotations/reference_KOG_hierarchy.txt',header=T,sep='\t',col.names = c('KOG_lvl1_id','KOG','KOG_lvl1'))
kog_hier %>% head
orth=read.csv('../data/Assemblies_and_annotations/Final_annotations.txt',header=T,sep='\t') %>% .[.$flag=='good_annot',c('genome','qseqid','KOG','Description')] %>% unique
colnames(orth)[which(colnames(orth)=='qseqid')]='mRNA'
orth$KOG=strsplit(orth$KOG, split=', ')
orth=orth %>% unnest(.,KOG)
map_of_good_descriptions=aggregate(orth$Description, by=list(orth$KOG), FUN = shortest_string) %>% as.data.frame %>% setNames(c("KOG",'Description'))
orth=merge(orth[,colnames(orth)[which(colnames(orth)!='Description')]],map_of_good_descriptions,by='KOG',all.x=T)
orth=merge(orth,kog_hier, by=c('KOG'), all.x=T)
orth %>% head

orth=merge(orth,df[,c('genome','mRNA','orth_id')] %>% unique, by=c('genome','mRNA'), all.y=T) %>% .[,c('orth_id','KOG','Description','KOG_lvl1')] %>% unique

tmp=read.csv('../data/Gene-body_methylation/venn_genebody_methylation_orthologs.txt',header=T,sep='\t')
tmp=merge(tmp,orth,by=c('orth_id')) %>% unique
tmp$L1=factor(tmp$L1, levels=c("Pec","R07B", "P08H", "Pec..R07B", "R07B..P08H", 
 "Pec..P08H", "Pec..R07B..P08H" ))
tmp$meth=factor(tmp$meth, levels=unique(tmp$meth))
tmp %>% str

tmp$orth_id %>% unique %>% length

data1=reshape2::dcast(tmp[tmp$meth=="methylated_gene-body",],orth_id+KOG_lvl1~L1,fun.aggregate = presence,drop = T)

data2=reshape2::dcast(tmp[tmp$meth=="unmethylated_gene-body",],orth_id+KOG_lvl1~L1,fun.aggregate = presence,drop = T)

data=merge(data1,data2,by=c('orth_id','KOG_lvl1'),all=T,suffixes = c('_meth','_unmeth'))

data$orth_id %>% unique %>% length
data %>% head

data$Core=1
data=data[,c('orth_id','KOG_lvl1','Core',colnames(data)[!(colnames(data) %in% c('orth_id','KOG_lvl1','Core'))])]
data[is.na(data)]=0
data=data[data$KOG_lvl1!=0,]

pval_tmp<-run_multihypergeometric_montecarlo_tests(data[,-1],'KOG_lvl1')


# dd=data[data$genome==genome,-c(1,2)] %>%
dd=data[,-c(1)] %>%
  aggregate(. ~ KOG_lvl1,., sum)

rownames(dd)=dd$KOG_lvl1
dd[1]<-NULL
breakdown_tmp<-run_fisher_tests(dd)

pval=pval_tmp
breakdown=breakdown_tmp

breakdown %>% str

breakdown_KOG=breakdown[(breakdown$symbol %in% c('***','**')),]
breakdown_KOG %>% str

breakdown_KOG$comp=factor(breakdown_KOG$comp,levels=levels(breakdown_KOG$comp))

breakdown_KOG_min=breakdown_KOG[breakdown_KOG$comp %in% pval[pval$symbol %in% c('**','***'),c('comp')],]

breakdown_KOG_min=breakdown_KOG_min[breakdown_KOG_min$sign=='enrichment' & breakdown_KOG_min$comp %in% c('Pec..R07B..P08H_meth','Pec..R07B..P08H_unmeth'),] %>% melt(.,id.vars=c('categories','pvalue','diff','symbol','sign'),value.name ='comp')

breakdown_KOG_min[breakdown_KOG_min$comp=='Pec..R07B..P08H_unmeth','diff']=-1*breakdown_KOG_min[breakdown_KOG_min$comp=='Pec..R07B..P08H_unmeth','diff']

ordered_cat=breakdown_KOG_min[,'categories'][order(breakdown_KOG_min[,'diff'])]

breakdown_KOG_min$categories=factor(breakdown_KOG_min$categories,levels=ordered_cat %>% unique)
breakdown_KOG_min$comp= factor(breakdown_KOG_min$comp, levels=c('Pec..R07B..P08H_meth','Pec..R07B..P08H_unmeth'), labels=c('high','low-intermediate')) 

breakdown_KOG_min$comp %>% unique
breakdown_KOG_min %>% str
breakdown_KOG_min$genome='All three species'

breakdown_KOG_min_gene_body=breakdown_KOG_min

#### get Fisher for all data ####

breakdown_KOG2=read.csv('../data/Gene-body_methylation/Fisher_tests_results_genebody.txt',header=T,sep='\t')
# breakdown_KOG2$comp=factor(breakdown_KOG2$comp,levels=levels(breakdown_KOG$comp))

breakdown_KOG2_min=breakdown_KOG2[breakdown_KOG2$sign=='enrichment' & breakdown_KOG2$categories!='not annotated',]

breakdown_KOG2_min[breakdown_KOG2_min$comp=='low-intermediate','diff']=-1*breakdown_KOG2_min[breakdown_KOG2_min$comp=='low-intermediate','diff']

colnames(breakdown_KOG_min)
colnames(breakdown_KOG2_min)

breakdown_KOG3_min=rbind(breakdown_KOG_min[,intersect(colnames(breakdown_KOG_min),colnames(breakdown_KOG2_min))],breakdown_KOG2_min[,intersect(colnames(breakdown_KOG_min),colnames(breakdown_KOG2_min))])


categories_order=breakdown_KOG3_min %>% group_by(categories) %>% summarise(mean=mean(diff)) %>% as.data.frame() %>%  .[order(.$mean),'categories'] %>%  unique

breakdown_KOG3_min$categories=factor(breakdown_KOG3_min$categories,levels=categories_order)
#### plot ####
plot_KOG=ggplot(breakdown_KOG_min)+
  geom_bar(aes(x=diff,y=categories,fill=comp),color='grey',stat='identity')+
  labs(y='KOG functional categories', fill='',x='normalised differences between expected and observed\nrelative abundances of functional categories')+
  # scale_fill_manual(values=c("Pec..R07B..P08H_meth"='red',"Pec..R07B..P08H_unmeth"='blue'),labels=c('Orthologs methylated in all species','Orthologs unmethylated in all species'))+
  facet_wrap(~genome)+
  guides(fill=guide_legend(title.position="top", title.hjust = 0,nrow=1))+
  scale_x_continuous(labels = abs)+
  theme_bw()+
  theme(
        strip.background.x=element_blank(),
        strip.text.y=element_text(angle=0),
        legend.position='top')

plot_KOG

```

```{r KEGG functional enrichment of orthologs, eval=FALSE, include=FALSE}

shortest_string <- function(s) {
  s=s[which(s!='---NA---')]
  s[which.min(nchar(s))]
}
presence <- function(x) {
  if(length(x) > 0) {1}
  else {0}}

brite_hier=read.csv('../data/Assemblies_and_annotations/reference_BRITE_hierarchy_clean.tsv',header=T,sep='\t')
brite_hier %>% str

orth=read.csv('../data/Assemblies_and_annotations/Final_annotations.txt',header=T,sep='\t') %>% .[.$flag=='good_annot',c('genome','qseqid','KO','Description')] %>% unique
colnames(orth)[which(colnames(orth)=='qseqid')]='mRNA'
orth$KO=strsplit(orth$KO, split=', ')
orth=orth %>% unnest(.,KO)
orth=merge(orth,brite_hier,by='KO',all.x=T)
map_of_good_descriptions=aggregate(orth$Description, by=list(orth$KO), FUN = shortest_string) %>% as.data.frame %>% setNames(c("KO",'Description'))
orth=merge(orth[,colnames(orth)[which(colnames(orth)!='Description')]],map_of_good_descriptions,by='KO',all.x=T)
orth %>% str

orth=merge(orth,df[,c('genome','mRNA','orth_id')] %>% unique, by=c('genome','mRNA'), all.y=T) %>% .[,c('orth_id','KO','Description','KEGG_lvl1','KEGG_lvl2','KEGG_longname')] %>% unique

tmp=read.csv('../data/Gene-body_methylation/venn_genebody_methylation_orthologs.txt',header=T,sep='\t')
tmp=merge(tmp,orth,by=c('orth_id'), all.x=T) %>% unique
tmp$L1=factor(tmp$L1, levels=c("Pec","R07B", "P08H", "Pec..R07B", "R07B..P08H", 
 "Pec..P08H", "Pec..R07B..P08H" ))
tmp$meth=factor(tmp$meth, levels=unique(tmp$meth))
tmp %>% str

bad_KEGG_lvl1=c("Brite Hierarchies","Human Diseases", "Not Included in Pathway or Brite") # remove irrelevant KEGG_lvl1 categories
bad_KEGG_lvl2=c("Biosynthesis of other secondary metabolites", "Cellular community - prokaryotes", 
"Information processing in viruses", "Membrane transport", "Metabolism of terpenoids and polyketides", 
"Not included in regular maps") #poorly represented KEGG_lvl2 categories (with less than 50 genes in each worm genomes)

tmp=tmp[!(tmp$KEGG_lvl1 %in% bad_KEGG_lvl1 | tmp$KEGG_lvl2 %in% bad_KEGG_lvl2), ]
tmp %>% head

#######

breakdown=data.frame()
pval=data.frame()
for (group in levels(tmp$meth)){

data=reshape2::dcast(tmp[tmp$meth==group,],orth_id+KEGG_lvl2~L1,fun.aggregate = presence,drop = T)

data$Core=1
data=data[,c('orth_id','KEGG_lvl2','Core',levels(tmp$L1))]

pval_tmp<-run_multihypergeometric_montecarlo_tests(data[,-1],'KEGG_lvl2')

pval_tmp$group=group
pval=rbind(pval,pval_tmp)

# dd=data[data$genome==genome,-c(1,2)] %>%
dd=data[,-c(1)] %>%
  aggregate(. ~ KEGG_lvl2,., sum)

rownames(dd)=dd$KEGG_lvl2
dd[1]<-NULL
breakdown_tmp<-run_fisher_tests(dd)
breakdown_tmp$group=group

breakdown=rbind(breakdown,breakdown_tmp)
}

pval

breakdown %>% str

breakdown$diff2=breakdown$diff
breakdown[breakdown$diff<0, 'diff2' ]=NA
breakdown[breakdown$group=="unmethylated_gene-body" & breakdown$diff>=0, 'diff2']=-1*breakdown[breakdown$group=="unmethylated_gene-body" & breakdown$diff>=0, 'diff']

breakdown_KEGG=breakdown[(breakdown$symbol %in% c('***','**')),]
breakdown_KEGG %>% str

breakdown_KEGG$comp=factor(breakdown_KEGG$comp,levels=levels(breakdown_KEGG$comp))

breakdown_KEGG_min=merge(pval[pval$symbol %in% c('**','***'),c('comp','group')],breakdown_KEGG, by=c('group','comp'),all.x=T)

plot_KEGG<-ggplot(breakdown_KEGG_min)+
  geom_bar(aes(x=diff2,y=categories,fill=group),color='grey',stat='identity')+
  labs(y='KEGG functional categories', fill='',x='normalised differences between expected and observed\nrelative abundances of functional categories')+
  facet_grid(.~comp,scales='free')+
  guides(fill=guide_legend(title.position="top", title.hjust = 0,nrow=1))+
  scale_x_continuous(labels = abs)+
  theme_bw()+
  theme(strip.text.x = element_text(face = 'bold.italic'),
        strip.background.x=element_blank(),
        strip.text.y=element_text(angle=0),
        legend.position='top')

plot_KEGG

breakdown_KEGG %>% str
plot_KEGG2<-ggplot(breakdown_KEGG)+
  geom_bar(aes(x=diff,y=categories,fill=sign),color='grey',stat='identity')+
  labs(y='KEGG functional categories', fill='',x='normalised differences between expected and observed\nrelative abundances of functional categories')+
  facet_grid(group~comp,scales='free')+
  guides(fill=guide_legend(title.position="top", title.hjust = 0,nrow=1))+
  scale_x_continuous(labels = abs)+
  theme_bw()+
  theme(strip.text.x = element_text(face = 'bold.italic'),
        strip.background.x=element_blank(),
        strip.text.y=element_text(angle=0),
        legend.position='top')

plot_KEGG2

tmp %>% str
tmp$KEGG_lvl2 %>% unique

ggarrange(plot_KOG,plot_KEGG,nrow=2)



```


### Promoters

```{r promoter demethylation in Ppalmiformis and tubeworms, eval=FALSE, include=FALSE}
#########################
#### Do the genes that loose promoter methylation in P.palm compared to the tubeworms also gain expression ? Transcriptomic data from R. piscesae are not used because of their low quality.
df %>% str

tmp=read.csv('../data/Gene-body_methylation/venn_genebody_methylation_orthologs.txt',header=T,sep='\t')
tail(tmp)
core_highmeth_orth=tmp[tmp$meth=='methylated_gene-body' & tmp$L1=='Pec..R07B..P08H','orth_id']
  
good_core_highmeth_orth=df[(df$orth_id %in% core_highmeth_orth) & df$binary_idx %in% c('11','01'),c("orth_id","flank_meth_binary","binary_idx")] %>% group_by(orth_id) %>%
  mutate(dupe = n()==3) %>% .[.$dupe==T,] %>% .$orth_id %>% unique # keep only orth with conserved highly methylated gene-body and methylation data for the promoter region

df[(df$orth_id %in% core_highmeth_orth) & df$binary_idx %in% c('11','01'),c("orth_id","flank_meth_binary","binary_idx")] %>% group_by(orth_id) %>%
  mutate(dupe = n()==3) %>% .[.$dupe==F,]

df[df$orth_id=='5084',c("genome","orth_id","flank_meth_binary","binary_idx")]

good_core_highmeth_orth %>% unique %>% length

Pec_11_ortho=(df[df$orth_id %in% good_core_highmeth_orth &
                 df$genome %in% c('P. echinospica') & df$binary_idx=='11',]$orth_id %>% unique)

Pec_01_ortho=(df[df$orth_id %in% good_core_highmeth_orth &
                 df$genome %in% c('P. echinospica') & df$binary_idx=='01',]$orth_id %>% unique)
R07_11_ortho=(df[df$orth_id %in% good_core_highmeth_orth &
                 df$genome %in% c('R. piscesae') & df$binary_idx=='11',]$orth_id %>% unique)
R07_01_ortho=(df[df$orth_id %in% good_core_highmeth_orth & 
                 df$genome %in% c('R. piscesae') & df$binary_idx=='01',]$orth_id %>% unique)
P08_11_ortho=(df[df$orth_id %in% good_core_highmeth_orth & 
                 df$genome %in% c('P. palmiformis') & df$binary_idx=='11',]$orth_id %>% unique)
P08_01_ortho=(df[df$orth_id %in% good_core_highmeth_orth & 
                 df$genome %in% c('P. palmiformis') & df$binary_idx=='01',]$orth_id %>% unique)

c(Pec_01_ortho,Pec_11_ortho,R07_01_ortho,R07_11_ortho,P08_01_ortho,P08_11_ortho) %>% unique %>% length

```

#### venn diagram
```{r venn diagram orthologs promoter methylation, eval=FALSE, include=FALSE}
library(ggVennDiagram)
library(ggtext)


dat_01=list(Pec=Pec_01_ortho,
      R07B=R07_01_ortho,
      P08H=P08_01_ortho)
      
dat_11=list(Pec=Pec_11_ortho,
      R07B=R07_11_ortho,
      P08H=P08_11_ortho)


venn_11 <- Venn(dat_11)
data_11 <- process_data(venn_11)
venn_01 <- Venn(dat_01)
data_01 <- process_data(venn_01)

dat_11 %>% str
venn_11 %>% str
data_11 %>% str

line_colors_bright=c('high promoter methylation'='#e8000b','low promoter methylation'='#ff7c00')
# line_colors_bright=c('high UG'='#8c0800','low U high G'='#ff7c00')

venn_plot_promoter=ggplot() +
  geom_sf(data = venn_region(data_11),fill=NA)+
  geom_sf_text(aes(label = c('P. echinospica','R. piscesae','P. palmiformis')), fontface = "bold.italic", data = venn_setlabel(data_11))+
  geom_sf_text(aes(label=count,color='high promoter methylation'),fontface='bold',nudge_y = 0.5, data = venn_region(data_11)) +
  geom_sf_text(aes(label=count,color='low promoter methylation'),fontface='bold',nudge_y = -0.5, data = venn_region(data_01)) +
  scale_color_manual(values=line_colors_bright,
                     labels = paste("<span style='color:",
                                   line_colors_bright,
                                   "'>",
                                   names(line_colors_bright),
                                   "</span>"),
                     name='',
                     guide=guide_legend(override.aes = list(color = c(NA,NA)) ))+
  coord_sf(xlim = c(-6,11))+
  theme_void()+
  theme(legend.position = c(1,1),
        legend.justification = c(1, 1),
        legend.text = element_markdown(face = 'bold'))

venn_plot_promoter

ggsave('../figures/Venn_orthology_promoter_methylation.png',venn_plot_promoter,width=8.49,height=5.18, units = 'in', dpi=300)

ggsave('../figures/Venn_orthology_promoter_methylation.pdf',venn_plot_promoter,width=8.49,height=5.18, units = 'in')

###########

methylated_venn=data_11@region$item %>% reshape2::melt(value.name = 'orth_id')
methylated_venn$L1=factor(methylated_venn$L1, levels=unique(methylated_venn$L1), labels=data_11@region$name)
methylated_venn$meth='methylated_promoter'
unmethylated_venn=data_01@region$item %>% reshape2::melt(value.name = 'orth_id')
unmethylated_venn$L1=factor(unmethylated_venn$L1, levels=unique(unmethylated_venn$L1), labels=data_01@region$name)
unmethylated_venn$meth='unmethylated_promoter'

methylated_venn %>% group_by(L1) %>% summarise(count=n())
unmethylated_venn %>% group_by(L1) %>% summarise(count=n())

venn_data=rbind(methylated_venn,unmethylated_venn)
venn_data %>% head

write.table(venn_data,'../data/Gene-body_methylation/venn_promoter_methylation_orthologs.txt',col.names = T,sep='\t',row.names = F)


########
venn_plot=egg::ggarrange(venn_plot_genebody+labs(title="Core orthologs (n = 4942)")+
                           theme(plot.title = element_text(hjust = 0.5, vjust=0)),
                         venn_plot_promoter+labs(title="Core orthologs with conserved\nhigh gene-body methylation (n = 2906)")+
                           theme(plot.title = element_text(hjust = 0.5, vjust=0)),nrow = 1,ncol = 2)

venn_plot

ggsave('../figures/Venn_orthology_methylation.png',venn_plot,width=11,height=5.18, units = 'in', dpi=300)

ggsave('../figures/Venn_orthology_methylation.pdf',venn_plot,width=8.49,height=5.18, units = 'in')


```

#### expression
```{r summarise data per genome2, eval=FALSE, include=FALSE}
demet_TPM1=df[df$orth_id %in% c(
  intersect(Pec_11_ortho,P08_11_ortho),
  intersect(Pec_11_ortho,P08_01_ortho),
  intersect(Pec_01_ortho,P08_11_ortho),
  intersect(Pec_01_ortho,P08_01_ortho)
  ),] %>% unique
demet_TPM1$demet=NA
demet_TPM1$demet[demet_TPM1$orth_id %in% intersect(Pec_11_ortho,P08_11_ortho)]='control_met'
demet_TPM1$demet[demet_TPM1$orth_id %in% intersect(Pec_11_ortho,P08_01_ortho)]='demet'
demet_TPM1$demet[demet_TPM1$orth_id %in% intersect(Pec_01_ortho,P08_11_ortho)]='remet'
demet_TPM1$demet[demet_TPM1$orth_id %in% intersect(Pec_01_ortho,P08_01_ortho)]='control_unmet'
demet_TPM1$demet=factor(demet_TPM1$demet,levels=c('control_met','demet','remet','control_unmet'))
demet_TPM1=demet_TPM1[demet_TPM1$genome != 'R. piscesae',]
demet_TPM1 %>% group_by(genome,demet) %>% summarise(count=n()) # gene counts 
demet_TPM1 %>% group_by(genome) %>% summarise(count=n()) # gene counts 
demet_TPM1 %>% nrow

demet_TPM2=df[df$orth_id %in% c(
  intersect(Pec_11_ortho,R07_11_ortho),
  intersect(Pec_11_ortho,R07_01_ortho),
  intersect(Pec_01_ortho,R07_11_ortho),
  intersect(Pec_01_ortho,R07_01_ortho)),] %>% unique
demet_TPM2$demet=NA
demet_TPM2$demet[demet_TPM2$orth_id %in% intersect(Pec_11_ortho,R07_11_ortho)]='control_met'
demet_TPM2$demet[demet_TPM2$orth_id %in% intersect(Pec_11_ortho,R07_01_ortho)]='demet'
demet_TPM2$demet[demet_TPM2$orth_id %in% intersect(Pec_01_ortho,R07_11_ortho)]='remet'
demet_TPM2$demet[demet_TPM2$orth_id %in% intersect(Pec_01_ortho,R07_01_ortho)]='control_unmet'
demet_TPM2$demet=factor(demet_TPM2$demet,levels=c('control_met','demet','remet','control_unmet'))
demet_TPM2=demet_TPM2[demet_TPM2$genome != 'P. palmiformis',]
demet_TPM2 %>% group_by(genome,demet) %>% summarise(count=n()) # gene counts 
demet_TPM2 %>% group_by(genome) %>% summarise(count=n()) # gene counts 
demet_TPM2 %>% nrow

demet_TPM3=df[df$orth_id %in% c(
  intersect(R07_11_ortho,P08_11_ortho),
  intersect(R07_01_ortho,P08_11_ortho),
  intersect(R07_11_ortho,P08_01_ortho),
  intersect(R07_01_ortho,P08_01_ortho)),] %>% unique
demet_TPM3$demet=NA
demet_TPM3$demet[demet_TPM3$orth_id %in% intersect(R07_11_ortho,P08_11_ortho)]='control_met'
demet_TPM3$demet[demet_TPM3$orth_id %in% intersect(R07_01_ortho,P08_11_ortho)]='demet'
demet_TPM3$demet[demet_TPM3$orth_id %in% intersect(R07_11_ortho,P08_01_ortho)]='remet'
demet_TPM3$demet[demet_TPM3$orth_id %in% intersect(R07_01_ortho,P08_01_ortho)]='control_unmet'
demet_TPM3$demet=factor(demet_TPM3$demet,levels=c('control_met','demet','remet','control_unmet'))
demet_TPM3=demet_TPM3[demet_TPM3$genome != 'P. echinospica',]
demet_TPM3 %>% group_by(genome,demet) %>% summarise(count=n())
demet_TPM3 %>% group_by(genome) %>% summarise(count=n())



```
```{r observed exp2, eval=FALSE, include=FALSE}

make_boxplot_plot=function(test1th,sp1,sp2,res,method='KS'){
  library(ggtext)  # remotes::install_github("clauswilke/ggtext")

e <- function(x) {
  r <- quantile(x, probs = c(0.05, 0.25, NA, 0.75, 0.95))
  names(r) <- c("ymin", "lower", "middle", "upper", "ymax")
  r
}

f <- function(x) {
  r <- quantile(x, probs = c(NA, 0.4, 0.5, 0.6, NA))
  names(r) <- c("ymin", "lower", "middle", "upper", "ymax")
  r
}

alpha_scale=c(0.4,0.6)
names(alpha_scale)=c('50','20')
ngenes=test1th[!(is.na(test1th$logfd)),] %>% group_by(demet) %>% summarise(count=n()) %>% as.data.frame

fill_scale=c("#F8766D", "#7CAE00", "#00BFC4", "#C77CFF")
names(fill_scale)=levels(test1th$demet)

fill_scale_labels=c(paste0("hypermethylated in both species (n = ",ngenes[1,2],')'),
                    paste0("hypermethylated in *",sp1,"* only (n = ",ngenes[2,2],')'),
                    paste0('hypermethylated in *',sp2,"* only (n = ",ngenes[3,2],')'),
                    paste0("low-intermediate methylation in both species  (n = ",ngenes[4,2],')'))

res_title=paste0("Summary of pairwise ",method," tests: ",
       paste(c(unlist(rbind(paste("<span style = 'color: ",fill_scale[strsplit(res[res$method==method,'string'],split = " ")[[1]][c(1,3,5,7)]], ";'>",'&#11044;',"</span>"),strsplit(res[res$method==method,'string'],split = " ")[[1]][c(2,4,6,8)]))) %>% .[1:7],collapse=''))

plot=ggplot(test1th)+
    stat_summary(aes(y=logfd, x=demet,fill=demet,alpha='50'),
                 fun.data = e, geom="boxplot",
                 width=0.8,
                 position=position_dodge(1))+
    stat_summary(aes(y=logfd, x=demet,fill=demet,alpha='20'),
                 fun.data = f, geom="boxplot",
                 width=0.8,
                 position=position_dodge(1))+
    # geom_flat_violin(aes(y=logfd, x=demet))+
    scale_y_continuous(name=paste0('Log10 expression (*',sp2,'* / *',sp1,'*)'))+
  scale_x_discrete(limits = rev(levels(test1th$demet)))+
    labs(title=res_title)+
   scale_alpha_manual(values=alpha_scale,name='Interpercentile range')+
  scale_fill_manual(values=fill_scale,labels=fill_scale_labels,name='Methylation state of promoter region')+
    coord_flip(ylim = c(-2,2))+
    guides(alpha = guide_legend(override.aes = list(fill = 'grey20', alpha=c(0.2,0.6))))+
    theme_bw()+
    theme(axis.ticks.y=element_blank(),
      axis.title.y = element_blank(),
          axis.text.y = element_blank(),
          legend.text = element_markdown(),
          axis.title.x = element_markdown(),
      plot.title = element_markdown(size=rel(0.8)))

return(plot)
}
make_plot=function(test1th,sp1,sp2){
  demet_labels=c('control_unmet'='Unmethylated in both species','remet'='Methylated in species 2 only','demet'='Methylated in species 1 only','control_met'='Methylated in both species')
  ngenes=test1th %>% nrow
  t1 <- test1th %>% group_by(demet) %>% slice_sample(n=10)

  n=0.5
  plot1 <- ggplot(t1,aes(x=demet, ymin=species1, ymax=species2, group = row.names(t1)))+
    geom_linerange(position = position_dodge(n))+
    geom_point(aes(y=species1,color=paste0('species 1 : ', sp1)),position = position_dodge(n),alpha=1,size=2)+
    geom_point(aes(y=species2,color=paste0('species 2 : ', sp2)),position = position_dodge(n),alpha=1,size=2)+
    scale_y_continuous(trans=pseudo_log_trans(),name='Expression (TPM)')+
    scale_x_discrete(name='Gene-body methylation profiles\nof orthologous genes\nacross species',labels=demet_labels)+
    scale_color_discrete(name='')+
    labs(title='Expression of representative orthologous genes across species\n(random sample of 10 genes per category amongst expressed genes)')+
    coord_flip()+
    theme_bw()+
    theme(legend.position='bottom',
          plot.title = element_text(size=rel(0.8)))

source("https://raw.githubusercontent.com/datavizpyr/data/master/half_flat_violinplot.R")
  plot2<-ggplot(test1th)+
    geom_flat_violin(aes(y=logfd, x=demet))+
    scale_y_continuous(name='Log10 expression (species 2/species 1)')+
    labs(title=paste0('Distribution of expression fold change of orthologous genes across\nspecies (all orthologous genes expressed in both species, n = ',ngenes,')'))+
    coord_flip()+
    theme_bw()+
    theme(axis.title.y = element_blank(),
          axis.text.y = element_blank(),
          plot.title = element_text(size=rel(0.8)))

  plot=egg::ggarrange(plot1,plot2,nrow = 1,ncol = 2)

  return(plot)

  }
run_KS_test= function(fold_diff_test1){
cols=unique(fold_diff_test1$demet)
diff_test_higher=matrix(rep(1,length(cols)^2),ncol=length(cols),nrow=length(cols)) # is x>y? yes if pval <0.01
diff_test_lower=matrix(rep(1,length(cols)^2),ncol=length(cols),nrow=length(cols)) # is x<y? yes if pval <0.01

colnames(diff_test_higher)=cols
rownames(diff_test_higher)=cols
colnames(diff_test_lower)=cols
rownames(diff_test_lower)=cols

#### KS test ####
pairs=expand.grid(cols,cols)
for (i in c(1:nrow(pairs))){
l1=pairs[i,1]
l2=pairs[i,2]

set1=fold_diff_test1[fold_diff_test1$demet==l1,'logfd']
set2=fold_diff_test1[fold_diff_test1$demet==l2,'logfd']
val=ks.test(set1,set2,alternative='greater')
val=val$p.value

diff_test_lower[l1,l2]<-val
}
for (i in c(1:nrow(pairs))){
l1=pairs[i,1]
l2=pairs[i,2]

set1=fold_diff_test1[fold_diff_test1$demet==l1,'logfd']
set2=fold_diff_test1[fold_diff_test1$demet==l2,'logfd']
val=ks.test(set1,set2,alternative='less')
val=val$p.value

diff_test_higher[l2,l1]<-val
}

#### sumarise results  ####
thresh=0.05
diff_test_lower=ifelse(diff_test_lower<thresh,1,0)
diff_test_higher=ifelse(diff_test_higher<thresh,1,0)

# print(diff_test_higher*diff_test_lower)

str_list=rowSums(diff_test_higher*diff_test_lower)[rev(order(rowSums(diff_test_higher*diff_test_lower)))] %>% as.data.frame
colnames(str_list)[1]='count'
str_list$id=rownames(str_list)

str_list=str_list %>%
  group_by(count) %>%
  summarise(loc=min(count), id=paste(id, collapse=" = "))


return(t(c(comp=fold_diff_test1$comp[1],pair=fold_diff_test1$pair[1],string=paste0(rev(str_list$id),collapse =" < "))))
}
run_wilcox_test= function(fold_diff_test1){
cols=unique(fold_diff_test1$demet)
diff_test_higher=matrix(rep(1,length(cols)^2),ncol=length(cols),nrow=length(cols)) # is x>y? yes if pval <0.01
diff_test_lower=matrix(rep(1,length(cols)^2),ncol=length(cols),nrow=length(cols)) # is x<y? yes if pval <0.01

colnames(diff_test_higher)=cols
rownames(diff_test_higher)=cols
colnames(diff_test_lower)=cols
rownames(diff_test_lower)=cols

#### wilcox ####
pairs=expand.grid(cols,cols)
wilcox_res=data.frame()
for (i in c(1:nrow(pairs))){
l1=pairs[i,1]
l2=pairs[i,2]

set1=fold_diff_test1[fold_diff_test1$demet==l1,'logfd']
set2=fold_diff_test1[fold_diff_test1$demet==l2,'logfd']
val=wilcox.test(set1,set2,alternative='less')
val=val$p.value

diff_test_lower[l1,l2]<-val
}
for (i in c(1:nrow(pairs))){
l1=pairs[i,1]
l2=pairs[i,2]

set1=fold_diff_test1[fold_diff_test1$demet==l1,'logfd']
set2=fold_diff_test1[fold_diff_test1$demet==l2,'logfd']
val=wilcox.test(set1,set2,alternative='greater')
val=val$p.value

diff_test_higher[l2,l1]<-val
}
diff_test_lower=ifelse(diff_test_lower<0.01,1,0)
diff_test_higher=ifelse(diff_test_higher<0.01,1,0)

### summarise results ####
str_list=rowSums(diff_test_higher*diff_test_lower)[rev(order(rowSums(diff_test_higher*diff_test_lower)))] %>% as.data.frame
colnames(str_list)[1]='count'
str_list$id=rownames(str_list)

str_list=str_list %>%
  group_by(count) %>%
  summarise(loc=min(count), id=paste(id, collapse=" = "))

return(t(c(comp=fold_diff_test1$comp[1],pair=fold_diff_test1$pair[1],string=paste0(rev(str_list$id),collapse =" < ")))) 

}
run_test=function(test1,filter,sp1,sp2){
  run_KS_wilcox_tests=function(test1){
  ks_res=as.data.frame(run_KS_test(test1))
  wilcox_res=as.data.frame(run_wilcox_test(test1))
  ks_res$method='KS'
  wilcox_res$method='Wilcox'
  
  res=rbind(ks_res,wilcox_res)
  res$filter=filter
  return(res)
  }

  res=run_KS_wilcox_tests(test1)
  plot=make_boxplot_plot(test1,sp1,sp2,res)

  return(list('data'=test1,'plot'=plot,'res'=res))
}

pseudocounts=0.0000001
pseudocounts=0
##############

sp1='P. echinospica'
sp2='P. palmiformis'
test1th=demet_TPM1[!(is.na(demet_TPM1$TPM_value)),] %>% dcast(.,demet+orth_id~genome,value.var='TPM_value', drop=T)
test1th$comp=paste0(sp2,'/',sp1)
test1th=test1th[,c('demet','orth_id',sp1,sp2,'comp')]
colnames(test1th)=c("demet", "orth_id", "species1", "species2",'comp')

test1th$logfd=log10((test1th$`species2`+pseudocounts)/(test1th$species1+pseudocounts))


test1th %>% tail(n=10)
test1th %>% str
test1th[!(is.na(test1th$logfd)),] %>% group_by(demet,.drop = T) %>% summarise(count=n())
test1th[(is.na(test1th$species1) & is.na(test1th$species2)),] %>% group_by(demet,.drop = T) %>% summarise(count=n())
test1th[,] %>% group_by(demet,.drop = T) %>% summarise(count=n())
test1th[,] %>% group_by(demet,.drop = T) %>% summarise(count=n())

test1th %>% head

filter='none'
t1=run_test(test1th,filter,sp1,sp2)
t1$res
t1$plot

# filter='TPM>0'
# test1=test1th[!(is.na(test1th$species1) & is.na(test1th$species2)) & test1th$species1>0 & test1th$species2>0,] %>% .[!(is.na(.$demet)),]
# t1=run_test(test1,filter,sp1,sp2)
# t1$res
# t1$plot


ggsave('../figures/PecPpalm_orthologs_promoter_methylation.png',t1$plot,width=10,height=3, units = 'in', dpi=300)


```

#### comparisons across taxonomic groups 

```{r summarise data per taxa, eval=FALSE, include=FALSE}

tubeworms_11_ortho=intersect(Pec_11_ortho,R07_11_ortho)
tubeworms_01_ortho=intersect(Pec_01_ortho,R07_01_ortho)
end_11_ortho=intersect(R07_11_ortho,P08_11_ortho)
end_01_ortho=intersect(R07_01_ortho,P08_01_ortho)

intersect(Pec_01_ortho,Pec_11_ortho) #orthologs with both promoter methylation profiles in Pec (when an ortholog is represented by more than one gene)
intersect(R07_01_ortho,R07_11_ortho)
intersect(P08_01_ortho,P08_11_ortho)

promoter_demethylation_in_P08=intersect(tubeworms_11_ortho,P08_01_ortho) 
promoter_nodemethylation_in_P08=intersect(tubeworms_11_ortho,P08_11_ortho) 
promoter_remethylation_in_P08=intersect(tubeworms_01_ortho,P08_11_ortho) 
promoter_noremethylation_in_P08=intersect(tubeworms_01_ortho,P08_01_ortho) 

promoter_demethylation_in_R07=intersect(Pec_11_ortho,R07_01_ortho) 
promoter_nodemethylation_in_R07=intersect(Pec_11_ortho,R07_11_ortho) 
promoter_remethylation_in_R07=intersect(Pec_01_ortho,R07_11_ortho)
promoter_noremethylation_in_R07=intersect(Pec_01_ortho,R07_01_ortho) 

promoter_demethylation_in_end=intersect(Pec_11_ortho,end_01_ortho)
promoter_nodemethylation_in_end=intersect(Pec_11_ortho,end_11_ortho)
promoter_remethylation_in_end=intersect(Pec_01_ortho,end_11_ortho) 
promoter_noremethylation_in_end=intersect(Pec_01_ortho,end_01_ortho) 

#### TPM1 ####
demet_TPM1=df[df$orth_id %in% c(promoter_nodemethylation_in_P08,promoter_demethylation_in_P08,promoter_remethylation_in_P08,promoter_noremethylation_in_P08),] %>% unique
demet_TPM1$demet=NA
demet_TPM1$demet[demet_TPM1$orth_id %in% promoter_nodemethylation_in_P08]='control_met'
demet_TPM1$demet[demet_TPM1$orth_id %in% promoter_demethylation_in_P08]='demet'
demet_TPM1$demet[demet_TPM1$orth_id %in% promoter_remethylation_in_P08]='remet'
demet_TPM1$demet[demet_TPM1$orth_id %in% promoter_noremethylation_in_P08]='control_unmet'
demet_TPM1$demet=factor(demet_TPM1$demet,levels=c('control_met','demet','remet','control_unmet'))
# demet_TPM1=demet_TPM1[demet_TPM1$genome != 'R. piscesae',]
demet_TPM1 %>% group_by(genome,demet) %>% summarise(count=n()) # gene counts for orthologs in each genome and group


#### TPM2 ####

demet_TPM2=df[df$orth_id %in% c(promoter_demethylation_in_R07, promoter_nodemethylation_in_R07,promoter_remethylation_in_R07,promoter_noremethylation_in_R07),] %>% unique
demet_TPM2$demet=NA
demet_TPM2$demet[demet_TPM2$orth_id %in% promoter_nodemethylation_in_R07]='control_met'
demet_TPM2$demet[demet_TPM2$orth_id %in% promoter_demethylation_in_R07]='demet'
demet_TPM2$demet[demet_TPM2$orth_id %in% promoter_remethylation_in_R07]='remet'
demet_TPM2$demet[demet_TPM2$orth_id %in% promoter_noremethylation_in_R07]='control_unmet'
demet_TPM2$demet=factor(demet_TPM2$demet,levels=c('control_met','demet','remet','control_unmet'))
demet_TPM2=demet_TPM2[demet_TPM2$genome != 'P. palmiformis',]
demet_TPM2 %>% group_by(genome,demet) %>% summarise(count=n()) 
# bad_orth=demet_TPM2 %>% 
#   group_by(genome, orth_id) %>% 
#   mutate(dupe = n()>1) %>% .[.$dupe==T,] %>% unique %>% .$orth_id
# demet_TPM2=demet_TPM2[!(demet_TPM2$orth_id %in% bad_orth),]


#### TPM3 ####
demet_TPM3=df[df$orth_id %in% c(promoter_demethylation_in_end, promoter_nodemethylation_in_end,promoter_remethylation_in_end,promoter_noremethylation_in_end),] %>% unique
demet_TPM3$demet=NA
demet_TPM3$demet[demet_TPM3$orth_id %in% promoter_nodemethylation_in_end]='control_met'
demet_TPM3$demet[demet_TPM3$orth_id %in% promoter_demethylation_in_end]='demet'
demet_TPM3$demet[demet_TPM3$orth_id %in% promoter_remethylation_in_end]='remet'
demet_TPM3$demet[demet_TPM3$orth_id %in% promoter_noremethylation_in_end]='control_unmet'
demet_TPM3$demet=factor(demet_TPM3$demet,levels=c('control_met','demet','remet','control_unmet'))
# demet_TPM3=demet_TPM3[demet_TPM3$genome != 'R. piscesae',]
# demet_TPM3 %>% group_by(genome,demet) %>% summarise(count=n()) 
# bad_orth=demet_TPM3 %>% 
#   group_by(genome, orth_id) %>% 
#   mutate(dupe = n()>1) %>% .[.$dupe==T,] %>% unique %>% .$orth_id
# demet_TPM3=demet_TPM3[!(demet_TPM3$orth_id %in% bad_orth),]


```

```{r param distrib figures2, eval=FALSE, include=FALSE}
e <- function(x) {
  r <- quantile(x, probs = c(0.0025, 0.25, 0.5, 0.75, 0.9975))
  names(r) <- c("ymin", "lower", "middle", "upper", "ymax")
  r
}

library('ggimage')


#### plot tpm1 ####
demet_TPM1=demet_TPM1[demet_TPM1$genome != 'R. piscesae',]

demet_TPM1[demet_TPM1$demet=='control_met','image']='../figures/c11.svg'
demet_TPM1[demet_TPM1$demet=='control_unmet','image']='../figures/c01.svg'
demet_TPM1[demet_TPM1$demet=='demet' & demet_TPM1$genome=='P. echinospica','image']='../figures/e11.svg'
demet_TPM1[demet_TPM1$demet=='demet' & demet_TPM1$genome!='P. echinospica','image']='../figures/p01.svg'
demet_TPM1[demet_TPM1$demet=='remet' & demet_TPM1$genome=='P. echinospica','image']='../figures/e01.svg'
demet_TPM1[demet_TPM1$demet=='remet' & demet_TPM1$genome!='P. echinospica','image']='../figures/p11.svg'

count_TPM1=demet_TPM1 %>% group_by(genome,demet) %>% summarise(count=n()) %>% as.data.frame
count_TPM1_control_met=count_TPM1[count_TPM1$demet=='control_met','count'] %>% unique %>% as.character
count_TPM1_control_unmet=count_TPM1[count_TPM1$demet=='control_unmet','count'] %>% unique %>% as.character
count_TPM1_demet=count_TPM1[count_TPM1$demet=='demet','count'] %>% unique %>% as.character
count_TPM1_remet=count_TPM1[count_TPM1$demet=='remet','count'] %>% unique %>% as.character

p11<-ggplot(demet_TPM1,fill=demet,color=demet)+
  geom_image(aes(y=10000,x=demet, image=image,fill=demet),size=0.2,by='width')+
  scale_size_identity()+
  stat_summary(aes(x=demet,y=TPM_value,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=TPM_value),shape=21,color='black',fill=NA,width=0.2,alpha=0.1)+
  scale_y_continuous(trans=scales::pseudo_log_trans(base = 10),
                     labels = label_number(accuracy = 1),
                     breaks=c(1,10,100,1000,10000),
                     minor_breaks=c(0,5,50,500,5000),
                     expand = c(0,0.6),
                     name='Gene expression\n(TPM)')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved promoter methylation (n=',count_TPM1_control_met,')'),
    paste0('with unmethylated promoter in *P. palmiformis* (n=',count_TPM1_demet,')'),
    paste0('with unmethylated promoter in tubeworms (n=',count_TPM1_remet,')'),
    paste0('with conserved  lack of promoter methylation (n=',count_TPM1_control_unmet,')')),
    name='Highly methylated gene orthologs:')+
guides(color=guide_legend(title.position="top", title.hjust = 0,nrow=2))+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_text(face = 'bold.italic'),
        strip.background = element_blank(),
        legend.position='top',
        legend.text=element_markdown())

demet_TPM1 %>% str

p12<-ggplot(demet_TPM1)+
  stat_summary(aes(x=demet,y=cds_Length,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=cds_Length),shape=21,color='black',fill=NA,width=0.2,alpha=0.1)+
  scale_y_continuous(trans=scales::pseudo_log_trans(base = 10),
                     labels = label_number(accuracy = 1),
                     breaks=c(1,10,100,1000,10000),
                     minor_breaks=c(0,5,50,500,5000),
                     name='cds length (bp)')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved promoter methylation (n=',count_TPM1_control_met,')'),
    paste0('with unmethylated promoter in P. palmiformis (n=',count_TPM1_demet,')'),
    paste0('with unmethylated promoter in tubeworms (n=',count_TPM1_remet,')'),
    paste0('with conserved  lack of promoter methylation (n=',count_TPM1_control_unmet,')')),
    name='Highly methylated gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
          legend.position='none')

p13<-ggplot(demet_TPM1)+
  stat_summary(aes(x=demet,y=flank_CG*100,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=flank_CG*100),shape=21,color='black',fill=NA,width=0.2,alpha=0.1)+
  scale_y_continuous(name='promoter CG%')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved promoter methylation (n=',count_TPM1_control_met,')'),
    paste0('with unmethylated promoter in P. palmiformis (n=',count_TPM1_demet,')'),
    paste0('with unmethylated promoter in tubeworms (n=',count_TPM1_remet,')'),
    paste0('with conserved  lack of promoter methylation (n=',count_TPM1_control_unmet,')')),
    name='Highly methylated gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
        legend.position='none')

p14<-ggplot(demet_TPM1)+
  stat_summary(aes(x=demet,y=flank_CpGoe,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=flank_CpGoe),shape=21,color='black',fill=NA,width=0.2,alpha=0.1)+
  scale_y_continuous(name='Promoter\nCpG obs/exp')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved promoter methylation (n=',count_TPM1_control_met,')'),
    paste0('with unmethylated promoter in P. palmiformis (n=',count_TPM1_demet,')'),
    paste0('with unmethylated promoter in tubeworms (n=',count_TPM1_remet,')'),
    paste0('with conserved  lack of promoter methylation (n=',count_TPM1_control_unmet,')')),
    name='Highly methylated gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
        legend.position='none')
#### plot tpm2 ####

demet_TPM2[demet_TPM2$demet=='control_met','image']='../figures/c11.svg'
demet_TPM2[demet_TPM2$demet=='control_unmet','image']='../figures/c01.svg'
demet_TPM2[demet_TPM2$demet=='demet' & demet_TPM2$genome=='P. echinospica','image']='../figures/e11.svg'
demet_TPM2[demet_TPM2$demet=='demet' & demet_TPM2$genome!='P. echinospica','image']='../figures/p01.svg'
demet_TPM2[demet_TPM2$demet=='remet' & demet_TPM2$genome=='P. echinospica','image']='../figures/e01.svg'
demet_TPM2[demet_TPM2$demet=='remet' & demet_TPM2$genome!='P. echinospica','image']='../figures/p11.svg'

count_TPM2=demet_TPM2 %>% group_by(genome,demet) %>% summarise(count=n()) %>% as.data.frame
count_TPM2_control_met=count_TPM2[count_TPM2$demet=='control_met','count'] %>% unique %>% as.character
count_TPM2_control_unmet=count_TPM2[count_TPM2$demet=='control_unmet','count'] %>% unique %>% as.character
count_TPM2_demet=count_TPM2[count_TPM2$demet=='demet','count'] %>% unique %>% as.character
count_TPM2_remet=count_TPM2[count_TPM2$demet=='remet','count'] %>% unique %>% as.character

p21<-ggplot(demet_TPM2,fill=demet,color=demet)+
  geom_image(aes(y=10000,x=demet, image=image,fill=demet),size=0.2,by='width')+
  scale_size_identity()+
  stat_summary(aes(x=demet,y=TPM_value,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=TPM_value),shape=21,color='black',fill=NA,width=0.2,alpha=0.01)+
  scale_y_continuous(trans=scales::pseudo_log_trans(base = 10),
                     labels = label_number(accuracy = 1),
                     breaks=c(1,10,100,1000,10000),
                     minor_breaks=c(0,5,50,500,5000),
                     expand = c(0,0.6),
                     name='Gene expression\n(TPM)')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved promoter methylation (n=',count_TPM2_control_met,')'),
    paste0('with unmethylated promoter in *R. piscesae* (n=',count_TPM2_demet,')'),
    paste0('with unmethylated promoter in *P. echinospica* (n=',count_TPM2_remet,')'),
    paste0('with conserved  lack of promoter methylation (n=',count_TPM2_control_unmet,')')),
    name='Highly methylated gene orthologs:')+
guides(color=guide_legend(title.position="top", title.hjust = 0,nrow=2))+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_text(face = 'bold.italic'),
        strip.background = element_blank(),
        legend.position='top',
        legend.text=element_markdown())

p22<-ggplot(demet_TPM2)+
  stat_summary(aes(x=demet,y=cds_Length,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=cds_Length),shape=21,color='black',fill=NA,width=0.2,alpha=0.01)+
  scale_y_continuous(trans=scales::pseudo_log_trans(base = 10),
                     labels = label_number(accuracy = 1),
                     breaks=c(1,10,100,1000,10000),
                     minor_breaks=c(0,5,50,500,5000),
                     name='cds length (bp)')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved promoter methylation (n=',count_TPM2_control_met,')'),
    paste0('with unmethylated promoter in R. piscesae (n=',count_TPM2_demet,')'),
    paste0('with unmethylated promoter in P. echinospica (n=',count_TPM2_remet,')'),
    paste0('with conserved  lack of promoter methylation (n=',count_TPM2_control_unmet,')')),
    name='Highly methylated gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
          legend.position='none')

p23<-ggplot(demet_TPM2)+
  stat_summary(aes(x=demet,y=flank_CG*100,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=flank_CG*100),shape=21,color='black',fill=NA,width=0.2,alpha=0.01)+
  scale_y_continuous(name='promoter CG%')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved promoter methylation (n=',count_TPM2_control_met,')'),
    paste0('with unmethylated promoter in R. piscesae (n=',count_TPM2_demet,')'),
    paste0('with unmethylated promoter in P. echinospica (n=',count_TPM2_remet,')'),
    paste0('with conserved  lack of promoter methylation (n=',count_TPM2_control_unmet,')')),
    name='Highly methylated gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
        legend.position='none')

p24<-ggplot(demet_TPM2)+
  stat_summary(aes(x=demet,y=flank_CpGoe,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=flank_CpGoe),shape=21,color='black',fill=NA,width=0.2,alpha=0.01)+
  scale_y_continuous(name='Promoter\nCpG obs/exp')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved promoter methylation (n=',count_TPM2_control_met,')'),
    paste0('with unmethylated promoter in R. piscesae (n=',count_TPM2_demet,')'),
    paste0('with unmethylated promoter in P. echinospica (n=',count_TPM2_remet,')'),
    paste0('with conserved  lack of promoter methylation (n=',count_TPM2_control_unmet,')')),
    name='Highly methylated gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
        legend.position='none')


#### plot tpm3 ####
demet_TPM3=demet_TPM3[demet_TPM3$genome != 'R. piscesae',]

demet_TPM3[demet_TPM3$demet=='control_met','image']='../figures/c11.svg'
demet_TPM3[demet_TPM3$demet=='control_unmet','image']='../figures/c01.svg'
demet_TPM3[demet_TPM3$demet=='demet' & demet_TPM3$genome=='P. echinospica','image']='../figures/e11.svg'
demet_TPM3[demet_TPM3$demet=='demet' & demet_TPM3$genome!='P. echinospica','image']='../figures/p01.svg'
demet_TPM3[demet_TPM3$demet=='remet' & demet_TPM3$genome=='P. echinospica','image']='../figures/e01.svg'
demet_TPM3[demet_TPM3$demet=='remet' & demet_TPM3$genome!='P. echinospica','image']='../figures/p11.svg'
count_TPM3=demet_TPM3 %>% group_by(genome,demet) %>% summarise(count=n()) %>% as.data.frame
count_TPM3_control_met=count_TPM3[count_TPM3$demet=='control_met','count'] %>% unique %>% as.character
count_TPM3_control_unmet=count_TPM3[count_TPM3$demet=='control_unmet','count'] %>% unique %>% as.character
count_TPM3_demet=count_TPM3[count_TPM3$demet=='demet','count'] %>% unique %>% as.character
count_TPM3_remet=count_TPM3[count_TPM3$demet=='remet','count'] %>% unique %>% as.character


p31<-ggplot(demet_TPM3,fill=demet,color=demet)+
  geom_image(aes(y=10000,x=demet, image=image,fill=demet),size=0.2,by='width')+
  scale_size_identity()+
  stat_summary(aes(x=demet,y=TPM_value,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=TPM_value),shape=21,color='black',fill=NA,width=0.2,alpha=0.01)+
  scale_y_continuous(trans=scales::pseudo_log_trans(base = 10),
                     labels = label_number(accuracy = 1),
                     breaks=c(1,10,100,1000,10000),
                     minor_breaks=c(0,5,50,500,5000),
                     expand = c(0,0.6),
                     name='Gene expression\n(TPM)')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved promoter methylation (n=',count_TPM3_control_met,')'),
    paste0('with unmethylated promoter in vents worms (n=',count_TPM3_demet,')'),
    paste0('with unmethylated promoter in *P. echinospica* (n=',count_TPM3_remet,')'),
    paste0('with conserved  lack of promoter methylation (n=',count_TPM3_control_unmet,')')),
    name='Highly methylated gene orthologs:')+
guides(color=guide_legend(title.position="top", title.hjust = 0,nrow=2))+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_text(face = 'bold.italic'),
        strip.background = element_blank(),
        legend.position='top',
        legend.text=element_markdown())

p32<-ggplot(demet_TPM3)+
  stat_summary(aes(x=demet,y=cds_Length,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=cds_Length),shape=21,color='black',fill=NA,width=0.2,alpha=0.01)+
  scale_y_continuous(trans=scales::pseudo_log_trans(base = 10),
                     labels = label_number(accuracy = 1),
                     breaks=c(1,10,100,1000,10000),
                     minor_breaks=c(0,5,50,500,5000),
                     name='cds length (bp)')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved promoter methylation (n=',count_TPM3_control_met,')'),
    paste0('with unmethylated promoter in vent worms (n=',count_TPM3_demet,')'),
    paste0('with unmethylated promoter in P. echinospica (n=',count_TPM3_remet,')'),
    paste0('with conserved  lack of promoter methylation (n=',count_TPM3_control_unmet,')')),
    name='Highly methylated gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
          legend.position='none')

p33<-ggplot(demet_TPM3)+
  stat_summary(aes(x=demet,y=flank_CG*100,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=flank_CG*100),shape=21,color='black',fill=NA,width=0.2,alpha=0.01)+
  scale_y_continuous(name='promoter CG%')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved promoter methylation (n=',count_TPM3_control_met,')'),
    paste0('with unmethylated promoter in vent worms (n=',count_TPM3_demet,')'),
    paste0('with unmethylated promoter in P. echinospica (n=',count_TPM3_remet,')'),
    paste0('with conserved  lack of promoter methylation (n=',count_TPM3_control_unmet,')')),
    name='Highly methylated gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
        legend.position='none')

p34<-ggplot(demet_TPM3)+
  stat_summary(aes(x=demet,y=flank_CpGoe,color=demet),
                 fun.data = e, geom="boxplot",
                 fatten = 1)+
  geom_jitter(aes(x=demet,y=flank_CpGoe),shape=21,color='black',fill=NA,width=0.2,alpha=0.01)+
  scale_y_continuous(name='Promoter\nCpG obs/exp')+
  scale_color_manual(values=c('#1b9e77','#d95f02','#e7298a','#7570b3'),labels=c(
    paste0('with conserved promoter methylation (n=',count_TPM3_control_met,')'),
    paste0('with unmethylated promoter in vent worms (n=',count_TPM3_demet,')'),
    paste0('with unmethylated promoter in P. echinospica (n=',count_TPM3_remet,')'),
    paste0('with conserved  lack of promoter methylation (n=',count_TPM3_control_unmet,')')),
    name='Highly methylated gene orthologs:')+
  facet_grid(.~genome)+
  theme_bw()+
  theme(axis.title.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        strip.text=element_blank(),
        strip.background = element_blank(),
        legend.position='none')



#### PLOT ####
plot_demet1<-egg::ggarrange(p11,p12,p13,p14, nrow=4, ncol=1,heights = c(3, 2, 2, 2))
plot_demet2<-egg::ggarrange(p21,p22,p23,p24, nrow=4, ncol=1,heights = c(3, 2, 2, 2))
plot_demet3<-egg::ggarrange(p31,p32,p33,p34, nrow=4, ncol=1,heights = c(3, 2, 2, 2))


ggsave('../figures/Tbws_vs_Ppalm_orthology_promoter_methylation.png',plot_demet1, width=7.5,height=6, units = 'in',dpi=300)
ggsave('../figures/Tbws_vs_Ppalm_orthology_promoter_methylation.pdf',plot_demet1, width=7.5,height=6, units = 'in')


ggsave('../figures/Pec_vs_Rid_orthology_promoter_methylation.png',plot_demet2, width=7.5,height=6, units = 'in',dpi=300)
ggsave('../figures/Pec_vs_Rid_orthology_promoter_methylation.pdf',plot_demet2, width=7.5,height=6, units = 'in')

# ggsave('../figures/Pec_vs_endeavour_orthology_promoter_methylation.png',plot_demet3, width=7.5,height=6, units = 'in',dpi=300)
# ggsave('../figures/Pec_vs_endeavour_orthology_promoter_methylation.pdf',plot_demet3, width=7.5,height=6, units = 'in')

```

```{r genetic-epigenetic coupling2, eval=FALSE, include=FALSE}
demet_TPM1 %>% str
demet_TPM1$conservation=ifelse(demet_TPM1$demet %in% c('control_met','control_unmet'),'conserved','unconserved')
demet_TPM1$comp='inter-fam'
demet_TPM2$conservation=ifelse(demet_TPM2$demet %in% c('control_met','control_unmet'),'conserved','unconserved')
demet_TPM2$comp='tubeworms'
demet_TPM3$conservation=ifelse(demet_TPM3$demet %in% c('control_met','control_unmet'),'conserved','unconserved')
demet_TPM3$comp='inter-habitat'


#### concatenate data frames ####
demet_TPM=rbind(demet_TPM1,demet_TPM2)

demet_TPM$group=paste(demet_TPM$comp,demet_TPM$conservation,sep='_')
demet_TPM %>% str

demet_TPM %>% group_by(comp,conservation) %>% summarise(count=n())
demet_TPM %>% group_by(comp,conservation,flank_meth_binary) %>% summarise(count=n())


e <- function(x) {
  r <- quantile(x, probs = c(0.0025, 0.25, 0.5, 0.75, 0.9975))
  names(r) <- c("ymin", "lower", "middle", "upper", "ymax")
  r
}

demet_TPM$group=demet_TPM$group=factor(demet_TPM$group,
                     levels=c("tubeworms_unconserved","tubeworms_conserved",
                              "inter-fam_unconserved", "inter-fam_conserved"),
                     labels=c('Contrasting methylation\namongst tubeworms',
                             "Conserved methylation\namongst tubeworms",
                             'Contrasting methylation\nacross families',
                             "Conserved methylation\nacross families"))

inspect=demet_TPM[,c('group','comp','genome','orth_id','cds_Length')] %>% dcast(group+orth_id~genome,value.var = 'cds_Length')

inspect$rel=round(apply(inspect[,c(3:5)], 1,FUN=sd,na.rm=T)/inspect$`P. echinospica`*100)

# % standard deviation of cds_length of orthologs between tubeworms and Pec normalised to Pec cds length
ggplot(inspect,aes(x=rel))+
  stat_ecdf(geom = "step")+
  scale_x_continuous(trans=pseudo_log_trans(),limits=c(0,100))+
  facet_wrap(~group)  

inspect %>% group_by(group) %>% summarise(quant=quantile(rel,probs = c(0.25, 0.5, 0.75, 0.8,0.9,0.95,0.99,1), na.rm=T)) %>% as.data.frame()

inspect %>% group_by(group) %>% summarise(quant=quantile(rel,probs = c(0.85), na.rm=T)) %>% as.data.frame() #85% of othologs have 40% or less product length difference

inspect %>% group_by(group) %>% summarise(quant=quantile(rel,probs = c(0.75), na.rm=T)) %>% as.data.frame() #75% of othologs have 27% or less product length difference

df_n=demet_TPM %>% group_by(comp,conservation,flank_meth_binary) %>% summarise(count=n()) %>% as.data.frame

textsize=ggplot2::.pt

p0<-ggplot(demet_TPM)+
  stat_summary(aes(x=as.factor(flank_meth_binary),y=flank_CpGoe, color=group),
               fun.data = e, geom="boxplot",
               fatten = 1)+
  geom_jitter(aes(x=as.factor(flank_meth_binary),y=flank_CpGoe),
              shape=21,color='black',fill=NA,width=0.2,alpha=0.02)+
  geom_text(data=df_n,aes(x=as.factor(flank_meth_binary),y=2,label=paste0('n = ',as.character(count))),size=textsize,vjust=0,color='grey20')+
  scale_x_discrete(labels=c("Low","High"), name='Promoter methylation level')+
  scale_y_continuous(name='Promoter CpG observed/expected')+
  facet_grid(comp~conservation,
             labeller = labeller(conservation=setNames(c("Contrasting\npromoter methylation",
                                                         "Conserved\npromoter methylation"),
                                                       c('unconserved', 'conserved')),
                                 comp=setNames(c("Across families","Amongst tubeworms","Across habitats"),
                                               c("inter-fam", 'tubeworms','inter-habitat'))))+
  coord_flip()+
  theme_bw()+
  theme(legend.position='none',
        axis.title.x=element_text(size=rel(0.8)),
        axis.title.y=element_text(size=rel(0.8)))

p0

#### 1000 bootstraps ####

set.seed(2022)    # for reproduciblility
iter=1000
tot_res=data.frame()
demet_TPM$group=paste(demet_TPM$comp,demet_TPM$conservation,sep='_')

demet_TPM$group=demet_TPM$group=factor(demet_TPM$group,
                     levels=c("tubeworms_unconserved","tubeworms_conserved",
                              "inter-fam_unconserved", "inter-fam_conserved"),
                     labels=c('Contrasting methylation\namongst tubeworms',
                             "Conserved methylation\namongst tubeworms",
                             'Contrasting methylation\nacross families',
                             "Conserved methylation\nacross families"))

sample_count=demet_TPM %>% group_by(genome,group) %>% summarise(count = n()) # get number of samples
orth_id_tbl=demet_TPM[demet_TPM$genome=='P. echinospica',c('comp','conservation','group','orth_id')] # get orth_id table


orth_id_tbl %>% group_by(group) %>% summarise(count=n())

library('purrr')

for (i in (1:iter)){
  if(i%%100==0){print(i)}
  
  tmp=demet_TPM
  
  orth_id_map=orth_id_tbl %>% 
  nest(-group) %>% 
  left_join(sample_count, by = "group") %>%
  mutate(Sample = map2(data, count, replace = T, sample_n)) %>%
  unnest(Sample,names_repair='universal') %>% .[,c('orth_id','comp','conservation')] # select randomly (with replacement) as many ortholog ids as in the original dataset.

  # orth_id_map %>% group_by(comp,conservation) %>% summarise(count=n())
  # orth_id_map[duplicated(orth_id_map),]
  # orth_id_map[orth_id_map$orth_id=='3837',]
  
  tmp=merge(tmp[,colnames(tmp)[colnames(tmp)!='conservation']],orth_id_map,by=c('orth_id','comp'),all.y=T) #get data for the bootstrapped orthologs id
  
  
  res=tmp %>% nest(-group) %>% 
  mutate(fit = purrr::map(data, ~ glm(as.numeric(flank_meth_binary) ~ flank_CpGoe, data = .,family = binomial)),
         results = map(fit, glance)) %>% 
  unnest(results) %>% 
  summarise(group=group,r.squared=1 - deviance/null.deviance) 
  
  res$i=i
  tot_res=rbind(tot_res,res)
}

tot_res %>% summary

write.table(tot_res,'../data/Gene-body_methylation/Bootstrap_results_promoter_coupling.txt',col.names=T,row.names = F,sep='\t',quote = T)

ggplot(tot_res)+
  geom_histogram(aes(x=r.squared))+
  facet_grid(.~group, scales='free')

error_bar=tot_res %>% group_by(group) %>% summarise(sd=sd(r.squared),mean=mean(r.squared),sdmin=mean(r.squared)-sd(r.squared),sdmax=mean(r.squared)+sd(r.squared))

res_glm=demet_TPM %>% nest(-group) %>% 
  mutate(fit = map(data, ~ glm(as.numeric(flank_meth_binary) ~ flank_CpGoe, data = .,family = binomial)),
         results = map(fit, glance)) %>% 
  unnest(results) %>% 
  summarise(group=group,r.squared=1 - deviance/null.deviance) %>% as.data.frame

# ggplot(res_glm)+
#   geom_bar(aes(x=r.squared,y=group,fill=group),stat='identity')


res_glm=merge(res_glm,error_bar ,by='group',all.x=T)

p1<-ggplot(res_glm,aes(y=group,xmin=sdmin,xmax=sdmax))+
  geom_bar(aes(fill=group,x=mean),stat='identity')+
  # geom_bar(aes(fill=group,,x=r.squared),stat='identity')+
  geom_errorbar(width = 0.5)+
  labs(x='Strength of correlation between\npromoter methylation and its genetic signature')+
  theme_bw()+
  theme(axis.title.y=element_blank(),
        legend.position='none',
        axis.title.x=element_text(size=rel(0.8)))
p1

#### mount plot ####

plot=egg::ggarrange(p0,p1,ncol=2)

plot

ggsave('../figures/promoterCpGoe_vs_promotermeth.png',plot,width=9,height=3.5, units = 'in',dpi=300)
ggsave('../figures/promoterCpGoe_vs_promotermeth.pdf',plot,width=9,height=3.5, units = 'in')
```

```{r}
#### inspect bootstraps ####

tot_res_gene=read.csv('../data/Gene-body_methylation/Bootstrap_results_gene-body_coupling.txt',header=T,sep='\t',allowEscapes = T)
tot_res_gene$region='gene-body'

tot_res_promoter=read.csv('../data/Gene-body_methylation/Bootstrap_results_promoter_coupling.txt',header=T,sep='\t',allowEscapes = F)
tot_res_promoter$region='promoter'

tot_res2=rbind(tot_res_gene,tot_res_promoter)

tot_res2 %>% summary
tot_res2 %>% str

ggplot(tot_res2)+
  geom_histogram(aes(x=r.squared))+
  facet_grid(region~group, scales='free')

ggplot(tot_res2)+
  geom_bar(aes(x=r.squared,y=region,color=group))+
  theme_bw()

error_bar=tot_res2 %>% group_by(region,group) %>% summarise(sd=sd(r.squared),mean=mean(r.squared),sdmin=mean(r.squared)-sd(r.squared),sdmax=mean(r.squared)+sd(r.squared))


coupling_plot=ggplot(error_bar,aes(y=rev(group),xmin=sdmin,xmax=sdmax))+
  geom_bar(aes(fill=group,x=mean),stat='identity')+
  # geom_bar(aes(fill=group,,x=r.squared),stat='identity')+
  geom_errorbar(width = 0.5)+
  labs(x='Strength of correlation between\nmethylation level and its genetic signature')+
  guides(fill=guide_legend(nrow=2,byrow=F))+
  facet_grid(region~.)+
  theme_bw()+
  theme(axis.title.y=element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        legend.position='top',
        legend.title = element_blank(),
        axis.title.x=element_text(size=rel(0.8)),
        strip.text.y = element_text(angle = 0,face = 'bold'),
        strip.background = element_blank())

coupling_plot

ggsave('../figures/coupling_CpGoe_vs_rmeth.png',coupling_plot,width=5,height=3.5, units = 'in',dpi=300)
ggsave('../figures/coupling_CpGoe_vs_meth.pdf',coupling_plot,width=5,height=3.5, units = 'in')


tot_res3=tot_res2
tot_res3$group=paste(tot_res2$region,tot_res2$group)
run_KS_test= function(tot_res){
cols=unique(tot_res$group)
diff_test_higher=matrix(rep(1,length(cols)^2),ncol=length(cols),nrow=length(cols)) # is x>y? yes if pval <0.01
diff_test_lower=matrix(rep(1,length(cols)^2),ncol=length(cols),nrow=length(cols)) # is x<y? yes if pval <0.01

colnames(diff_test_higher)=cols
rownames(diff_test_higher)=cols
colnames(diff_test_lower)=cols
rownames(diff_test_lower)=cols

#### KS test ####
pairs=expand.grid(cols,cols)
for (i in c(1:nrow(pairs))){
l1=pairs[i,1]
l2=pairs[i,2]

set1=tot_res[tot_res$group==l1,'r.squared'] %>% .[!(is.na(.))]
set2=tot_res[tot_res$group==l2,'r.squared'] %>% .[!(is.na(.))]
val=ks.test(set1,set2,alternative='greater')
val=val$p.value

diff_test_lower[l1,l2]<-val
}
for (i in c(1:nrow(pairs))){
l1=pairs[i,1]
l2=pairs[i,2]

set1=tot_res[tot_res$group==l1,'r.squared'] %>% .[!(is.na(.))]
set2=tot_res[tot_res$group==l2,'r.squared'] %>% .[!(is.na(.))]
val=ks.test(set1,set2,alternative='less')
val=val$p.value

diff_test_higher[l2,l1]<-val
}

#### sumarise results  ####
thresh=0.01
diff_test_lower=ifelse(diff_test_lower<thresh,1,0)
diff_test_higher=ifelse(diff_test_higher<thresh,1,0)

# print(diff_test_higher*diff_test_lower)

str_list=rowSums(diff_test_higher*diff_test_lower)[rev(order(rowSums(diff_test_higher*diff_test_lower)))] %>% as.data.frame
colnames(str_list)[1]='count'
str_list$id=rownames(str_list)

str_list=str_list %>%
  group_by(count) %>%
  summarise(loc=min(count), id=paste(id, collapse=" = "))


return(t(c(comp=tot_res$comp[1],pair=tot_res$pair[1],string=paste0(rev(str_list$id),collapse =" < "))))
}

pairwise_KS_test_results=run_KS_test(tot_res3)



```
#### KOG/KEGG fisher tests on core orthologs

```{r KOG functional enrichment of orthologs core promoter, eval=FALSE, include=FALSE}
#### prepare data frame ####

shortest_string <- function(s) {
  s=s[which(s!='---NA---')]
  s[which.min(nchar(s))]
}
presence <- function(x) {
  if(length(x) > 0) {1}
  else {0}}

kog_hier=read.csv('../data/Assemblies_and_annotations/reference_KOG_hierarchy.txt',header=T,sep='\t',col.names = c('KOG_lvl1_id','KOG','KOG_lvl1'))
kog_hier %>% head
orth=read.csv('../data/Assemblies_and_annotations/Final_annotations.txt',header=T,sep='\t') %>% .[.$flag=='good_annot',c('genome','qseqid','KOG','Description')] %>% unique
colnames(orth)[which(colnames(orth)=='qseqid')]='mRNA'
orth$KOG=strsplit(orth$KOG, split=', ')
orth=orth %>% unnest(.,KOG)
map_of_good_descriptions=aggregate(orth$Description, by=list(orth$KOG), FUN = shortest_string) %>% as.data.frame %>% setNames(c("KOG",'Description'))
orth=merge(orth[,colnames(orth)[which(colnames(orth)!='Description')]],map_of_good_descriptions,by='KOG',all.x=T)
orth=merge(orth,kog_hier, by=c('KOG'), all.x=T)
orth %>% head

orth=merge(orth,df[,c('genome','mRNA','orth_id')] %>% unique, by=c('genome','mRNA'), all.y=T) %>% .[,c('orth_id','KOG','Description','KOG_lvl1')] %>% unique

tmp=read.csv('../data/Gene-body_methylation/venn_promoter_methylation_orthologs.txt',header=T,sep='\t')
tmp=merge(tmp,orth,by=c('orth_id')) %>% unique
tmp$L1=factor(tmp$L1, levels=c("Pec","R07B", "P08H", "Pec..R07B", "R07B..P08H", 
 "Pec..P08H", "Pec..R07B..P08H" ))
tmp$meth=factor(tmp$meth, levels=unique(tmp$meth))
tmp %>% str

breakdown=data.frame()
pval=data.frame()

group=levels(tmp$meth)[[2]]

#### run hypergeometric test and fisher tests for each promoter methylation condition ####

for (group in levels(tmp$meth)){

# levels(tmp$L1)

data=reshape2::dcast(tmp[tmp$meth==group,],orth_id+KOG_lvl1~L1,fun.aggregate = presence,drop = T)

data$Core=1
data=data[,c('orth_id','KOG_lvl1','Core',levels(tmp$L1))]

pval_tmp<-run_multihypergeometric_montecarlo_tests(data[,-1],'KOG_lvl1')

pval_tmp$group=group
pval=rbind(pval,pval_tmp)

# dd=data[data$genome==genome,-c(1,2)] %>%
dd=data[,-c(1)] %>%
  aggregate(. ~ KOG_lvl1,., sum)

rownames(dd)=dd$KOG_lvl1
dd[1]<-NULL
breakdown_tmp<-run_fisher_tests(dd)
breakdown_tmp$group=group

breakdown=rbind(breakdown,breakdown_tmp)
}

pval

breakdown %>% str

breakdown$diff2=breakdown$diff
breakdown[breakdown$diff<0, 'diff2' ]=NA
breakdown[breakdown$group=="unmethylated_promoter" & breakdown$diff>=0, 'diff2']=-1*breakdown[breakdown$group=="unmethylated_promoter" & breakdown$diff>=0, 'diff']

breakdown_KOG=breakdown[(breakdown$symbol %in% c('***','**','*')),]
breakdown_KOG %>% str

breakdown_KOG$comp=factor(breakdown_KOG$comp,levels=levels(breakdown_KOG$comp))

pval$group=rownames(pval)

dat=merge(pval[pval$symbol %in% c('**','***'),c('comp','group')],breakdown_KOG, by.x=c('group'),by.y=c('comp'),all.x=T)

# plot_KOG<-ggplot(breakdown_KOG[!(breakdown_KOG$comp %in% c('Pec','R07B','P08H')),])+
plot_KOG<-ggplot(dat)+
  geom_bar(aes(x=diff2,y=categories,fill=group),color='grey',stat='identity')+
  labs(y='KOG functional categories', fill='',x='normalised differences between expected and observed\nrelative abundances of functional categories')+
  facet_grid(.~comp,scales='free')+
  guides(fill=guide_legend(title.position="top", title.hjust = 0,nrow=1))+
  scale_x_continuous(labels = abs)+
  theme_bw()+
  theme(strip.text.x = element_text(face = 'bold.italic'),
        strip.background.x=element_blank(),
        strip.text.y=element_text(angle=0),
        legend.position='top')

plot_KOG

# breakdown_KOG[!(breakdown_KOG$comp %in% c('Pec','R07B','P08H')),]
plot_KOG2<-ggplot(breakdown_KOG)+
  geom_bar(aes(x=diff,y=categories,fill=group),color='grey',stat='identity')+
  labs(y='KOG functional categories', fill='',x='normalised differences between expected and observed\nrelative abundances of functional categories')+
  facet_grid(group~comp,scales='free')+
  guides(fill=guide_legend(title.position="top", title.hjust = 0,nrow=1))+
  scale_x_continuous(labels = abs)+
  theme_bw()+
  theme(strip.text.x = element_text(face = 'bold.italic'),
        strip.background.x=element_blank(),
        strip.text.y=element_text(angle=0),
        legend.position='top')

plot_KOG2

tmp %>% str
tmp$KOG_lvl1 %>% unique

inspect=tmp[tmp$L1 %in% c('R07B..P08H') &
            tmp$KOG_lvl1 %in% c("Translation, ribosomal structure and biogenesis " ,"RNA processing and modification ") &
             tmp$meth %in% c('unmethylated_promoter'), c('orth_id','Description')] %>% unique


```

```{r KEGG functional enrichment of orthologs core promoter, eval=FALSE, include=FALSE}

shortest_string <- function(s) {
  s=s[which(s!='---NA---')]
  s[which.min(nchar(s))]
}
presence <- function(x) {
  if(length(x) > 0) {1}
  else {0}}

brite_hier=read.csv('../data/Assemblies_and_annotations/reference_BRITE_hierarchy_clean.tsv',header=T,sep='\t')
brite_hier %>% str

orth=read.csv('../data/Assemblies_and_annotations/Final_annotations.txt',header=T,sep='\t') %>% .[.$flag=='good_annot',c('genome','qseqid','KO','Description')] %>% unique
colnames(orth)[which(colnames(orth)=='qseqid')]='mRNA'
orth$KO=strsplit(orth$KO, split=', ')
orth=orth %>% unnest(.,KO)
orth=merge(orth,brite_hier,by='KO',all.x=T)
map_of_good_descriptions=aggregate(orth$Description, by=list(orth$KO), FUN = shortest_string) %>% as.data.frame %>% setNames(c("KO",'Description'))
orth=merge(orth[,colnames(orth)[which(colnames(orth)!='Description')]],map_of_good_descriptions,by='KO',all.x=T)
orth %>% str

orth=merge(orth,df[,c('genome','mRNA','orth_id')] %>% unique, by=c('genome','mRNA'), all.y=T) %>% .[,c('orth_id','KO','Description','KEGG_lvl1','KEGG_lvl2','KEGG_longname')] %>% unique

tmp=read.csv('../data/Gene-body_methylation/venn_promoter_methylation_orthologs.txt',header=T,sep='\t')
tmp=merge(tmp,orth,by=c('orth_id'), all.x=T) %>% unique
tmp$L1=factor(tmp$L1, levels=c("Pec","R07B", "P08H", "Pec..R07B", "R07B..P08H", 
 "Pec..P08H", "Pec..R07B..P08H" ))
tmp$meth=factor(tmp$meth, levels=unique(tmp$meth))
tmp %>% str


# KEGG_lvl2=c("seq_id","start", "end", "CpGoe","CpGs_count", "mRNA", 
# "callCpGs_count", "meth_meandepth", "meth_meancov", "meth_meanfreq", 
#  "gene_Length",'cds_Length', "genome",'KEGG_lvl1','KEGG_lvl2')

bad_KEGG_lvl1=c("Brite Hierarchies","Human Diseases", "Not Included in Pathway or Brite") # remove irrelevant KEGG_lvl1 categories
bad_KEGG_lvl2=c("Biosynthesis of other secondary metabolites", "Cellular community - prokaryotes", 
"Information processing in viruses", "Membrane transport", "Metabolism of terpenoids and polyketides", 
"Not included in regular maps") #poorly represented KEGG_lvl2 categories (with less than 50 genes in each worm genomes)

tmp=tmp[!(tmp$KEGG_lvl1 %in% bad_KEGG_lvl1 | tmp$KEGG_lvl2 %in% bad_KEGG_lvl2), ]
tmp %>% head

#######

breakdown=data.frame()
pval=data.frame()
for (group in levels(tmp$meth)){

data=reshape2::dcast(tmp[tmp$meth==group,],orth_id+KEGG_lvl2~L1,fun.aggregate = presence,drop = T)

data$Core=1
data=data[,c('orth_id','KEGG_lvl2','Core',levels(tmp$L1))]

pval_tmp<-run_multihypergeometric_montecarlo_tests(data[,-1],'KEGG_lvl2')

pval_tmp$group=group
pval=rbind(pval,pval_tmp)

# dd=data[data$genome==genome,-c(1,2)] %>%
dd=data[,-c(1)] %>%
  aggregate(. ~ KEGG_lvl2,., sum)

rownames(dd)=dd$KEGG_lvl2
dd[1]<-NULL
breakdown_tmp<-run_fisher_tests(dd)
breakdown_tmp$group=group

breakdown=rbind(breakdown,breakdown_tmp)
}

pval

breakdown %>% str

breakdown$diff2=breakdown$diff
breakdown[breakdown$diff<0, 'diff2' ]=NA
breakdown[breakdown$group=="unmethylated_promoter" & breakdown$diff>=0, 'diff2']=-1*breakdown[breakdown$group=="unmethylated_promoter" & breakdown$diff>=0, 'diff']

breakdown_KEGG=breakdown[(breakdown$symbol %in% c('***','**','*')),]
breakdown_KEGG %>% str

breakdown_KEGG$comp=factor(breakdown_KEGG$comp,levels=levels(breakdown_KEGG$comp))

dat_KEGG=merge(pval[pval$symbol %in% c('**','***'),c('comp','group')],breakdown_KEGG, by=c('group','comp'),all.x=T)
# breakdown_KEGG[!(breakdown_KEGG$comp %in% c('Pec','R07B','P08H')),]
plot_KEGG<-ggplot(breakdown_KEGG)+
  geom_bar(aes(x=diff2,y=categories,fill=group),color='grey',stat='identity')+
  labs(y='KEGG functional categories', fill='',x='normalised differences between expected and observed\nrelative abundances of functional categories')+
  facet_grid(.~comp,scales='free')+
  guides(fill=guide_legend(title.position="top", title.hjust = 0,nrow=1))+
  scale_x_continuous(labels = abs)+
  theme_bw()+
  theme(strip.text.x = element_text(face = 'bold.italic'),
        strip.background.x=element_blank(),
        strip.text.y=element_text(angle=0),
        legend.position='top')

# plot_KEGG

breakdown_KEGG %>% str
plot_KEGG2<-ggplot(breakdown_KEGG)+
  geom_bar(aes(x=diff,y=categories,fill=group),color='grey',stat='identity')+
  labs(y='KEGG functional categories', fill='',x='normalised differences between expected and observed\nrelative abundances of functional categories')+
  facet_grid(group~comp,scales='free')+
  guides(fill=guide_legend(title.position="top", title.hjust = 0,nrow=1))+
  scale_x_continuous(labels = abs)+
  theme_bw()+
  theme(strip.text.x = element_text(face = 'bold.italic'),
        strip.background.x=element_blank(),
        strip.text.y=element_text(angle=0),
        legend.position='top')

# plot_KEGG2
tmp %>% str
tmp$KEGG_lvl2 %>% unique

ggarrange(plot_KOG2,plot_KEGG2,nrow=2)

tmp$KEGG_lvl2 %>% unique 

inspect=tmp[tmp$L1 %in% c('R07B..P08H') &
            tmp$KOG_lvl1 %in% c("Cell growth and death" ,"RNA processing and modification ") &
             tmp$meth %in% c('unmethylated_promoter'), c('orth_id','Description')] %>% unique


```

```{r inspect, eval=FALSE, include=FALSE}
shortest_string <- function(s) {
  s=s[which(s!='---NA---')]
  s[which.min(nchar(s))]
}
presence <- function(x) {
  if(length(x) > 0) {1}
  else {0}}

kog_hier=read.csv('../data/Assemblies_and_annotations/reference_KOG_hierarchy.txt',header=T,sep='\t',col.names = c('KOG_lvl1_id','KOG','KOG_lvl1'))
kog_hier %>% head
orth=read.csv('../data/Assemblies_and_annotations/Final_annotations.txt',header=T,sep='\t') %>% .[.$flag=='good_annot',c('genome','qseqid','KOG','Description')] %>% unique
colnames(orth)[which(colnames(orth)=='qseqid')]='mRNA'
orth$KOG=strsplit(orth$KOG, split=', ')
orth=orth %>% unnest(.,KOG)
map_of_good_descriptions=aggregate(orth$Description, by=list(orth$KOG), FUN = shortest_string) %>% as.data.frame %>% setNames(c("KOG",'Description'))
orth=merge(orth[,colnames(orth)[which(colnames(orth)!='Description')]],map_of_good_descriptions,by='KOG',all.x=T)
orth=merge(orth,kog_hier, by=c('KOG'), all.x=T)
orth %>% head

orth=merge(orth,df[,c('genome','mRNA','orth_id')] %>% unique, by=c('genome','mRNA'), all.y=T) %>% .[,c('orth_id','KOG','Description','KOG_lvl1')] %>% unique

tmp=read.csv('../data/Gene-body_methylation/venn_promoter_methylation_orthologs.txt',header=T,sep='\t')
tmp=merge(tmp,orth,by=c('orth_id'),all.x=T) %>% unique
tmp$L1=factor(tmp$L1, levels=c("Pec","R07B", "P08H", "Pec..R07B", "R07B..P08H", 
 "Pec..P08H", "Pec..R07B..P08H" ))
tmp$meth=factor(tmp$meth, levels=unique(tmp$meth))
tmp %>% str


brite_hier=read.csv('../data/Assemblies_and_annotations/reference_BRITE_hierarchy_clean.tsv',header=T,sep='\t')
orth=read.csv('../data/Assemblies_and_annotations/Final_annotations.txt',header=T,sep='\t') %>% .[.$flag=='good_annot',c('genome','qseqid','KO','Description')] %>% unique
colnames(orth)[which(colnames(orth)=='qseqid')]='mRNA'
orth$KO=strsplit(orth$KO, split=', ')
orth=orth %>% unnest(.,KO)
orth=merge(orth,brite_hier,by='KO',all.x=T)
map_of_good_descriptions=aggregate(orth$Description, by=list(orth$KO), FUN = shortest_string) %>% as.data.frame %>% setNames(c("KO",'Description'))
orth=merge(orth[,colnames(orth)[which(colnames(orth)!='Description')]],map_of_good_descriptions,by='KO',all.x=T)
orth %>% str

orth=merge(orth,df[,c('genome','mRNA','orth_id')] %>% unique, by=c('genome','mRNA'), all.y=T) %>% .[,c('orth_id','KO','Description','KEGG_lvl1','KEGG_lvl2','KEGG_longname')] %>% unique

tmp=merge(tmp,orth,by=c('orth_id','Description'), all.x=T) %>% unique
tmp$L1=factor(tmp$L1, levels=c("Pec","R07B", "P08H", "Pec..R07B", "R07B..P08H", 
 "Pec..P08H", "Pec..R07B..P08H" ))
tmp$meth=factor(tmp$meth, levels=unique(tmp$meth))
tmp %>% str


# KEGG_lvl2=c("seq_id","start", "end", "CpGoe","CpGs_count", "mRNA", 
# "callCpGs_count", "meth_meandepth", "meth_meancov", "meth_meanfreq", 
#  "gene_Length",'cds_Length', "genome",'KEGG_lvl1','KEGG_lvl2')

bad_KEGG_lvl1=c("Brite Hierarchies","Human Diseases", "Not Included in Pathway or Brite") # remove irrelevant KEGG_lvl1 categories
bad_KEGG_lvl2=c("Biosynthesis of other secondary metabolites", "Cellular community - prokaryotes", 
"Information processing in viruses", "Membrane transport", "Metabolism of terpenoids and polyketides", 
"Not included in regular maps") #poorly represented KEGG_lvl2 categories (with less than 50 genes in each worm genomes)

tmp=tmp[!(tmp$KEGG_lvl1 %in% bad_KEGG_lvl1 | tmp$KEGG_lvl2 %in% bad_KEGG_lvl2), ]
tmp %>% head

tmp$KOG_lvl1 %>% unique 
tmp$KEGG_lvl2 %>% unique 
tmp$L1 %>% unique 

orth_to_inspect=tmp[tmp$meth %in% c('methylated_promoter') &
            tmp$L1 %in% c('Pec..R07B') &
            (tmp$KOG_lvl1 %in% c("Cytoskeleton ","Signal transduction mechanisms ") |
            tmp$KEGG_lvl2 %in% c("Signal transduction","Cellular community - eukaryotes")),'orth_id'] %>% unique

orth_to_inspect=tmp[tmp$meth %in% c('methylated_promoter') &
            tmp$L1 %in% c('Pec..R07B') &
            (tmp$KOG_lvl1 %in% c("Signal transduction mechanisms ") |
            tmp$KEGG_lvl2 %in% c("Signal transduction")),'orth_id'] %>% unique

inspect=tmp[tmp$meth %in% c('methylated_promoter') &
    tmp$L1 %in% c('Pec..R07B') &
    tmp$orth_id %in% orth_to_inspect,c('orth_id','Description')] %>% .[complete.cases(.),] %>% unique


```

```{r fig-promoter-methylation-fisher-KOG-deprecated, eval=FALSE, fig.align='center', fig.cap='KOG functional enrichment of genes orthologs with unmethylated (low U high G) and methylated (high UG) promoters as compared to all highly methylated genes in each species. Only significant enrichments (Fisher test pvalue < 0.05) are represented.', fig.height=4, fig.scap='KOG functional enrichment of highly methylated gene orthologs with methylated and unmethylated promoters.', fig.width=8, message=FALSE, warning=FALSE, fig=FALSE, include=FALSE}

knitr::include_graphics("../figures/promotermethylation_fisher_KOG.png")

```

## Perspective

```{r eval=FALSE, include=FALSE}
library(boot)
x <- rnorm(5000,mean=0,sd=1)
##then the SE for the mean is 1/sqrt(5000)
se <- 1/sqrt(5000)

```



